{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Training machine learning models on pairs of substrates in individual organisms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from subpred.util import load_df\n",
    "from subpred.graph import preprocess_data, get_substrate_matrix\n",
    "from subpred.pssm import calculate_pssm_feature\n",
    "from subpred.compositions import calculate_aac, calculate_paac\n",
    "import pandas as pd\n",
    "from subpred.cdhit import cd_hit\n",
    "from sklearn.pipeline import make_pipeline\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.preprocessing import StandardScaler, LabelEncoder\n",
    "from sklearn.feature_selection import SelectPercentile\n",
    "from sklearn.model_selection import GridSearchCV, train_test_split\n",
    "from sklearn.metrics import classification_report\n",
    "from subpred.custom_transformers import FeatureCombinator, get_feature_type_combinations\n",
    "\n",
    "from subpred.util import load_df\n",
    "import networkx as nx\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_classification_task(\n",
    "    organism_ids: set,\n",
    "    labels: set,\n",
    "    clustering_threshold: int = None,\n",
    "    dataset_folder_path: str = \"../data/datasets\",\n",
    ") -> pd.DataFrame:\n",
    "    # TODO handling for multi-substrate\n",
    "    # TODO ability to use go terms or chebi terms (compare sample count, performance)\n",
    "\n",
    "    (\n",
    "        df_uniprot,\n",
    "        df_uniprot_goa,\n",
    "        graph_go_filtered,\n",
    "        graph_chebi_filtered,\n",
    "    ) = preprocess_data(\n",
    "        organism_ids=organism_ids, datasets_folder_path=dataset_folder_path\n",
    "    )\n",
    "    # TODO go through method code\n",
    "    df_substrate_overlaps, dict_chebi_to_uniprot = get_substrate_matrix(\n",
    "        datasets_folder_path=dataset_folder_path,\n",
    "        graph_chebi=graph_chebi_filtered,\n",
    "        graph_go=graph_go_filtered,\n",
    "        df_uniprot_goa=df_uniprot_goa,\n",
    "        min_overlap=0,\n",
    "        max_overlap=int(1e6),\n",
    "    )\n",
    "    assert df_substrate_overlaps.shape[0] == len(dict_chebi_to_uniprot.keys())\n",
    "    chebi_name_to_term = {\n",
    "        name: term for term, name in graph_chebi_filtered.nodes(data=\"name\")\n",
    "    }\n",
    "    chebi_term_to_name = {\n",
    "        term: name for term, name in graph_chebi_filtered.nodes(data=\"name\")\n",
    "    }\n",
    "    molecule_counts = {\n",
    "        chebi_term_to_name[term]: len(proteins)\n",
    "        for term, proteins in dict_chebi_to_uniprot.items()\n",
    "    }\n",
    "    print(sorted(molecule_counts.items(), key=lambda item: item[1], reverse=True))\n",
    "\n",
    "    protein_to_label = list()\n",
    "    for label in labels:\n",
    "        label_proteins = dict_chebi_to_uniprot[chebi_name_to_term[label]]\n",
    "        for protein in label_proteins:\n",
    "            protein_to_label.append([protein, label])\n",
    "\n",
    "    df_labels = pd.DataFrame.from_records(\n",
    "        protein_to_label, columns=[\"Uniprot\", \"label\"], index=\"Uniprot\"\n",
    "    )\n",
    "\n",
    "    df_labels = df_labels[~df_labels.index.duplicated()]  # TODO series?\n",
    "    print(df_labels.label.value_counts())\n",
    "    df_sequences = df_uniprot.loc[df_labels.index].sequence.to_frame()\n",
    "    print(\"number of sequences\", df_sequences.shape[0])\n",
    "    if clustering_threshold:\n",
    "        cluster_representatives = cd_hit(\n",
    "            df_sequences.sequence, identity_threshold=clustering_threshold\n",
    "        )\n",
    "        print(cluster_representatives)\n",
    "        df_sequences = df_sequences.loc[cluster_representatives]\n",
    "        df_labels = df_labels.loc[cluster_representatives]\n",
    "    return pd.concat([df_sequences, df_labels], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_features(series_sequences:pd.Series):\n",
    "    df_aac = calculate_aac(series_sequences)\n",
    "    df_paac = calculate_paac(series_sequences)\n",
    "    df_pssm_50_1 = calculate_pssm_feature(\n",
    "        series_sequences,\n",
    "        tmp_folder=\"../data/intermediate/blast/pssm_uniref50_1it\",\n",
    "        blast_db=\"../data/raw/uniref/uniref50/uniref50.fasta\",\n",
    "        iterations=1,\n",
    "        psiblast_threads=-1,\n",
    "        verbose=False,\n",
    "        feature_name=\"PSSM_50_1\"\n",
    "    )\n",
    "    df_pssm_50_3 = calculate_pssm_feature(\n",
    "        series_sequences,\n",
    "        tmp_folder=\"../data/intermediate/blast/pssm_uniref50_3it\",\n",
    "        blast_db=\"../data/raw/uniref/uniref50/uniref50.fasta\",\n",
    "        iterations=3,\n",
    "        psiblast_threads=-1,\n",
    "        verbose=False,\n",
    "        feature_name=\"PSSM_50_3\"\n",
    "    )\n",
    "    df_pssm_90_1 = calculate_pssm_feature(\n",
    "        series_sequences,\n",
    "        tmp_folder=\"../data/intermediate/blast/pssm_uniref90_3it\",\n",
    "        blast_db=\"../data/raw/uniref/uniref90/uniref90.fasta\",\n",
    "        iterations=1,\n",
    "        psiblast_threads=-1,\n",
    "        verbose=False,\n",
    "        feature_name=\"PSSM_90_1\"\n",
    "    )\n",
    "    df_pssm_90_3 = calculate_pssm_feature(\n",
    "        series_sequences,\n",
    "        tmp_folder=\"../data/intermediate/blast/pssm_uniref90_3it\",\n",
    "        blast_db=\"../data/raw/uniref/uniref90/uniref90.fasta\",\n",
    "        iterations=3,\n",
    "        psiblast_threads=-1,\n",
    "        verbose=False,\n",
    "        feature_name=\"PSSM_90_3\"\n",
    "    )\n",
    "    df_features = pd.concat(\n",
    "        [\n",
    "            df_aac,\n",
    "            df_paac,\n",
    "            df_pssm_50_1,\n",
    "            df_pssm_50_3,\n",
    "            df_pssm_90_1,\n",
    "            df_pssm_90_3,\n",
    "        ], axis=1\n",
    "    )\n",
    "    return df_features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Eval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO try removing worst sample according to percentages\n",
    "# TODO feature selection, regularization\n",
    "# TODO cd-hit\n",
    "# TODO determinism\n",
    "# TODO also comparative analysis of features?\n",
    "# TODO compare to protein embeddings and BLAST\n",
    "# TODO parameter for using featurecombinator\n",
    "# TODO separate functions\n",
    "\n",
    "def evaluate(df_dataset, df_features):\n",
    "\n",
    "    # converting data to numpy\n",
    "    label_encoder = LabelEncoder()\n",
    "    label_encoder.fit(sorted(df_dataset.label.unique()))\n",
    "    sample_names = df_features.index.values\n",
    "    feature_names = df_features.columns.values\n",
    "    X = df_features.values\n",
    "    y = label_encoder.transform(df_dataset.label)\n",
    "    # train test eval split\n",
    "    (\n",
    "        X_train,\n",
    "        X_eval,\n",
    "        y_train,\n",
    "        y_eval,\n",
    "        sample_names_train,\n",
    "        sample_names_eval,\n",
    "    ) = train_test_split(X, y, sample_names, test_size=0.2, random_state=1, stratify=y)\n",
    "\n",
    "    feature_type_combinations = get_feature_type_combinations(feature_names=feature_names)\n",
    "    feature_combinator = FeatureCombinator(feature_names=df_features.columns)\n",
    "    model = make_pipeline(\n",
    "        StandardScaler(), feature_combinator, SVC(random_state=1, probability=True)\n",
    "    )\n",
    "    param_grid = {\n",
    "        \"svc__C\": [0.1, 1, 10],\n",
    "        # \"svc__gamma\": [\"scale\", \"auto\"],\n",
    "        # \"svc__class_weight\": [\"balanced\", None],\n",
    "        \"featurecombinator__feature_types\": feature_type_combinations,\n",
    "        # \"selectpercentile__percentile\": list(range(1, 101, 5)),\n",
    "    }\n",
    "\n",
    "    # hyperparam optim & crossval\n",
    "    gridsearch = GridSearchCV(\n",
    "        estimator=model,\n",
    "        param_grid=param_grid,\n",
    "        scoring=\"f1\",\n",
    "        cv=5,\n",
    "        n_jobs=-1,\n",
    "        return_train_score=True,\n",
    "        # verbose=20\n",
    "    )\n",
    "    gridsearch.fit(X_train, y_train)\n",
    "    print(\"Best train score:\", gridsearch.best_score_)\n",
    "    print(\"Best train params\", gridsearch.best_params_)\n",
    "    model_optim = gridsearch.best_estimator_\n",
    "\n",
    "    # eval\n",
    "    y_pred = model_optim.predict(X_eval)\n",
    "    print(classification_report(y_true=y_eval, y_pred=y_pred))\n",
    "    print(model_optim.predict_proba(X_eval))  # TODO compare with actual labels\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## New dataset creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_id_update_dict(graph, field=\"alt_id\"):\n",
    "    dict_update_id = dict()\n",
    "    for node, alt_ids in graph.nodes(data=field):\n",
    "        if not alt_ids:\n",
    "            continue\n",
    "        for alt_id in alt_ids:\n",
    "            dict_update_id[alt_id] = node\n",
    "\n",
    "    return dict_update_id\n",
    "\n",
    "\n",
    "def get_go_subgraph(root_term_name: str, edge_keys: set = {\"is_a\"}):\n",
    "    # TODO filtering by aspect\n",
    "    graph_go = load_df(\"go_obo\")\n",
    "    go_name_to_id = {name: node for node, name in list(graph_go.nodes(data=\"name\"))}\n",
    "    root_term_id = go_name_to_id[root_term_name]\n",
    "\n",
    "    root_term_descendants = nx.ancestors(graph_go, root_term_id)\n",
    "    graph_go_sub = graph_go.subgraph(root_term_descendants)\n",
    "    graph_go_sub_isa = graph_go_sub.edge_subgraph(\n",
    "        [\n",
    "            (node1, node2, key)\n",
    "            for node1, node2, key in graph_go.edges(keys=True)\n",
    "            if key in edge_keys\n",
    "        ]\n",
    "    )\n",
    "    return graph_go_sub_isa.copy()\n",
    "\n",
    "\n",
    "def get_goa_subset(\n",
    "    root_term_name: str,\n",
    "    proteins_filter: set = None,\n",
    "    qualifiers_filter: set = None,\n",
    "    aspects_filter: set = None,\n",
    "    evidence_codes_exclude: set = None,\n",
    "):\n",
    "    # creates subset of go annotations below a root node, and filtered by a set of proteins.\n",
    "    graph_go = load_df(\"go_obo\")\n",
    "\n",
    "    go_name_to_id = {name: node for node, name in list(graph_go.nodes(data=\"name\"))}\n",
    "    go_id_to_name = {node: name for node, name in list(graph_go.nodes(data=\"name\"))}\n",
    "    root_term_id = go_name_to_id[root_term_name]\n",
    "\n",
    "    # use is_a key for retrieving ancestors and descendants, to avoid crossing into different aspect\n",
    "    graph_go_isa = graph_go.edge_subgraph(\n",
    "        [\n",
    "            (node1, node2, key)\n",
    "            for node1, node2, key in graph_go.edges(keys=True)\n",
    "            if key == \"is_a\"\n",
    "        ]\n",
    "    )\n",
    "    root_term_descendants = nx.ancestors(graph_go_isa, root_term_id)\n",
    "\n",
    "    df_goa = load_df(\"go\")\n",
    "\n",
    "    # update go terms\n",
    "    dict_update_ids = get_id_update_dict(graph_go, field=\"alt_id\")\n",
    "    df_goa.go_id.apply(\n",
    "        lambda id: dict_update_ids[id] if id in dict_update_ids.keys() else id\n",
    "    )\n",
    "    # filter annotations\n",
    "    df_goa = df_goa[df_goa.go_id.isin(root_term_descendants)]\n",
    "    if proteins_filter:\n",
    "        df_goa = df_goa[df_goa.Uniprot.isin(proteins_filter)]\n",
    "    if qualifiers_filter:\n",
    "        df_goa = df_goa[df_goa.qualifier.isin(qualifiers_filter)]\n",
    "    if aspects_filter:\n",
    "        df_goa = df_goa[df_goa.aspect.isin(aspects_filter)]\n",
    "    if evidence_codes_exclude:\n",
    "        df_goa = df_goa[~df_goa.evidence_code.isin(evidence_codes_exclude)]\n",
    "\n",
    "    df_goa = df_goa.drop_duplicates().reset_index(drop=True)\n",
    "    df_goa[\"go_term\"] = df_goa.go_id.map(go_id_to_name)\n",
    "    df_goa[\"ancestors\"] = df_goa.go_id.apply(\n",
    "        lambda x: set(nx.descendants(graph_go_isa, x) & root_term_descendants)\n",
    "    )\n",
    "    df_goa = df_goa.explode(\"ancestors\").reset_index(drop=True)\n",
    "    df_goa[\"ancestor_term\"] = df_goa.ancestors.map(go_id_to_name)\n",
    "\n",
    "    return df_goa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>go_id</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>aspect</th>\n",
       "      <th>go_term</th>\n",
       "      <th>ancestors</th>\n",
       "      <th>ancestor_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>B4E2Q0</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0005388</td>\n",
       "      <td>IEA</td>\n",
       "      <td>F</td>\n",
       "      <td>P-type calcium transporter activity</td>\n",
       "      <td>GO:0015662</td>\n",
       "      <td>P-type ion transporter activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B4E2Q0</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0005388</td>\n",
       "      <td>IEA</td>\n",
       "      <td>F</td>\n",
       "      <td>P-type calcium transporter activity</td>\n",
       "      <td>GO:0140358</td>\n",
       "      <td>P-type transmembrane transporter activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>B4E2Q0</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0005388</td>\n",
       "      <td>IEA</td>\n",
       "      <td>F</td>\n",
       "      <td>P-type calcium transporter activity</td>\n",
       "      <td>GO:0022890</td>\n",
       "      <td>inorganic cation transmembrane transporter act...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>B4E2Q0</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0005388</td>\n",
       "      <td>IEA</td>\n",
       "      <td>F</td>\n",
       "      <td>P-type calcium transporter activity</td>\n",
       "      <td>GO:0015085</td>\n",
       "      <td>calcium ion transmembrane transporter activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B4E2Q0</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0005388</td>\n",
       "      <td>IEA</td>\n",
       "      <td>F</td>\n",
       "      <td>P-type calcium transporter activity</td>\n",
       "      <td>GO:0008324</td>\n",
       "      <td>monoatomic cation transmembrane transporter ac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1124</th>\n",
       "      <td>Q9P0L9</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0015269</td>\n",
       "      <td>IDA</td>\n",
       "      <td>F</td>\n",
       "      <td>calcium-activated potassium channel activity</td>\n",
       "      <td>GO:0015075</td>\n",
       "      <td>monoatomic ion transmembrane transporter activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1125</th>\n",
       "      <td>Q9P0L9</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0015269</td>\n",
       "      <td>IDA</td>\n",
       "      <td>F</td>\n",
       "      <td>calcium-activated potassium channel activity</td>\n",
       "      <td>GO:0022839</td>\n",
       "      <td>monoatomic ion gated channel activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1126</th>\n",
       "      <td>Q9P0L9</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0015269</td>\n",
       "      <td>IDA</td>\n",
       "      <td>F</td>\n",
       "      <td>calcium-activated potassium channel activity</td>\n",
       "      <td>GO:0046873</td>\n",
       "      <td>metal ion transmembrane transporter activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1127</th>\n",
       "      <td>Q9P0L9</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0015269</td>\n",
       "      <td>IDA</td>\n",
       "      <td>F</td>\n",
       "      <td>calcium-activated potassium channel activity</td>\n",
       "      <td>GO:1901702</td>\n",
       "      <td>salt transmembrane transporter activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1128</th>\n",
       "      <td>Q9P0L9</td>\n",
       "      <td>enables</td>\n",
       "      <td>GO:0015269</td>\n",
       "      <td>IDA</td>\n",
       "      <td>F</td>\n",
       "      <td>calcium-activated potassium channel activity</td>\n",
       "      <td>GO:0015079</td>\n",
       "      <td>potassium ion transmembrane transporter activity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1129 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Uniprot qualifier       go_id evidence_code aspect  \\\n",
       "0     B4E2Q0   enables  GO:0005388           IEA      F   \n",
       "1     B4E2Q0   enables  GO:0005388           IEA      F   \n",
       "2     B4E2Q0   enables  GO:0005388           IEA      F   \n",
       "3     B4E2Q0   enables  GO:0005388           IEA      F   \n",
       "4     B4E2Q0   enables  GO:0005388           IEA      F   \n",
       "...      ...       ...         ...           ...    ...   \n",
       "1124  Q9P0L9   enables  GO:0015269           IDA      F   \n",
       "1125  Q9P0L9   enables  GO:0015269           IDA      F   \n",
       "1126  Q9P0L9   enables  GO:0015269           IDA      F   \n",
       "1127  Q9P0L9   enables  GO:0015269           IDA      F   \n",
       "1128  Q9P0L9   enables  GO:0015269           IDA      F   \n",
       "\n",
       "                                           go_term   ancestors  \\\n",
       "0              P-type calcium transporter activity  GO:0015662   \n",
       "1              P-type calcium transporter activity  GO:0140358   \n",
       "2              P-type calcium transporter activity  GO:0022890   \n",
       "3              P-type calcium transporter activity  GO:0015085   \n",
       "4              P-type calcium transporter activity  GO:0008324   \n",
       "...                                            ...         ...   \n",
       "1124  calcium-activated potassium channel activity  GO:0015075   \n",
       "1125  calcium-activated potassium channel activity  GO:0022839   \n",
       "1126  calcium-activated potassium channel activity  GO:0046873   \n",
       "1127  calcium-activated potassium channel activity  GO:1901702   \n",
       "1128  calcium-activated potassium channel activity  GO:0015079   \n",
       "\n",
       "                                          ancestor_term  \n",
       "0                       P-type ion transporter activity  \n",
       "1             P-type transmembrane transporter activity  \n",
       "2     inorganic cation transmembrane transporter act...  \n",
       "3        calcium ion transmembrane transporter activity  \n",
       "4     monoatomic cation transmembrane transporter ac...  \n",
       "...                                                 ...  \n",
       "1124  monoatomic ion transmembrane transporter activity  \n",
       "1125              monoatomic ion gated channel activity  \n",
       "1126       metal ion transmembrane transporter activity  \n",
       "1127            salt transmembrane transporter activity  \n",
       "1128   potassium ion transmembrane transporter activity  \n",
       "\n",
       "[1129 rows x 8 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(organism_ids:set, labels: set):\n",
    "    # TODO get rid of unnecessary prints\n",
    "    df_dataset = get_classification_task(\n",
    "        organism_ids=organism_ids,\n",
    "        labels=labels,\n",
    "        clustering_threshold=70,\n",
    "    )\n",
    "\n",
    "    # TODO this is a quickfix, redo pipeline\n",
    "    tmtp_proteins = get_goa_subset(\n",
    "        root_term_name=\"transmembrane transporter activity\",\n",
    "        qualifiers_filter=[\"enables\"],\n",
    "        aspects_filter=[\"F\"],\n",
    "        proteins_filter=df_dataset.index.tolist()\n",
    "    ).Uniprot.unique()\n",
    "    df_dataset = df_dataset[df_dataset.index.isin(set(tmtp_proteins))]\n",
    "    # TODO quickfix end\n",
    "\n",
    "    df_features = get_features(df_dataset.sequence)\n",
    "\n",
    "    df_features = df_features.loc[df_features.index.sort_values()]\n",
    "    df_dataset = df_dataset.loc[df_features.index]\n",
    "    df_features = df_features.loc[df_dataset.index]\n",
    "\n",
    "    print(df_dataset.shape, df_features.shape)\n",
    "\n",
    "    evaluate(df_dataset=df_dataset, df_features=df_features)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "human\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43248\n",
      "164519\n",
      "60547\n",
      "1995\n",
      "477\n",
      "474\n",
      "[('monoatomic ion', 1455), ('monoatomic cation', 1138), ('inorganic cation', 1061), ('metal cation', 852), ('calcium(2+)', 356), ('potassium(1+)', 289), ('organic anion', 266), ('sodium(1+)', 239), ('monoatomic anion', 238), ('inorganic anion', 234), ('proton', 197), ('chloride', 187), ('organic acid', 186), ('carboxylic acid anion', 183), ('chemical entity', 144), ('amino acid', 111), ('L-alpha-amino acid zwitterion', 86), ('carbohydrate derivative', 78), ('sulfur molecular entity', 66), ('ion', 58), ('monocarboxylic acid', 58), ('amide', 55), ('transition element cation', 49), ('carbohydrate', 40), ('organic cation', 40), ('amino acid derivative', 38), ('organic phosphate', 35), ('nucleotide-sugar', 32), ('biomacromolecule', 32), ('nucleotide', 31), ('pyrimidine nucleotide-sugar', 30), ('dicarboxylic acid', 28), ('purine nucleotide', 27), ('monosaccharide', 27), ('glucose', 25), ('hexose', 25), ('hydrogencarbonate', 24), ('adenyl nucleotide', 24), ('purine ribonucleotide', 23), ('zinc(2+)', 23), ('quaternary nitrogen compound', 22), ('ammonium', 21), ('bile acid', 18), ('sulfate', 17), ('water', 16), ('riboflavin(1-)', 16), ('peptide', 15), ('magnesium cation', 15), ('C4-dicarboxylate', 15), ('alanine zwitterion', 15), ('ATP(4-)', 14), ('azole', 14), ('icosanoid', 13), ('L-argininium(1+)', 13), ('nucleoside', 13), ('ADP(3-)', 13), ('urea', 12), ('glutathione derivative', 12), ('prostaglandin', 12), ('branched-chain amino acid', 11), ('copper cation', 11), ('polyol', 11), ('sulfur-containing amino acid', 11), ('monoamine', 11), ('choline', 10), ('D-glucose', 10), ('phosphate', 10), ('L-glutamate(1-)', 10), ('aromatic amino acid', 10), ('glycine zwitterion', 10), ('gamma-aminobutyric acid zwitterion', 10), ('lipid', 9), ('vitamin B1', 9), ('fatty acid anion', 9), ('L-aspartate(1-)', 9), ('heme', 9), ('L-lysinium(1+)', 9), ('serine zwitterion', 9), ('L-glutamine zwitterion', 9), ('inorganic phosphate', 8), ('iron cation', 8), ('manganese cation', 8), ('fructose', 8), ('oxalate(2-)', 8), ('phosphate ion', 8), ('copper(1+)', 8), ('polyamine macromolecule', 8), ('L-alanine zwitterion', 8), ('glutamate(1-)', 7), ('glycerol', 7), ('nucleobase', 7), ('proline', 7), ('alcohol', 7), ('L-serine zwitterion', 7), ('dehydroascorbide(1-)', 7), ('L-proline zwitterion', 7), ('carnitine', 6), ('oligopeptide', 6), ('purine nucleobase', 6), ('pyrimidine nucleoside', 6), ('succinate(2-)', 6), ('retinol', 6), ('L-leucine zwitterion', 6), ('folate(2-)', 6), ('zinc cation', 6), ('alkanesulfonate oxoanion', 5), ('pyruvate', 5), ('lactate', 5), ('iron(2+)', 5), ('oleate', 5), ('L-ornithinium(1+)', 5), ('taurine zwitterion', 5), ('glutathionate(1-)', 5), ('creatine zwitterion', 5), ('carbon dioxide', 4), ('uridine', 4), ('2-oxoglutarate(2-)', 4), ('citrate(3-)', 4), ('myo-inositol', 4), ('sucrose', 4), ('tricarboxylic acid', 4), ('galactose', 4), ('1,4-butanediammonium', 4), ('L-cystine zwitterion', 4), ('dipeptide', 4), ('L-histidine zwitterion', 4), ('spermidine(3+)', 4), ('L-asparagine zwitterion', 4), ('guanyl nucleotide', 4), ('acetylcholine', 3), ('malate(2-)', 3), ('(R)-carnitine', 3), ('iodide', 3), ('oxaloacetate(2-)', 3), ('adenine', 3), ('glucoside', 3), ('pyrimidine nucleobase', 3), ('amine', 3), ('serotonin(1+)', 3), ('cysteine zwitterion', 3), ('methotrexate(2-)', 3), ('coenzyme A(4-)', 3), ('acetyl-CoA(4-)', 3), ('biotinate', 3), ('UDP-N-acetyl-alpha-D-glucosamine(2-)', 3), ('L-tryptophan zwitterion', 3), ('L-isoleucine zwitterion', 3), ('short-chain fatty acid anion', 3), ('dopaminium(1+)', 3), ('cadmium cation', 3), ('(R)-noradrenaline(1+)', 3), ('NAD', 2), ('D-glucose 6-phosphate', 2), ('sterol', 2), ('O-acylcarnitine', 2), ('cobalt cation', 2), ('flavin adenine dinucleotide', 2), ('organophosphorus compound', 2), ('pentose', 2), ('purine nucleoside', 2), ('pyrimidine nucleotide', 2), ('copper(2+)', 2), ('cobalamin', 2), ('nucleic acid', 2), ('ribonucleic acid', 2), ('fucose', 2), ('aspartate(1-)', 2), ('L-ascorbate', 2), (\"adenosine 5'-monophosphate(2-)\", 2), ('hexose phosphate', 2), ('L-phenylalanine zwitterion', 2), (\"3'-phosphonato-5'-adenylyl sulfate(4-)\", 2), ('thiamine(1+) diphosphate(3-)', 2), ('alpha-glucoside', 2), ('3-sulfino-L-alanine(1-)', 2), ('GDP-fucose', 2), ('c-GMP-AMP(2-)', 2), ('peptidoglycan', 2), ('formate', 1), ('D-glucuronate', 1), ('D-fructose', 1), ('D-mannose', 1), ('guanine', 1), ('nitric oxide', 1), ('monoacylglycerol phosphate', 1), ('propionate', 1), ('cytidine', 1), ('nitrate', 1), ('glycine betaine', 1), ('glucosiduronic acid', 1), ('mevalonate', 1), ('nickel cation', 1), ('sialic acid', 1), ('tricarboxylic acid trianion', 1), ('(R)-pantothenate', 1), ('arsenite(1-)', 1), ('fumarate(2-)', 1), ('D-aspartate(1-)', 1), ('iron-sulfur cluster', 1), ('nicotinate', 1), ('boric acid', 1), ('vanadium cation', 1), ('cystine zwitterion', 1), ('purine', 1), ('5-ammoniolevulinate', 1), ('molybdate', 1), ('mannose', 1), ('tripeptide', 1), ('lithium(1+)', 1), ('UDP-alpha-D-xylose(2-)', 1), ('L-valine zwitterion', 1), ('CMP-N-acetyl-beta-neuraminate(2-)', 1), ('L-methionine zwitterion', 1), ('L-threonine zwitterion', 1), ('L-kynurenine zwitterion', 1), ('beta-alanine zwitterion', 1), ('bilirubin(2-)', 1), ('FMN(3-)', 1), ('L-tyrosine zwitterion', 1), ('UDP-D-glucose(2-)', 1), ('methylammonium', 1), ('S-adenosyl-L-methionine zwitterion', 1), ('D-alpha-amino acid zwitterion', 1), ('lead ion', 1), ('silver cation', 1), ('chromium ion', 1), ('4-hydroxyproline zwitterion', 1), ('fatty acyl-CoA(4-)', 1), ('enterobactin(1-)', 1), ('tetracycline zwitterion', 1)]\n",
      "label\n",
      "calcium(2+)      356\n",
      "potassium(1+)    266\n",
      "Name: count, dtype: int64\n",
      "number of sequences 622\n",
      "cd-hit: clustered 622 sequences into 373 clusters at threshold 70\n",
      "['Q8TDD5', 'P78417', 'Q12981', 'O75762', 'P24588', 'Q9H1D0', 'A0A1X7SBR8', 'P06756', 'P32418', 'O75052', 'P08913', 'Q8IU99', 'Q9NPB3', 'P62955', 'J3QTB0', 'P36544', 'Q9NZN3', 'O95153', 'Q14524', 'P35372', 'Q9HBA0', 'B7ZMI3', 'O60733', 'P98161', 'P16298', 'Q9BX84', 'Q9NR96', 'P31415', 'P55042', 'O95180', 'P51674', 'Q8TEL6', 'Q9Y210', 'Q96SN7', 'E9PJN4', 'P49768', 'A0A804HIR0', 'Q9NZU7', 'Q9P246', 'P23634', 'Q9NP86', 'P14416', 'P0DP24', 'P63165', 'P56704', 'P05026', 'P57796', 'A0A087X080', 'Q9P0X4', 'Q05586', 'A0A1B0GTP6', 'Q96RD7', 'Q12797', 'Q96D31', 'Q99571', 'Q8NHX9', 'P54284', 'A0A1B0GTN0', 'P30626', 'P19634', 'B1AL88', 'P61619', 'Q12879', 'Q96GC9', 'Q9H3K2', 'O75628', 'Q6ZMT1', 'P05067', 'Q14644', 'O95202', 'P17612', 'Q13023', 'P55916', 'P07355', 'P63098', 'A0A1C7CYY9', 'P21917', 'Q07343', 'Q9ULQ1', 'O60840', 'P41146', 'Q15858', 'Q8NE86', 'P34998', 'O00305', 'Q01484', 'O75323', 'P48995', 'P56539', 'P62258', 'Q9BR39', 'Q8IZK6', 'A0A804HKC4', 'P29475', 'Q9UBN4', 'Q9Y6N3', 'Q96P56', 'P08133', 'Q14571', 'Q96QT4', 'Q9UM00', 'P0DN84', 'Q01814', 'Q9GZU1', 'A0A1B0GV16', 'Q13698', 'Q15878', 'Q06787', 'Q8WVX3', 'Q8NFF2', 'O43497', 'Q14957', 'P35498', 'Q9UBN1', 'P57103', 'Q92736', 'Q71RS6', 'O14775', 'P84074', 'Q96MF2', 'Q14573', 'Q99469', 'F5GY28', 'P62166', 'Q8NER1', 'Q9Y698', 'Q9HC07', 'Q9UGM1', 'K7EJY1', 'Q6XR72', 'Q8NEC5', 'A0A1B0GTK1', 'Q06432', 'Q01118', 'P54652', 'Q9UL62', 'P28222', 'Q8TDX9', 'O14958', 'Q6J4K2', 'Q08289', 'O00631', 'P16885', 'Q7RTX7', 'P68106', 'Q8NET8', 'Q13507', 'P21817', 'O75185', 'Q13224', 'Q96MG2', 'P55040', 'Q86VW0', 'Q9P0L9', 'E7EMC0', 'Q9NZV5', 'P16615', 'P11532', 'Q08499', 'Q8WXH2', 'O15399', 'E9PGK7', 'Q93034', 'Q7Z443', 'Q96JJ6', 'Q8TD43', 'P54257', 'Q86XK2', 'Q7Z442', 'Q8IZS8', 'Q13557', 'Q13563', 'Q9Y5S1', 'P42858', 'P23327', 'Q14643', 'Q9NZQ8', 'P08195', 'F5H483', 'Q8TCU5', 'Q9GZZ6', 'Q7Z2W7', 'Q00535', 'Q9UI33', 'A0A140LJL2', 'A0A1B0GU42', 'Q9UMX0', 'O15247', 'O00499', 'P54289', 'P49281', 'C9JVC9', 'G0XQ39', 'Q9UI40', 'Q6YHU6', 'Q8TDI8', 'Q09666', 'P35499', 'P11413', 'P28161', 'A0A7P0T8M7', 'Q9BXT2', 'Q9BRQ5', 'Q86TL2', 'P15382', 'Q15493', 'Q7Z4N2', 'P26678', 'Q9HDC5', 'Q7Z3S7', 'O60721', 'Q02641', 'Q15077', 'Q13061', 'B4E2Q0', 'Q9Y6J6', 'Q9Y5Y9', 'P01160', 'Q9BSW2', 'Q00975', 'Q9NYG8', 'Q8N7C0', 'P54709', 'Q14500', 'Q99712', 'P29972', 'Q6PIU1', 'O95069', 'P48051', 'P55017', 'Q96ER9', 'P56696', 'Q9NPC2', 'Q9Y257', 'Q16322', 'O95747', 'Q86WK6', 'Q8N5I3', 'Q15842', 'Q8WWG9', 'Q99996', 'Q9NVV0', 'O00180', 'P51164', 'P39086', 'Q12959', 'P55085', 'Q92806', 'Q9BXP2', 'Q9Y2W7', 'O60741', 'Q5SVJ8', 'A0A0A0MT26', 'Q8TDN2', 'Q09428', 'O43525', 'Q6ZSA7', 'A0A590UJ62', 'P48549', 'P48050', 'Q9H3M0', 'P48764', 'P55011', 'Q8TAE7', 'P63027', 'Q9NPA1', 'A0A087WYJ0', 'Q9UJ90', 'Q8TDN1', 'A0A590UJU4', 'O00308', 'Q9UJ96', 'A0A804HHV7', 'Q9NS40', 'Q7Z418', 'Q9UHW9', 'Q9UNX9', 'Q9UL51', 'Q9BQ31', 'Q8IVB4', 'A8MYU2', 'P07492', 'Q2I0M4', 'A0A0D9SEY3', 'P14415', 'P22460', 'P54710', 'Q5I7T1', 'Q6TFL4', 'Q13642', 'Q86W47', 'P46934', 'Q92581', 'P22466', 'Q16558', 'A0A1W2PPZ1', 'P78508', 'Q8NCM2', 'Q9Y3Q4', 'O43603', 'Q12809', 'Q9NPI9', 'Q14722', 'P54707', 'Q9Y2U2', 'O43526', 'A0A0D9SG49', 'Q96T83', 'Q9H252', 'A0A804HIW7', 'O95279', 'Q13621', 'Q9Y376', 'P35609', 'Q12955', 'P48048', 'Q14940', 'Q6PIL6', 'Q9UIX4', 'P22001', 'Q03721', 'Q96T55', 'Q9NZC2', 'Q9UK17', 'A0A3F2YNY5', 'O60939', 'O14649', 'P17658', 'Q14721', 'Q6UVM3', 'Q9Y691', 'Q9BSA9', 'Q9P1Z3', 'Q9Y6H6', 'P50570', 'Q5VT99', 'Q9H6F2', 'Q9NS61', 'Q9NZI2', 'P60880', 'P20648', 'O60706', 'F8WEA4', 'O15554', 'Q6UWJ1', 'Q96PU5', 'Q9UEW8', 'Q9Y2E8', 'A0AV02', 'Q9H2X9', 'Q03135', 'Q96KK3', 'Q96SW2', 'Q6ZRF8', 'K7EKU4', 'O43687', 'P05556', 'P57789', 'P49286', 'O43448', 'O95259', 'A0A1W2PNZ3', 'Q9ULD8', 'A0A0G2JNW7', 'Q9ULS6', 'Q96PR1', 'O60928', 'Q9H427', 'Q9NZV8', 'P22459', 'P51787', 'Q9NSA2', 'A0A590UJI8']\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 46\u001b[0m\n\u001b[1;32m     44\u001b[0m \u001b[39mfor\u001b[39;00m dataset_name, organism_ids \u001b[39min\u001b[39;00m dataset_name_to_organism_ids\u001b[39m.\u001b[39mitems():\n\u001b[1;32m     45\u001b[0m     \u001b[39mprint\u001b[39m(dataset_name)\n\u001b[0;32m---> 46\u001b[0m     main(organism_ids\u001b[39m=\u001b[39;49morganism_ids, labels\u001b[39m=\u001b[39;49mlabels)\n",
      "Cell \u001b[0;32mIn[6], line 18\u001b[0m, in \u001b[0;36mmain\u001b[0;34m(organism_ids, labels)\u001b[0m\n\u001b[1;32m     10\u001b[0m tmtp_proteins \u001b[39m=\u001b[39m get_goa_subset(\n\u001b[1;32m     11\u001b[0m     root_term_name\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mtransmembrane transporter activity\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m     12\u001b[0m     qualifiers_filter\u001b[39m=\u001b[39m[\u001b[39m\"\u001b[39m\u001b[39menables\u001b[39m\u001b[39m\"\u001b[39m],\n\u001b[1;32m     13\u001b[0m     aspects_filter\u001b[39m=\u001b[39m[\u001b[39m\"\u001b[39m\u001b[39mF\u001b[39m\u001b[39m\"\u001b[39m],\n\u001b[1;32m     14\u001b[0m     proteins_filter\u001b[39m=\u001b[39mdf_dataset\u001b[39m.\u001b[39mindex\u001b[39m.\u001b[39mtolist()\n\u001b[1;32m     15\u001b[0m )\u001b[39m.\u001b[39mUniprot\u001b[39m.\u001b[39munique()\n\u001b[1;32m     16\u001b[0m df_dataset \u001b[39m=\u001b[39m df_dataset[df_dataset\u001b[39m.\u001b[39mindex\u001b[39m.\u001b[39misin(\u001b[39mset\u001b[39m(tmtp_proteins))]\n\u001b[0;32m---> 18\u001b[0m df_features \u001b[39m=\u001b[39m get_features(df_dataset\u001b[39m.\u001b[39;49msequence)\n\u001b[1;32m     20\u001b[0m df_features \u001b[39m=\u001b[39m df_features\u001b[39m.\u001b[39mloc[df_features\u001b[39m.\u001b[39mindex\u001b[39m.\u001b[39msort_values()]\n\u001b[1;32m     21\u001b[0m df_dataset \u001b[39m=\u001b[39m df_dataset\u001b[39m.\u001b[39mloc[df_features\u001b[39m.\u001b[39mindex]\n",
      "Cell \u001b[0;32mIn[3], line 4\u001b[0m, in \u001b[0;36mget_features\u001b[0;34m(series_sequences)\u001b[0m\n\u001b[1;32m      2\u001b[0m df_aac \u001b[39m=\u001b[39m calculate_aac(series_sequences)\n\u001b[1;32m      3\u001b[0m df_paac \u001b[39m=\u001b[39m calculate_paac(series_sequences)\n\u001b[0;32m----> 4\u001b[0m df_pssm_50_1 \u001b[39m=\u001b[39m calculate_pssm_feature(\n\u001b[1;32m      5\u001b[0m     series_sequences,\n\u001b[1;32m      6\u001b[0m     tmp_folder\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m../data/intermediate/blast/pssm_uniref50_1it\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m      7\u001b[0m     blast_db\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39m../data/raw/uniref/uniref50/uniref50.fasta\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m      8\u001b[0m     iterations\u001b[39m=\u001b[39;49m\u001b[39m1\u001b[39;49m,\n\u001b[1;32m      9\u001b[0m     psiblast_threads\u001b[39m=\u001b[39;49m\u001b[39m-\u001b[39;49m\u001b[39m1\u001b[39;49m,\n\u001b[1;32m     10\u001b[0m     verbose\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m,\n\u001b[1;32m     11\u001b[0m     feature_name\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mPSSM_50_1\u001b[39;49m\u001b[39m\"\u001b[39;49m\n\u001b[1;32m     12\u001b[0m )\n\u001b[1;32m     13\u001b[0m df_pssm_50_3 \u001b[39m=\u001b[39m calculate_pssm_feature(\n\u001b[1;32m     14\u001b[0m     series_sequences,\n\u001b[1;32m     15\u001b[0m     tmp_folder\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m../data/intermediate/blast/pssm_uniref50_3it\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     20\u001b[0m     feature_name\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mPSSM_50_3\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m     21\u001b[0m )\n\u001b[1;32m     22\u001b[0m df_pssm_90_1 \u001b[39m=\u001b[39m calculate_pssm_feature(\n\u001b[1;32m     23\u001b[0m     series_sequences,\n\u001b[1;32m     24\u001b[0m     tmp_folder\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m../data/intermediate/blast/pssm_uniref90_3it\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     29\u001b[0m     feature_name\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mPSSM_90_1\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m     30\u001b[0m )\n",
      "File \u001b[0;32m~/subpred4/subpred/pssm.py:177\u001b[0m, in \u001b[0;36mcalculate_pssm_feature\u001b[0;34m(sequences, tmp_folder, blast_db, iterations, psiblast_executable, psiblast_threads, verbose, feature_name)\u001b[0m\n\u001b[1;32m    175\u001b[0m \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(\u001b[39mlen\u001b[39m(sequences)):\n\u001b[1;32m    176\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 177\u001b[0m         pssm \u001b[39m=\u001b[39m __get_pssm_feature(\n\u001b[1;32m    178\u001b[0m             accession\u001b[39m=\u001b[39;49msequences\u001b[39m.\u001b[39;49mindex[i],\n\u001b[1;32m    179\u001b[0m             sequence\u001b[39m=\u001b[39;49msequences\u001b[39m.\u001b[39;49mvalues[i],\n\u001b[1;32m    180\u001b[0m             blastdb_fasta_file\u001b[39m=\u001b[39;49mblast_db,\n\u001b[1;32m    181\u001b[0m             pssm_folder_path\u001b[39m=\u001b[39;49mtmp_folder,\n\u001b[1;32m    182\u001b[0m             iterations\u001b[39m=\u001b[39;49miterations,\n\u001b[1;32m    183\u001b[0m             psiblast_location\u001b[39m=\u001b[39;49mpsiblast_executable,\n\u001b[1;32m    184\u001b[0m             threads\u001b[39m=\u001b[39;49mpsiblast_threads,\n\u001b[1;32m    185\u001b[0m             verbose\u001b[39m=\u001b[39;49mverbose,\n\u001b[1;32m    186\u001b[0m         )\n\u001b[1;32m    187\u001b[0m         accessions\u001b[39m.\u001b[39mappend(sequences\u001b[39m.\u001b[39mindex[i])\n\u001b[1;32m    188\u001b[0m         features\u001b[39m.\u001b[39mappend(pssm)\n",
      "File \u001b[0;32m~/subpred4/subpred/pssm.py:145\u001b[0m, in \u001b[0;36m__get_pssm_feature\u001b[0;34m(accession, sequence, blastdb_fasta_file, pssm_folder_path, psiblast_location, iterations, evalue, threads, verbose)\u001b[0m\n\u001b[1;32m    139\u001b[0m     \u001b[39mprint\u001b[39m(\n\u001b[1;32m    140\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mPSSM for accession \u001b[39m\u001b[39m{\u001b[39;00maccession\u001b[39m}\u001b[39;00m\u001b[39m was not found in tmp folder \u001b[39m\u001b[39m{\u001b[39;00mpssm_folder_path\u001b[39m}\u001b[39;00m\u001b[39m, calling psiblast\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    141\u001b[0m     )\n\u001b[1;32m    142\u001b[0m write_fasta(\n\u001b[1;32m    143\u001b[0m     fasta_file_name\u001b[39m=\u001b[39mfasta_file_name, fasta_data\u001b[39m=\u001b[39m[(\u001b[39m\"\u001b[39m\u001b[39m>\u001b[39m\u001b[39m\"\u001b[39m \u001b[39m+\u001b[39m accession, sequence)]\n\u001b[1;32m    144\u001b[0m )\n\u001b[0;32m--> 145\u001b[0m __create_pssm_file(\n\u001b[1;32m    146\u001b[0m     psiblast_location\u001b[39m=\u001b[39;49mpsiblast_location,\n\u001b[1;32m    147\u001b[0m     fasta_file_name\u001b[39m=\u001b[39;49mfasta_file_name,\n\u001b[1;32m    148\u001b[0m     pssm_file_name\u001b[39m=\u001b[39;49mpssm_file_name,\n\u001b[1;32m    149\u001b[0m     blastdb_location\u001b[39m=\u001b[39;49mblastdb_fasta_file,\n\u001b[1;32m    150\u001b[0m     iterations\u001b[39m=\u001b[39;49miterations,\n\u001b[1;32m    151\u001b[0m     evalue\u001b[39m=\u001b[39;49mevalue,\n\u001b[1;32m    152\u001b[0m     threads\u001b[39m=\u001b[39;49mthreads,\n\u001b[1;32m    153\u001b[0m )\n\u001b[1;32m    154\u001b[0m pssm \u001b[39m=\u001b[39m __process_pssm_file(pssm_file_name, sequence)\n\u001b[1;32m    155\u001b[0m \u001b[39mif\u001b[39;00m verbose:\n",
      "File \u001b[0;32m~/subpred4/subpred/pssm.py:90\u001b[0m, in \u001b[0;36m__create_pssm_file\u001b[0;34m(psiblast_location, fasta_file_name, pssm_file_name, blastdb_location, iterations, evalue, threads)\u001b[0m\n\u001b[1;32m     88\u001b[0m \u001b[39m# TODO create tmp files, then rename after program finished\u001b[39;00m\n\u001b[1;32m     89\u001b[0m log_file_name \u001b[39m=\u001b[39m \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mpssm_file_name\u001b[39m}\u001b[39;00m\u001b[39m.log\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m---> 90\u001b[0m subprocess\u001b[39m.\u001b[39;49mrun(\n\u001b[1;32m     91\u001b[0m     \u001b[39m\"\u001b[39;49m\u001b[39m{}\u001b[39;49;00m\u001b[39m -query \u001b[39;49m\u001b[39m{}\u001b[39;49;00m\u001b[39m -db \u001b[39;49m\u001b[39m{}\u001b[39;49;00m\u001b[39m -num_iterations \u001b[39;49m\u001b[39m{}\u001b[39;49;00m\u001b[39m -inclusion_ethresh \u001b[39;49m\u001b[39m{}\u001b[39;49;00m\u001b[39m -num_threads \u001b[39;49m\u001b[39m{}\u001b[39;49;00m\u001b[39m -save_pssm_after_last_round\u001b[39;49m\u001b[39m\\\u001b[39;49;00m\n\u001b[1;32m     92\u001b[0m \u001b[39m         -out_ascii_pssm \u001b[39;49m\u001b[39m{}\u001b[39;49;00m\u001b[39m -out \u001b[39;49m\u001b[39m{}\u001b[39;49;00m\u001b[39m -comp_based_stats \u001b[39;49m\u001b[39m{}\u001b[39;49;00m\u001b[39m\"\u001b[39;49m\u001b[39m.\u001b[39;49mformat(\n\u001b[1;32m     93\u001b[0m         psiblast_location,\n\u001b[1;32m     94\u001b[0m         fasta_file_name,\n\u001b[1;32m     95\u001b[0m         blastdb_location,\n\u001b[1;32m     96\u001b[0m         iterations,\n\u001b[1;32m     97\u001b[0m         evalue,\n\u001b[1;32m     98\u001b[0m         threads,\n\u001b[1;32m     99\u001b[0m         pssm_file_name,\n\u001b[1;32m    100\u001b[0m         log_file_name,\n\u001b[1;32m    101\u001b[0m         \u001b[39m2\u001b[39;49m\n\u001b[1;32m    102\u001b[0m         \u001b[39mif\u001b[39;49;00m iterations \u001b[39m==\u001b[39;49m \u001b[39m1\u001b[39;49m\n\u001b[1;32m    103\u001b[0m         \u001b[39melse\u001b[39;49;00m \u001b[39m1\u001b[39;49m,  \u001b[39m# default is 2, but not supported when matrix is PSSM instead of BLOSUM\u001b[39;49;00m\n\u001b[1;32m    104\u001b[0m     ),\n\u001b[1;32m    105\u001b[0m     check\u001b[39m=\u001b[39;49m\u001b[39mTrue\u001b[39;49;00m,\n\u001b[1;32m    106\u001b[0m     shell\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m \u001b[39mif\u001b[39;49;00m platform\u001b[39m.\u001b[39;49msystem() \u001b[39m==\u001b[39;49m \u001b[39m\"\u001b[39;49m\u001b[39mWindows\u001b[39;49m\u001b[39m\"\u001b[39;49m \u001b[39melse\u001b[39;49;00m \u001b[39mTrue\u001b[39;49;00m,\n\u001b[1;32m    107\u001b[0m     \u001b[39m# stdout=subprocess.DEVNULL,\u001b[39;49;00m\n\u001b[1;32m    108\u001b[0m     \u001b[39m# stderr=subprocess.DEVNULL,\u001b[39;49;00m\n\u001b[1;32m    109\u001b[0m )\n",
      "File \u001b[0;32m~/mambaforge/envs/subpred4/lib/python3.10/subprocess.py:505\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    503\u001b[0m \u001b[39mwith\u001b[39;00m Popen(\u001b[39m*\u001b[39mpopenargs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs) \u001b[39mas\u001b[39;00m process:\n\u001b[1;32m    504\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 505\u001b[0m         stdout, stderr \u001b[39m=\u001b[39m process\u001b[39m.\u001b[39;49mcommunicate(\u001b[39minput\u001b[39;49m, timeout\u001b[39m=\u001b[39;49mtimeout)\n\u001b[1;32m    506\u001b[0m     \u001b[39mexcept\u001b[39;00m TimeoutExpired \u001b[39mas\u001b[39;00m exc:\n\u001b[1;32m    507\u001b[0m         process\u001b[39m.\u001b[39mkill()\n",
      "File \u001b[0;32m~/mambaforge/envs/subpred4/lib/python3.10/subprocess.py:1146\u001b[0m, in \u001b[0;36mPopen.communicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1144\u001b[0m         stderr \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mstderr\u001b[39m.\u001b[39mread()\n\u001b[1;32m   1145\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mstderr\u001b[39m.\u001b[39mclose()\n\u001b[0;32m-> 1146\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mwait()\n\u001b[1;32m   1147\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m   1148\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/mambaforge/envs/subpred4/lib/python3.10/subprocess.py:1209\u001b[0m, in \u001b[0;36mPopen.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1207\u001b[0m     endtime \u001b[39m=\u001b[39m _time() \u001b[39m+\u001b[39m timeout\n\u001b[1;32m   1208\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1209\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_wait(timeout\u001b[39m=\u001b[39;49mtimeout)\n\u001b[1;32m   1210\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m   1211\u001b[0m     \u001b[39m# https://bugs.python.org/issue25942\u001b[39;00m\n\u001b[1;32m   1212\u001b[0m     \u001b[39m# The first keyboard interrupt waits briefly for the child to\u001b[39;00m\n\u001b[1;32m   1213\u001b[0m     \u001b[39m# exit under the common assumption that it also received the ^C\u001b[39;00m\n\u001b[1;32m   1214\u001b[0m     \u001b[39m# generated SIGINT and will exit rapidly.\u001b[39;00m\n\u001b[1;32m   1215\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/mambaforge/envs/subpred4/lib/python3.10/subprocess.py:1959\u001b[0m, in \u001b[0;36mPopen._wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1957\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mreturncode \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m   1958\u001b[0m     \u001b[39mbreak\u001b[39;00m  \u001b[39m# Another thread waited.\u001b[39;00m\n\u001b[0;32m-> 1959\u001b[0m (pid, sts) \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_try_wait(\u001b[39m0\u001b[39;49m)\n\u001b[1;32m   1960\u001b[0m \u001b[39m# Check the pid and loop as waitpid has been known to\u001b[39;00m\n\u001b[1;32m   1961\u001b[0m \u001b[39m# return 0 even without WNOHANG in odd situations.\u001b[39;00m\n\u001b[1;32m   1962\u001b[0m \u001b[39m# http://bugs.python.org/issue14396.\u001b[39;00m\n\u001b[1;32m   1963\u001b[0m \u001b[39mif\u001b[39;00m pid \u001b[39m==\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpid:\n",
      "File \u001b[0;32m~/mambaforge/envs/subpred4/lib/python3.10/subprocess.py:1917\u001b[0m, in \u001b[0;36mPopen._try_wait\u001b[0;34m(self, wait_flags)\u001b[0m\n\u001b[1;32m   1915\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[39;00m\n\u001b[1;32m   1916\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1917\u001b[0m     (pid, sts) \u001b[39m=\u001b[39m os\u001b[39m.\u001b[39;49mwaitpid(\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mpid, wait_flags)\n\u001b[1;32m   1918\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mChildProcessError\u001b[39;00m:\n\u001b[1;32m   1919\u001b[0m     \u001b[39m# This happens if SIGCLD is set to be ignored or waiting\u001b[39;00m\n\u001b[1;32m   1920\u001b[0m     \u001b[39m# for child processes has otherwise been disabled for our\u001b[39;00m\n\u001b[1;32m   1921\u001b[0m     \u001b[39m# process.  This child is dead, we can't get the status.\u001b[39;00m\n\u001b[1;32m   1922\u001b[0m     pid \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpid\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "labels = {\"potassium(1+)\", \"calcium(2+)\"}\n",
    "\n",
    "dataset_name_to_organism_ids = {\n",
    "    \"human\": {9606},\n",
    "    \"athaliana\": {3702},\n",
    "    \"ecoli\": {83333},\n",
    "    \"yeast\": {559292},\n",
    "}\n",
    "dataset_name_to_organism_ids[\"all\"] = {\n",
    "    list(s)[0] for s in dataset_name_to_organism_ids.values() if len(s) == 1\n",
    "}\n",
    "\n",
    "\n",
    "test_cases = [\n",
    "    (\"athaliana\", \"potassium(1+)\", \"calcium(2+)\"),\n",
    "    (\"athaliana\", \"inorganic anion\", \"inorganic cation\"),\n",
    "    (\"athaliana\", \"carboxylic acid anion\", \"inorganic anion\"),\n",
    "    (\"ecoli\", \"carbohydrate derivate\", \"monosaccharide\"),\n",
    "    (\"ecoli\", \"monocarboxylic acid\", \"amino acid\"),\n",
    "    (\"human\", \"calcium(2+)\", \"sodium(1+)\"),\n",
    "    (\"human\", \"calcium(2+)\", \"potassium(1+)\"),\n",
    "    (\"human\", \"sodium(1+)\", \"potassium(1+)\"),\n",
    "    (\"human\", \"inorganic anion\", \"inorganic cation\"),\n",
    "    (\"yeast\", \"amide\", \"amino acid derivative\"),\n",
    "]\n",
    "\n",
    "# - athaliana\n",
    "#   - Ca2+ und K+\n",
    "#   - inorganic anion/cation\n",
    "#   - carboxylic acid anion/inorganic anion\n",
    "# - ecoli\n",
    "#   - carbohydrate derivate / monosaccharide\n",
    "#   - minicarboxylic acid / amino acid\n",
    "# - human\n",
    "#   - Ca2+ / Na1+\n",
    "#   - Ca2+ / K+\n",
    "#   - Na+ / K+\n",
    "#   - inorganic anion/cation\n",
    "# - hefe\n",
    "#   - amid / amino acid derivative\n",
    "\n",
    "# TODO P05556 protein does not belong in there?\n",
    "\n",
    "for dataset_name, organism_ids in dataset_name_to_organism_ids.items():\n",
    "    print(dataset_name)\n",
    "    main(organism_ids=organism_ids, labels=labels)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PSSM generation script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # TODO delete\n",
    "\n",
    "# dataset_name_to_organism_ids = {\n",
    "#     \"human\": {9606},\n",
    "#     \"athaliana\": {3702},\n",
    "#     \"ecoli\": {83333},\n",
    "#     \"yeast\": {559292},\n",
    "# }\n",
    "\n",
    "# test_cases = [\n",
    "#     (\"athaliana\", \"potassium(1+)\", \"calcium(2+)\"),\n",
    "#     (\"athaliana\", \"inorganic anion\", \"inorganic cation\"),\n",
    "#     (\"athaliana\", \"carboxylic acid anion\", \"inorganic anion\"),\n",
    "#     (\"ecoli\", \"carbohydrate derivate\", \"monosaccharide\"),\n",
    "#     (\"ecoli\", \"monocarboxylic acid\", \"amino acid\"),\n",
    "#     (\"human\", \"calcium(2+)\", \"sodium(1+)\"),\n",
    "#     (\"human\", \"calcium(2+)\", \"potassium(1+)\"),\n",
    "#     (\"human\", \"sodium(1+)\", \"potassium(1+)\"),\n",
    "#     (\"human\", \"inorganic anion\", \"inorganic cation\"),\n",
    "#     (\"yeast\", \"amide\", \"amino acid derivative\"),\n",
    "# ]\n",
    "\n",
    "# dataset_name_to_organism_ids[\"all\"] = {\n",
    "#     list(s)[0] for s in dataset_name_to_organism_ids.values() if len(s) == 1\n",
    "# }\n",
    "\n",
    "# for dataset_name, substrate1, substrate2 in test_cases:\n",
    "#     organism_ids = dataset_name_to_organism_ids[dataset_name]\n",
    "#     df_dataset = get_classification_task(\n",
    "#         organism_ids=organism_ids,\n",
    "#         labels={substrate1, substrate2},\n",
    "#         clustering_threshold=70,\n",
    "#     )\n",
    "\n",
    "# df_features = get_features(df_dataset.sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparisons\n",
    "\n",
    "Compare training results with: \n",
    "\n",
    "- Average sequence similarity\n",
    "- GO term similarity\n",
    "  - How many protein in common?\n",
    "  - Semantic similarity?\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "subpred4",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
