{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Training machine learning models on pairs of substrates in individual organisms\n",
    "\n",
    "After creating the set of primary ChEBI substrates from a set of transmembrane transport proteins, the next task is to create a pipeline for generating these datasets automatically. Next, the "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generalized dataset creation method\n",
    "\n",
    "- filter uniprot proteins generally\n",
    "- filter again for transmembrane transporter activity\n",
    "- get network of GO terms that are descendants of transmembrane transporter activity\n",
    "    - filter network for proteins with in dataset with term\n",
    "    - annotate with number of proteins\n",
    "    - annotate with ChEBI substrates\n",
    "- get chebi network\n",
    "    - filter for GO terms in filtered GO network\n",
    "- 70% sequence identity?\n",
    "- feature generation\n",
    "- svm pipeline\n",
    "\n",
    "End results: \n",
    "- Annotated GO network for protein dataset\n",
    "- Annotated chebi network for protein dataset\n",
    "- Dataframe with accession, organism id, sequence, transport go terms, and transported substrate chebi term"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "from subpred.util import load_df\n",
    "from subpred.graph import preprocess_data, get_substrate_matrix\n",
    "from subpred.pssm import calculate_pssm_feature\n",
    "from subpred.compositions import calculate_aac, calculate_paac\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'human': {9606},\n",
       " 'athaliana': {3702},\n",
       " 'ecoli': {83333},\n",
       " 'yeast': {559292},\n",
       " 'all': {3702, 9606, 83333, 559292}}"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DATASET_FOLDER_PATH = \"../data/datasets\"\n",
    "\n",
    "DATASET_NAME_TO_ORGANISM_IDS = {\"human\": {9606}, \"athaliana\": {3702}, \"ecoli\": {83333}, \"yeast\": {559292}}\n",
    "DATASET_NAME_TO_ORGANISM_IDS[\"all\"] = {list(s)[0] for s in DATASET_NAME_TO_ORGANISM_IDS.values() if len(s) == 1}\n",
    "DATASET_NAME_TO_ORGANISM_IDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43248\n",
      "164519\n",
      "60547\n"
     ]
    }
   ],
   "source": [
    "organism_ids = DATASET_NAME_TO_ORGANISM_IDS[\"athaliana\"]\n",
    "df_uniprot, df_uniprot_goa, graph_go_filtered, graph_chebi_filtered = preprocess_data(\n",
    "    organism_ids=organism_ids, datasets_folder_path=DATASET_FOLDER_PATH\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1995\n",
      "245\n",
      "244\n"
     ]
    }
   ],
   "source": [
    "# TODO rename? check what the method acutally does, do we need to change anything?\n",
    "\n",
    "df_substrate_overlaps, dict_chebi_to_uniprot = get_substrate_matrix(\n",
    "    datasets_folder_path=DATASET_FOLDER_PATH,\n",
    "    graph_chebi=graph_chebi_filtered,\n",
    "    graph_go=graph_go_filtered,\n",
    "    df_uniprot_goa=df_uniprot_goa,\n",
    "    min_overlap=0, max_overlap=int(1e6)\n",
    ")\n",
    "assert df_substrate_overlaps.shape[0] == len(dict_chebi_to_uniprot.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('monoatomic ion', 188),\n",
       " ('monoatomic cation', 162),\n",
       " ('inorganic cation', 159),\n",
       " ('metal cation', 83),\n",
       " ('proton', 82),\n",
       " ('organic anion', 70),\n",
       " ('carboxylic acid anion', 50),\n",
       " ('organic acid', 50),\n",
       " ('carbohydrate derivative', 46),\n",
       " ('carbohydrate', 43),\n",
       " ('inorganic anion', 40),\n",
       " ('biomacromolecule', 36),\n",
       " ('potassium(1+)', 35),\n",
       " ('chemical entity', 33),\n",
       " ('amino acid', 32),\n",
       " ('calcium(2+)', 27),\n",
       " ('transition element cation', 26),\n",
       " ('monoatomic anion', 22),\n",
       " ('amide', 22),\n",
       " ('water', 21),\n",
       " ('ion', 20),\n",
       " ('organic phosphate', 18),\n",
       " ('monosaccharide', 18),\n",
       " ('nitrate', 17),\n",
       " ('nucleotide-sugar', 16),\n",
       " ('sucrose', 15),\n",
       " ('L-alpha-amino acid zwitterion', 15),\n",
       " ('purine nucleotide', 14),\n",
       " ('sulfur molecular entity', 14),\n",
       " ('nucleotide', 14),\n",
       " ('adenyl nucleotide', 14),\n",
       " ('pyrimidine nucleotide-sugar', 13),\n",
       " ('purine ribonucleotide', 12),\n",
       " ('manganese cation', 11),\n",
       " ('ATP(4-)', 11),\n",
       " ('nucleobase', 10),\n",
       " ('nucleoside', 10),\n",
       " ('ADP(3-)', 10),\n",
       " ('hexose', 9),\n",
       " ('iron cation', 9),\n",
       " ('monocarboxylic acid', 9),\n",
       " ('peptide', 8),\n",
       " ('glucose', 8),\n",
       " ('inorganic phosphate', 8),\n",
       " ('phosphate ion', 8),\n",
       " ('chloride', 7),\n",
       " ('copper cation', 7),\n",
       " ('organic cation', 7),\n",
       " ('purine nucleobase', 7),\n",
       " ('sodium(1+)', 7),\n",
       " ('oligopeptide', 6),\n",
       " ('cadmium cation', 6),\n",
       " ('polyol', 5),\n",
       " ('purine nucleoside', 5),\n",
       " ('ammonium', 5),\n",
       " ('dicarboxylic acid', 5),\n",
       " ('UDP-D-glucose(2-)', 5),\n",
       " ('sulfate', 4),\n",
       " ('urea', 4),\n",
       " ('myo-inositol', 4),\n",
       " ('uracil', 4),\n",
       " ('pyrimidine nucleobase', 4),\n",
       " ('zinc(2+)', 4),\n",
       " ('L-glutamate(1-)', 4),\n",
       " ('L-lysinium(1+)', 4),\n",
       " ('aromatic amino acid', 4),\n",
       " ('L-proline zwitterion', 4),\n",
       " ('phosphoglycerate', 4),\n",
       " ('GDP-fucose', 4),\n",
       " ('malate(2-)', 3),\n",
       " ('allantoin', 3),\n",
       " ('arsenate ion', 3),\n",
       " ('glucoside', 3),\n",
       " ('fatty acid anion', 3),\n",
       " ('copper(2+)', 3),\n",
       " ('alcohol', 3),\n",
       " ('aldotriose phosphate', 3),\n",
       " (\"adenosine 5'-monophosphate(2-)\", 3),\n",
       " ('L-ornithinium(1+)', 3),\n",
       " ('copper(1+)', 3),\n",
       " ('UDP-alpha-D-xylose(2-)', 3),\n",
       " ('C4-dicarboxylate', 3),\n",
       " ('amino acid derivative', 3),\n",
       " ('NAD', 2),\n",
       " ('oxaloacetate(2-)', 2),\n",
       " ('glycerol', 2),\n",
       " ('glucosinolate', 2),\n",
       " ('pentose', 2),\n",
       " ('tricarboxylic acid', 2),\n",
       " ('galactose', 2),\n",
       " ('fructose', 2),\n",
       " ('amine', 2),\n",
       " ('boric acid', 2),\n",
       " ('molybdate', 2),\n",
       " ('magnesium cation', 2),\n",
       " ('dipeptide', 2),\n",
       " ('tripeptide', 2),\n",
       " ('L-histidine zwitterion', 2),\n",
       " ('methylammonium', 2),\n",
       " ('phytochelatin', 2),\n",
       " ('(+)-abscisic acid D-glucopyranosyl ester', 2),\n",
       " ('zinc cation', 2),\n",
       " ('azole', 2),\n",
       " ('D-glucose 6-phosphate', 1),\n",
       " ('xanthine', 1),\n",
       " ('choline', 1),\n",
       " ('pyruvate', 1),\n",
       " ('guanine', 1),\n",
       " ('adenine', 1),\n",
       " ('2-oxoglutarate(2-)', 1),\n",
       " ('citrate(3-)', 1),\n",
       " ('D-ribose', 1),\n",
       " ('maltose', 1),\n",
       " ('D-glucose', 1),\n",
       " ('salicin', 1),\n",
       " ('lipid', 1),\n",
       " ('xylose', 1),\n",
       " ('hydroquinone O-beta-D-glucopyranoside', 1),\n",
       " ('GDP-mannose', 1),\n",
       " ('arabinose', 1),\n",
       " ('gibberellin', 1),\n",
       " ('glutathione derivative', 1),\n",
       " ('organophosphorus compound', 1),\n",
       " ('quaternary nitrogen compound', 1),\n",
       " ('sulfur-containing amino acid', 1),\n",
       " ('arsenite(1-)', 1),\n",
       " ('glycolate', 1),\n",
       " ('mannitol', 1),\n",
       " ('heme', 1),\n",
       " ('glucitol', 1),\n",
       " ('L-argininium(1+)', 1),\n",
       " ('L-serine zwitterion', 1),\n",
       " ('glycerate', 1),\n",
       " ('serine zwitterion', 1),\n",
       " ('L-cystine zwitterion', 1),\n",
       " ('mannose', 1),\n",
       " ('L-ascorbate', 1),\n",
       " ('methotrexate(2-)', 1),\n",
       " ('polyamine macromolecule', 1),\n",
       " ('biotinate', 1),\n",
       " ('L-alanine zwitterion', 1),\n",
       " ('L-glutamine zwitterion', 1),\n",
       " ('phosphonatoenolpyruvate', 1),\n",
       " ('S-adenosyl-L-methionine zwitterion', 1),\n",
       " ('gamma-aminobutyric acid zwitterion', 1),\n",
       " ('2-cis-abscisate', 1),\n",
       " ('folate(2-)', 1),\n",
       " ('5-formyltetrahydrofolate(2-)', 1),\n",
       " ('alanine zwitterion', 1),\n",
       " ('fatty acyl-CoA(4-)', 1)]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chebi_name_to_term = {name : term for term, name in graph_chebi_filtered.nodes(data=\"name\")}\n",
    "chebi_term_to_name = {term : name for term, name in graph_chebi_filtered.nodes(data=\"name\")}\n",
    "molecule_counts = {\n",
    "    chebi_term_to_name[term] : len(proteins) for term, proteins in dict_chebi_to_uniprot.items()\n",
    "}\n",
    "sorted(molecule_counts.items(), key=lambda item: item[1], reverse=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniprot</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Q84WG1</th>\n",
       "      <td>potassium(1+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q93VD3</th>\n",
       "      <td>potassium(1+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9M7K4</th>\n",
       "      <td>potassium(1+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9LD18</th>\n",
       "      <td>potassium(1+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q8LBL1</th>\n",
       "      <td>potassium(1+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q945S5</th>\n",
       "      <td>calcium(2+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q39253</th>\n",
       "      <td>calcium(2+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9LEQ3</th>\n",
       "      <td>calcium(2+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q39254</th>\n",
       "      <td>calcium(2+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q9SZR1</th>\n",
       "      <td>calcium(2+)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 label\n",
       "Uniprot               \n",
       "Q84WG1   potassium(1+)\n",
       "Q93VD3   potassium(1+)\n",
       "Q9M7K4   potassium(1+)\n",
       "Q9LD18   potassium(1+)\n",
       "Q8LBL1   potassium(1+)\n",
       "...                ...\n",
       "Q945S5     calcium(2+)\n",
       "Q39253     calcium(2+)\n",
       "Q9LEQ3     calcium(2+)\n",
       "Q39254     calcium(2+)\n",
       "Q9SZR1     calcium(2+)\n",
       "\n",
       "[62 rows x 1 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO sequence clustering?\n",
    "# TODO turn into function labels -> df_labels\n",
    "\n",
    "labels = \"potassium(1+)\", 'calcium(2+)'\n",
    "\n",
    "protein_to_label = list()\n",
    "for label in labels:\n",
    "    label_proteins = dict_chebi_to_uniprot[chebi_name_to_term[label]]\n",
    "    for protein in label_proteins:\n",
    "        protein_to_label.append([protein, label])\n",
    "\n",
    "df_labels = pd.DataFrame.from_records(protein_to_label,columns=[\"Uniprot\", \"label\"], index=\"Uniprot\")\n",
    "\n",
    "df_labels"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create sequence-based features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sequences = df_uniprot.loc[df_labels.index].sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_aac = calculate_aac(df_sequences)\n",
    "df_paac = calculate_paac(df_sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pssm_50_1 = calculate_pssm_feature(\n",
    "    df_sequences,\n",
    "    tmp_folder=\"../data/intermediate/blast/pssm_uniref50_1it\",\n",
    "    blast_db=\"../data/raw/uniref/uniref50/uniref50.fasta\",\n",
    "    iterations=1,\n",
    "    psiblast_threads=16,\n",
    "    verbose=False\n",
    ")\n",
    "df_pssm_50_3 = calculate_pssm_feature(\n",
    "    df_sequences,\n",
    "    tmp_folder=\"../data/intermediate/blast/pssm_uniref50_3it\",\n",
    "    blast_db=\"../data/raw/uniref/uniref50/uniref50.fasta\",\n",
    "    iterations=3,\n",
    "    psiblast_threads=16,\n",
    "    verbose=False\n",
    ")\n",
    "df_pssm_90_1 = calculate_pssm_feature(\n",
    "    df_sequences,\n",
    "    tmp_folder=\"../data/intermediate/blast/pssm_uniref90_3it\",\n",
    "    blast_db=\"../data/raw/uniref/uniref90/uniref90.fasta\",\n",
    "    iterations=1,\n",
    "    psiblast_threads=16,\n",
    "    verbose=False\n",
    ")\n",
    "df_pssm_90_3 = calculate_pssm_feature(\n",
    "    df_sequences,\n",
    "    tmp_folder=\"../data/intermediate/blast/pssm_uniref90_3it\",\n",
    "    blast_db=\"../data/raw/uniref/uniref90/uniref90.fasta\",\n",
    "    iterations=3,\n",
    "    psiblast_threads=16,\n",
    "    verbose=False\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "subpred4",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
