{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Training machine learning models on pairs of substrates in individual organisms\n",
    "\n",
    "After creating the set of primary ChEBI substrates from a set of transmembrane transport proteins, the next task is to create a pipeline for generating these datasets automatically. Next, the "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generalized dataset creation method\n",
    "\n",
    "- filter uniprot proteins generally\n",
    "- filter again for transmembrane transporter activity\n",
    "- get network of GO terms that are descendants of transmembrane transporter activity\n",
    "    - filter network for proteins with in dataset with term\n",
    "    - annotate with number of proteins\n",
    "    - annotate with ChEBI substrates\n",
    "- get chebi network\n",
    "    - filter for GO terms in filtered GO network\n",
    "- 70% sequence identity\n",
    "- feature generation\n",
    "- svm pipeline\n",
    "\n",
    "End results: \n",
    "- Stats for organisms, substrates, features, train/test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from subpred.util import load_df\n",
    "from subpred.graph import preprocess_data, get_substrate_matrix\n",
    "from subpred.pssm import calculate_pssm_feature\n",
    "from subpred.compositions import calculate_aac, calculate_paac\n",
    "import pandas as pd\n",
    "from subpred.cdhit import cd_hit\n",
    "from sklearn.pipeline import make_pipeline\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.preprocessing import StandardScaler, LabelEncoder\n",
    "from sklearn.feature_selection import SelectPercentile\n",
    "from sklearn.model_selection import GridSearchCV, train_test_split\n",
    "from sklearn.metrics import classification_report\n",
    "from subpred.custom_transformers import FeatureCombinator, get_feature_type_combinations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_classification_task(\n",
    "    organism_ids: set,\n",
    "    labels: set,\n",
    "    clustering_threshold: int = None,\n",
    "    dataset_folder_path: str = \"../data/datasets\",\n",
    ") -> pd.DataFrame:\n",
    "    # TODO handling for multi-substrate\n",
    "    # TODO ability to use go terms or chebi terms (compare sample count, performance)\n",
    "\n",
    "    (\n",
    "        df_uniprot,\n",
    "        df_uniprot_goa,\n",
    "        graph_go_filtered,\n",
    "        graph_chebi_filtered,\n",
    "    ) = preprocess_data(\n",
    "        organism_ids=organism_ids, datasets_folder_path=dataset_folder_path\n",
    "    )\n",
    "    # TODO go through method code\n",
    "    df_substrate_overlaps, dict_chebi_to_uniprot = get_substrate_matrix(\n",
    "        datasets_folder_path=dataset_folder_path,\n",
    "        graph_chebi=graph_chebi_filtered,\n",
    "        graph_go=graph_go_filtered,\n",
    "        df_uniprot_goa=df_uniprot_goa,\n",
    "        min_overlap=0,\n",
    "        max_overlap=int(1e6),\n",
    "    )\n",
    "    assert df_substrate_overlaps.shape[0] == len(dict_chebi_to_uniprot.keys())\n",
    "    chebi_name_to_term = {\n",
    "        name: term for term, name in graph_chebi_filtered.nodes(data=\"name\")\n",
    "    }\n",
    "    chebi_term_to_name = {\n",
    "        term: name for term, name in graph_chebi_filtered.nodes(data=\"name\")\n",
    "    }\n",
    "    molecule_counts = {\n",
    "        chebi_term_to_name[term]: len(proteins)\n",
    "        for term, proteins in dict_chebi_to_uniprot.items()\n",
    "    }\n",
    "    print(sorted(molecule_counts.items(), key=lambda item: item[1], reverse=True))\n",
    "\n",
    "    protein_to_label = list()\n",
    "    for label in labels:\n",
    "        label_proteins = dict_chebi_to_uniprot[chebi_name_to_term[label]]\n",
    "        for protein in label_proteins:\n",
    "            protein_to_label.append([protein, label])\n",
    "\n",
    "    df_labels = pd.DataFrame.from_records(\n",
    "        protein_to_label, columns=[\"Uniprot\", \"label\"], index=\"Uniprot\"\n",
    "    )\n",
    "\n",
    "    df_labels = df_labels[~df_labels.index.duplicated()]  # TODO series?\n",
    "    print(df_labels.label.value_counts())\n",
    "    df_sequences = df_uniprot.loc[df_labels.index].sequence.to_frame()\n",
    "    print(\"number of sequences\", df_sequences.shape[0])\n",
    "    if clustering_threshold:\n",
    "        cluster_representatives = cd_hit(\n",
    "            df_sequences.sequence, identity_threshold=clustering_threshold\n",
    "        )\n",
    "        print(cluster_representatives)\n",
    "        df_sequences = df_sequences.loc[cluster_representatives]\n",
    "        df_labels = df_labels.loc[cluster_representatives]\n",
    "    return pd.concat([df_sequences, df_labels], axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_features(series_sequences:pd.Series):\n",
    "    df_aac = calculate_aac(series_sequences)\n",
    "    df_paac = calculate_paac(series_sequences)\n",
    "    df_pssm_50_1 = calculate_pssm_feature(\n",
    "        series_sequences,\n",
    "        tmp_folder=\"../data/intermediate/blast/pssm_uniref50_1it\",\n",
    "        blast_db=\"../data/raw/uniref/uniref50/uniref50.fasta\",\n",
    "        iterations=1,\n",
    "        psiblast_threads=-1,\n",
    "        verbose=False,\n",
    "        feature_name=\"PSSM_50_1\"\n",
    "    )\n",
    "    df_pssm_50_3 = calculate_pssm_feature(\n",
    "        series_sequences,\n",
    "        tmp_folder=\"../data/intermediate/blast/pssm_uniref50_3it\",\n",
    "        blast_db=\"../data/raw/uniref/uniref50/uniref50.fasta\",\n",
    "        iterations=3,\n",
    "        psiblast_threads=-1,\n",
    "        verbose=False,\n",
    "        feature_name=\"PSSM_50_3\"\n",
    "    )\n",
    "    df_pssm_90_1 = calculate_pssm_feature(\n",
    "        series_sequences,\n",
    "        tmp_folder=\"../data/intermediate/blast/pssm_uniref90_3it\",\n",
    "        blast_db=\"../data/raw/uniref/uniref90/uniref90.fasta\",\n",
    "        iterations=1,\n",
    "        psiblast_threads=-1,\n",
    "        verbose=False,\n",
    "        feature_name=\"PSSM_90_1\"\n",
    "    )\n",
    "    df_pssm_90_3 = calculate_pssm_feature(\n",
    "        series_sequences,\n",
    "        tmp_folder=\"../data/intermediate/blast/pssm_uniref90_3it\",\n",
    "        blast_db=\"../data/raw/uniref/uniref90/uniref90.fasta\",\n",
    "        iterations=3,\n",
    "        psiblast_threads=-1,\n",
    "        verbose=False,\n",
    "        feature_name=\"PSSM_90_3\"\n",
    "    )\n",
    "    df_features = pd.concat(\n",
    "        [\n",
    "            df_aac,\n",
    "            df_paac,\n",
    "            df_pssm_50_1,\n",
    "            df_pssm_50_3,\n",
    "            df_pssm_90_1,\n",
    "            df_pssm_90_3,\n",
    "        ], axis=1\n",
    "    )\n",
    "    return df_features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Eval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO try removing worst sample according to percentages\n",
    "# TODO feature selection, regularization\n",
    "# TODO cd-hit\n",
    "# TODO determinism\n",
    "# TODO also comparative analysis of features?\n",
    "# TODO compare to protein embeddings and BLAST\n",
    "# TODO parameter for using featurecombinator\n",
    "# TODO separate functions\n",
    "\n",
    "def evaluate(df_dataset, df_features):\n",
    "\n",
    "    # converting data to numpy\n",
    "    label_encoder = LabelEncoder()\n",
    "    label_encoder.fit(sorted(df_dataset.label.unique()))\n",
    "    sample_names = df_features.index.values\n",
    "    feature_names = df_features.columns.values\n",
    "    X = df_features.values\n",
    "    y = label_encoder.transform(df_dataset.label)\n",
    "    # train test eval split\n",
    "    (\n",
    "        X_train,\n",
    "        X_eval,\n",
    "        y_train,\n",
    "        y_eval,\n",
    "        sample_names_train,\n",
    "        sample_names_eval,\n",
    "    ) = train_test_split(X, y, sample_names, test_size=0.2, random_state=1, stratify=y)\n",
    "\n",
    "    feature_type_combinations = get_feature_type_combinations(feature_names=feature_names)\n",
    "    feature_combinator = FeatureCombinator(feature_names=df_features.columns)\n",
    "    model = make_pipeline(\n",
    "        StandardScaler(), feature_combinator, SVC(random_state=1, probability=True)\n",
    "    )\n",
    "    param_grid = {\n",
    "        \"svc__C\": [0.1, 1, 10],\n",
    "        # \"svc__gamma\": [\"scale\", \"auto\"],\n",
    "        # \"svc__class_weight\": [\"balanced\", None],\n",
    "        \"featurecombinator__feature_types\": feature_type_combinations,\n",
    "        # \"selectpercentile__percentile\": list(range(1, 101, 5)),\n",
    "    }\n",
    "\n",
    "    # hyperparam optim & crossval\n",
    "    gridsearch = GridSearchCV(\n",
    "        estimator=model,\n",
    "        param_grid=param_grid,\n",
    "        scoring=\"f1\",\n",
    "        cv=5,\n",
    "        n_jobs=-1,\n",
    "        return_train_score=True,\n",
    "        # verbose=20\n",
    "    )\n",
    "    gridsearch.fit(X_train, y_train)\n",
    "    print(\"Best train score:\", gridsearch.best_score_)\n",
    "    print(\"Best train params\", gridsearch.best_params_)\n",
    "    model_optim = gridsearch.best_estimator_\n",
    "\n",
    "    # eval\n",
    "    y_pred = model_optim.predict(X_eval)\n",
    "    print(classification_report(y_true=y_eval, y_pred=y_pred))\n",
    "    print(model_optim.predict_proba(X_eval))  # TODO compare with actual labels\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(organism_ids:set, labels: set):\n",
    "    # TODO get rid of unnecessary prints\n",
    "    df_dataset = get_classification_task(\n",
    "        organism_ids=organism_ids,\n",
    "        labels=labels,\n",
    "        clustering_threshold=70,\n",
    "    )\n",
    "\n",
    "    df_features = get_features(df_dataset.sequence)\n",
    "\n",
    "    df_features = df_features.loc[df_features.index.sort_values()]\n",
    "    df_dataset = df_dataset.loc[df_features.index]\n",
    "    df_features = df_features.loc[df_dataset.index]\n",
    "\n",
    "    print(df_dataset.shape, df_features.shape)\n",
    "\n",
    "    evaluate(df_dataset=df_dataset, df_features=df_features)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "human\n",
      "43248\n",
      "164519\n",
      "60547\n",
      "1995\n",
      "477\n",
      "474\n",
      "[('monoatomic ion', 1455), ('monoatomic cation', 1138), ('inorganic cation', 1061), ('metal cation', 852), ('calcium(2+)', 356), ('potassium(1+)', 289), ('organic anion', 266), ('sodium(1+)', 239), ('monoatomic anion', 238), ('inorganic anion', 234), ('proton', 197), ('chloride', 187), ('organic acid', 186), ('carboxylic acid anion', 183), ('chemical entity', 144), ('amino acid', 111), ('L-alpha-amino acid zwitterion', 86), ('carbohydrate derivative', 78), ('sulfur molecular entity', 66), ('ion', 58), ('monocarboxylic acid', 58), ('amide', 55), ('transition element cation', 49), ('carbohydrate', 40), ('organic cation', 40), ('amino acid derivative', 38), ('organic phosphate', 35), ('nucleotide-sugar', 32), ('biomacromolecule', 32), ('nucleotide', 31), ('pyrimidine nucleotide-sugar', 30), ('dicarboxylic acid', 28), ('purine nucleotide', 27), ('monosaccharide', 27), ('glucose', 25), ('hexose', 25), ('hydrogencarbonate', 24), ('adenyl nucleotide', 24), ('purine ribonucleotide', 23), ('zinc(2+)', 23), ('quaternary nitrogen compound', 22), ('ammonium', 21), ('bile acid', 18), ('sulfate', 17), ('water', 16), ('riboflavin(1-)', 16), ('peptide', 15), ('magnesium cation', 15), ('C4-dicarboxylate', 15), ('alanine zwitterion', 15), ('ATP(4-)', 14), ('azole', 14), ('icosanoid', 13), ('L-argininium(1+)', 13), ('nucleoside', 13), ('ADP(3-)', 13), ('urea', 12), ('glutathione derivative', 12), ('prostaglandin', 12), ('branched-chain amino acid', 11), ('copper cation', 11), ('polyol', 11), ('sulfur-containing amino acid', 11), ('monoamine', 11), ('choline', 10), ('D-glucose', 10), ('phosphate', 10), ('L-glutamate(1-)', 10), ('aromatic amino acid', 10), ('glycine zwitterion', 10), ('gamma-aminobutyric acid zwitterion', 10), ('lipid', 9), ('vitamin B1', 9), ('fatty acid anion', 9), ('L-aspartate(1-)', 9), ('heme', 9), ('L-lysinium(1+)', 9), ('serine zwitterion', 9), ('L-glutamine zwitterion', 9), ('inorganic phosphate', 8), ('iron cation', 8), ('manganese cation', 8), ('fructose', 8), ('oxalate(2-)', 8), ('phosphate ion', 8), ('copper(1+)', 8), ('polyamine macromolecule', 8), ('L-alanine zwitterion', 8), ('glutamate(1-)', 7), ('glycerol', 7), ('nucleobase', 7), ('proline', 7), ('alcohol', 7), ('L-serine zwitterion', 7), ('dehydroascorbide(1-)', 7), ('L-proline zwitterion', 7), ('carnitine', 6), ('oligopeptide', 6), ('purine nucleobase', 6), ('pyrimidine nucleoside', 6), ('succinate(2-)', 6), ('retinol', 6), ('L-leucine zwitterion', 6), ('folate(2-)', 6), ('zinc cation', 6), ('alkanesulfonate oxoanion', 5), ('pyruvate', 5), ('lactate', 5), ('iron(2+)', 5), ('oleate', 5), ('L-ornithinium(1+)', 5), ('taurine zwitterion', 5), ('glutathionate(1-)', 5), ('creatine zwitterion', 5), ('carbon dioxide', 4), ('uridine', 4), ('2-oxoglutarate(2-)', 4), ('citrate(3-)', 4), ('myo-inositol', 4), ('sucrose', 4), ('tricarboxylic acid', 4), ('galactose', 4), ('1,4-butanediammonium', 4), ('L-cystine zwitterion', 4), ('dipeptide', 4), ('L-histidine zwitterion', 4), ('spermidine(3+)', 4), ('L-asparagine zwitterion', 4), ('guanyl nucleotide', 4), ('acetylcholine', 3), ('malate(2-)', 3), ('(R)-carnitine', 3), ('iodide', 3), ('oxaloacetate(2-)', 3), ('adenine', 3), ('glucoside', 3), ('pyrimidine nucleobase', 3), ('amine', 3), ('serotonin(1+)', 3), ('cysteine zwitterion', 3), ('methotrexate(2-)', 3), ('coenzyme A(4-)', 3), ('acetyl-CoA(4-)', 3), ('biotinate', 3), ('UDP-N-acetyl-alpha-D-glucosamine(2-)', 3), ('L-tryptophan zwitterion', 3), ('L-isoleucine zwitterion', 3), ('short-chain fatty acid anion', 3), ('dopaminium(1+)', 3), ('cadmium cation', 3), ('(R)-noradrenaline(1+)', 3), ('NAD', 2), ('D-glucose 6-phosphate', 2), ('sterol', 2), ('O-acylcarnitine', 2), ('cobalt cation', 2), ('flavin adenine dinucleotide', 2), ('organophosphorus compound', 2), ('pentose', 2), ('purine nucleoside', 2), ('pyrimidine nucleotide', 2), ('copper(2+)', 2), ('cobalamin', 2), ('nucleic acid', 2), ('ribonucleic acid', 2), ('fucose', 2), ('aspartate(1-)', 2), ('L-ascorbate', 2), (\"adenosine 5'-monophosphate(2-)\", 2), ('hexose phosphate', 2), ('L-phenylalanine zwitterion', 2), (\"3'-phosphonato-5'-adenylyl sulfate(4-)\", 2), ('thiamine(1+) diphosphate(3-)', 2), ('alpha-glucoside', 2), ('3-sulfino-L-alanine(1-)', 2), ('GDP-fucose', 2), ('c-GMP-AMP(2-)', 2), ('peptidoglycan', 2), ('formate', 1), ('D-glucuronate', 1), ('D-fructose', 1), ('D-mannose', 1), ('guanine', 1), ('nitric oxide', 1), ('monoacylglycerol phosphate', 1), ('propionate', 1), ('cytidine', 1), ('nitrate', 1), ('glycine betaine', 1), ('glucosiduronic acid', 1), ('mevalonate', 1), ('nickel cation', 1), ('sialic acid', 1), ('tricarboxylic acid trianion', 1), ('(R)-pantothenate', 1), ('arsenite(1-)', 1), ('fumarate(2-)', 1), ('D-aspartate(1-)', 1), ('iron-sulfur cluster', 1), ('nicotinate', 1), ('boric acid', 1), ('vanadium cation', 1), ('cystine zwitterion', 1), ('purine', 1), ('5-ammoniolevulinate', 1), ('molybdate', 1), ('mannose', 1), ('tripeptide', 1), ('lithium(1+)', 1), ('UDP-alpha-D-xylose(2-)', 1), ('L-valine zwitterion', 1), ('CMP-N-acetyl-beta-neuraminate(2-)', 1), ('L-methionine zwitterion', 1), ('L-threonine zwitterion', 1), ('L-kynurenine zwitterion', 1), ('beta-alanine zwitterion', 1), ('bilirubin(2-)', 1), ('FMN(3-)', 1), ('L-tyrosine zwitterion', 1), ('UDP-D-glucose(2-)', 1), ('methylammonium', 1), ('S-adenosyl-L-methionine zwitterion', 1), ('D-alpha-amino acid zwitterion', 1), ('lead ion', 1), ('silver cation', 1), ('chromium ion', 1), ('4-hydroxyproline zwitterion', 1), ('fatty acyl-CoA(4-)', 1), ('enterobactin(1-)', 1), ('tetracycline zwitterion', 1)]\n",
      "label\n",
      "calcium(2+)      356\n",
      "potassium(1+)    266\n",
      "Name: count, dtype: int64\n",
      "number of sequences 622\n",
      "cd-hit: clustered 622 sequences into 373 clusters at threshold 70\n",
      "['Q14524', 'P16615', 'Q14957', 'Q86TL2', 'Q96GC9', 'Q9GZZ6', 'P62955', 'Q9H1D0', 'Q02641', 'Q9P0L9', 'B4E2Q0', 'P54284', 'Q14643', 'Q01814', 'Q9H3K2', 'O00305', 'Q8WXH2', 'Q7Z2W7', 'P16298', 'Q96MG2', 'P21917', 'A0A140LJL2', 'Q9NR96', 'Q01484', 'P01160', 'Q7Z442', 'G0XQ39', 'P49768', 'Q15858', 'P11532', 'Q8IU99', 'P54257', 'Q9BRQ5', 'P23634', 'Q8NE86', 'P62942', 'P34998', 'O15399', 'A0A1B0GTK1', 'P29475', 'O60733', 'Q9P246', 'P16885', 'Q8NET8', 'A0A7P0T8M7', 'A0A1B0GTN0', 'Q05586', 'Q96P56', 'Q9NZN3', 'Q9NZQ8', 'Q09666', 'P0DP23', 'P57796', 'Q9HBA0', 'Q71RS6', 'Q8IZK6', 'P35499', 'Q06432', 'P06756', 'Q07343', 'J3QTB0', 'Q12879', 'Q13698', 'Q9NZV5', 'Q9UI40', 'Q7Z4N2', 'P42858', 'A0A1C7CYY9', 'P62166', 'Q01118', 'P51674', 'Q9HDC5', 'A0A1B0GV16', 'K7EJY1', 'Q96LZ3', 'Q9NZU7', 'P32418', 'Q6J4K2', 'P15382', 'P56704', 'Q13557', 'Q96D31', 'O95202', 'Q6YHU6', 'Q14644', 'P11413', 'Q9Y5S1', 'Q9NPB3', 'A0A087X080', 'Q93034', 'F5H483', 'Q9Y5Y9', 'Q13507', 'Q92736', 'Q99469', 'Q96MF2', 'P36544', 'Q14571', 'Q9UBN4', 'P0DN84', 'Q9BXT2', 'E9PJN4', 'P31415', 'Q8TDX9', 'P98161', 'P35498', 'Q9GZU1', 'Q15077', 'B1AL88', 'E7EMC0', 'Q08289', 'A0A1B0GU42', 'O60840', 'Q6XR72', 'Q12797', 'O00631', 'P55042', 'P17612', 'Q8TEL6', 'Q6ZMT1', 'P08133', 'O60721', 'Q9Y6J6', 'F5GY28', 'Q7Z3S7', 'P26678', 'Q99571', 'Q96SN7', 'P28161', 'P57103', 'O75762', 'Q8TCU5', 'P05026', 'Q9BR39', 'O75052', 'P28222', 'O14958', 'P30626', 'Q9NP86', 'P61619', 'Q8TDD5', 'Q13061', 'O75628', 'Q9P0X4', 'P07355', 'A0A804HKC4', 'Q13023', 'Q15878', 'Q8WVX3', 'Q7Z443', 'A0A804HIR0', 'Q8IZS8', 'Q9BSW2', 'Q96RD7', 'P05067', 'P23327', 'P35372', 'Q8TD43', 'P54652', 'Q8NFF2', 'Q08499', 'Q12981', 'P55916', 'O43497', 'P14416', 'A0A1B0GTP6', 'P08195', 'Q00975', 'Q9UMX0', 'Q14573', 'P62258', 'Q8NHX9', 'P08913', 'B7ZMI3', 'P63165', 'Q7RTX7', 'Q9UGM1', 'P19634', 'O95153', 'Q8NER1', 'O95180', 'P78417', 'Q13224', 'Q8TDI8', 'Q9HC07', 'Q13563', 'P48995', 'O14775', 'Q96JJ6', 'Q8NEC5', 'Q9Y6N3', 'P84074', 'P49281', 'Q9UI33', 'Q9UBN1', 'A0A1X7SBR8', 'P54289', 'O15247', 'Q9Y698', 'O75323', 'P41146', 'C9JVC9', 'Q9Y210', 'P24588', 'P21817', 'Q9BX84', 'E9PGK7', 'O75185', 'Q9UM00', 'P56539', 'Q00535', 'Q86VW0', 'Q15493', 'Q9ULQ1', 'Q96QT4', 'P55040', 'Q9UL62', 'Q06787', 'O00499', 'Q86XK2', 'Q9NSA2', 'P56696', 'P49286', 'Q9ULS6', 'O60939', 'P35609', 'A0A1W2PPZ1', 'Q9NPC2', 'P51164', 'P48050', 'Q9NZI2', 'P55017', 'P17658', 'O00308', 'P48764', 'Q9BSA9', 'A0A804HHV7', 'A0A1W2PNZ3', 'P78508', 'Q9UL51', 'B7U540', 'P54709', 'Q8N7C0', 'O43448', 'O43687', 'O60928', 'Q6PIU1', 'P57789', 'A0A804HIW7', 'Q8N5I3', 'Q9Y666', 'Q13642', 'Q92581', 'Q9UNX9', 'Q9Y2E8', 'Q96KK3', 'Q9Y376', 'P54710', 'Q03135', 'Q9H252', 'P07492', 'O14649', 'A0A0D9SEY3', 'Q9UJ96', 'Q9Y3Q4', 'A0A590UJ62', 'P22001', 'Q9NPI9', 'Q9NS40', 'A0A590UJU4', 'Q86W47', 'O43603', 'Q96ER9', 'P51787', 'Q8TDN1', 'A0AV02', 'P48048', 'Q14722', 'O00180', 'Q5SVJ8', 'Q9UEW8', 'Q5VT99', 'Q96PR1', 'Q9NZC2', 'O95069', 'Q9BQ31', 'Q9Y2W7', 'A0A590UJI8', 'A0A0A0MT26', 'P39086', 'Q9H6F2', 'Q9UK17', 'O43526', 'P05556', 'A8MYU2', 'O60706', 'Q9Y6H6', 'K7EKU4', 'Q12959', 'Q5I7T1', 'Q9P1Z3', 'Q9UIX4', 'P22466', 'Q9H427', 'Q96T83', 'Q86WK6', 'O95279', 'Q9Y691', 'Q16558', 'O95259', 'Q2I0M4', 'P48051', 'A0A0D9SG49', 'A0A3F2YNY5', 'P50570', 'Q15842', 'Q13621', 'Q7Z418', 'Q9NPA1', 'P29972', 'Q99712', 'Q6ZSA7', 'F8WEA4', 'P63027', 'Q9BXP2', 'Q12809', 'Q96T55', 'O43525', 'A0A087WYJ0', 'Q6TFL4', 'Q6ZRF8', 'Q92806', 'P54707', 'P55011', 'Q8TDN2', 'Q9Y257', 'Q14940', 'P14415', 'P48549', 'Q14721', 'Q8IVB4', 'Q9UHW9', 'Q9ULD8', 'Q9UJ90', 'Q6UVM3', 'Q6PIL6', 'P60880', 'Q8WWG9', 'Q9NYG8', 'P46934', 'P22459', 'Q99996', 'Q9NS61', 'Q9Y2U2', 'Q03721', 'Q12955', 'O95747', 'Q6UWJ1', 'Q16322', 'Q9H2X9', 'Q9NZV8', 'Q8NCM2', 'P55085', 'Q09428', 'Q96PU5', 'O15554', 'Q8TAE7', 'P22460', 'P20648', 'Q9NVV0', 'O60741', 'Q9H3M0', 'Q96SW2']\n"
     ]
    }
   ],
   "source": [
    "labels = {\"potassium(1+)\", 'calcium(2+)'}\n",
    "\n",
    "dataset_name_to_organism_ids = {\"human\": {9606}, \"athaliana\": {3702}, \"ecoli\": {83333}, \"yeast\": {559292}}\n",
    "dataset_name_to_organism_ids[\"all\"] = {list(s)[0] for s in dataset_name_to_organism_ids.values() if len(s) == 1}\n",
    "\n",
    "for dataset_name, organism_ids in dataset_name_to_organism_ids.items():\n",
    "    print(dataset_name)\n",
    "    main(organism_ids=organism_ids, labels=labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Comparisons\n",
    "\n",
    "Compare training results with: \n",
    "\n",
    "- Average sequence similarity\n",
    "- GO term similarity\n",
    "  - How many protein in common?\n",
    "  - Semantic similarity?\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "subpred4",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
