{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clustering the protein dataset\n",
    "\n",
    "The goal is to cluster the data with several different methods, in order to find the annotations that match the clusters thee best. "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### TODOs\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating the datasets"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Protein datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found pickle, reading...\n",
      "(1021957, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_names</th>\n",
       "      <th>protein_names</th>\n",
       "      <th>reviewed</th>\n",
       "      <th>protein_existence</th>\n",
       "      <th>sequence</th>\n",
       "      <th>organism_id</th>\n",
       "      <th>go_ids</th>\n",
       "      <th>keyword_ids</th>\n",
       "      <th>keywords</th>\n",
       "      <th>tcdb_id</th>\n",
       "      <th>interpro</th>\n",
       "      <th>tcdb_class</th>\n",
       "      <th>tcdb_substrates</th>\n",
       "      <th>keywords_substrates</th>\n",
       "      <th>keywords_transport_related</th>\n",
       "      <th>keywords_location</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniprot</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A0A0C5B5G6</th>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>Mitochondrial-derived peptide MOTS-c (Mitochon...</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Evidence at protein level</td>\n",
       "      <td>MRWQEMGYIFYPRKLR</td>\n",
       "      <td>9606</td>\n",
       "      <td>GO:0001649; GO:0003677; GO:0005615; GO:0005634...</td>\n",
       "      <td>KW-0238; KW-0496; KW-0539; KW-0892; KW-1185; K...</td>\n",
       "      <td>DNA-binding;Mitochondrion;Nucleus;Osteogenesis...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A1B0GTW7</th>\n",
       "      <td>CIROP LMLN2</td>\n",
       "      <td>Ciliated left-right organizer metallopeptidase...</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Evidence at protein level</td>\n",
       "      <td>MLLLLLLLLLLPPLVLRVAASRCLHDETQKSVSLLRPPFSQLPSKS...</td>\n",
       "      <td>9606</td>\n",
       "      <td>GO:0004222; GO:0005737; GO:0006508; GO:0007155...</td>\n",
       "      <td>KW-0025; KW-0225; KW-0325; KW-1056; KW-0378; K...</td>\n",
       "      <td>Alternative splicing;Disease variant;Glycoprot...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IPR001577;</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0JNW5</th>\n",
       "      <td>BLTP3B KIAA0701 SHIP164 UHRF1BP1L</td>\n",
       "      <td>Bridge-like lipid transfer protein family memb...</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Evidence at protein level</td>\n",
       "      <td>MAGIIKKQILKHLSRFTKNLSPDKINLSTLKGEGELKNLELDEEVL...</td>\n",
       "      <td>9606</td>\n",
       "      <td>GO:0005769; GO:0005829; GO:0034498; GO:0042803...</td>\n",
       "      <td>KW-0025; KW-0175; KW-0963; KW-0967; KW-0597; K...</td>\n",
       "      <td>Alternative splicing;Coiled coil;Cytoplasm;End...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IPR026728;IPR026854;</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0JP26</th>\n",
       "      <td>POTEB3</td>\n",
       "      <td>POTE ankyrin domain family member B3</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Evidence at protein level</td>\n",
       "      <td>MVAEVCSMPAASAVKKPFDLRSKMGKWCHHRFPCCRGSGKSNMGTS...</td>\n",
       "      <td>9606</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KW-0025; KW-0040; KW-0175; KW-1185; KW-0677</td>\n",
       "      <td>Alternative splicing;ANK repeat;Coiled coil;Re...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IPR002110;IPR036770;IPR039497;</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0PK11</th>\n",
       "      <td>CLRN2</td>\n",
       "      <td>Clarin-2</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Evidence at protein level</td>\n",
       "      <td>MPGWFKKAWYGLASLLSFSSFILIIVALVVPHWLSGKILCQTGVDL...</td>\n",
       "      <td>9606</td>\n",
       "      <td>GO:0007605; GO:0032421; GO:0060088; GO:0060171...</td>\n",
       "      <td>KW-1003; KW-0966; KW-0209; KW-0225; KW-0325; K...</td>\n",
       "      <td>Cell membrane;Cell projection;Deafness;Disease...</td>\n",
       "      <td>9.A.46.1.2</td>\n",
       "      <td>IPR026748;</td>\n",
       "      <td>9.A</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   gene_names  \\\n",
       "Uniprot                                         \n",
       "A0A0C5B5G6                            MT-RNR1   \n",
       "A0A1B0GTW7                        CIROP LMLN2   \n",
       "A0JNW5      BLTP3B KIAA0701 SHIP164 UHRF1BP1L   \n",
       "A0JP26                                 POTEB3   \n",
       "A0PK11                                  CLRN2   \n",
       "\n",
       "                                                protein_names  reviewed  \\\n",
       "Uniprot                                                                   \n",
       "A0A0C5B5G6  Mitochondrial-derived peptide MOTS-c (Mitochon...  reviewed   \n",
       "A0A1B0GTW7  Ciliated left-right organizer metallopeptidase...  reviewed   \n",
       "A0JNW5      Bridge-like lipid transfer protein family memb...  reviewed   \n",
       "A0JP26                   POTE ankyrin domain family member B3  reviewed   \n",
       "A0PK11                                               Clarin-2  reviewed   \n",
       "\n",
       "                    protein_existence  \\\n",
       "Uniprot                                 \n",
       "A0A0C5B5G6  Evidence at protein level   \n",
       "A0A1B0GTW7  Evidence at protein level   \n",
       "A0JNW5      Evidence at protein level   \n",
       "A0JP26      Evidence at protein level   \n",
       "A0PK11      Evidence at protein level   \n",
       "\n",
       "                                                     sequence  organism_id  \\\n",
       "Uniprot                                                                      \n",
       "A0A0C5B5G6                                   MRWQEMGYIFYPRKLR         9606   \n",
       "A0A1B0GTW7  MLLLLLLLLLLPPLVLRVAASRCLHDETQKSVSLLRPPFSQLPSKS...         9606   \n",
       "A0JNW5      MAGIIKKQILKHLSRFTKNLSPDKINLSTLKGEGELKNLELDEEVL...         9606   \n",
       "A0JP26      MVAEVCSMPAASAVKKPFDLRSKMGKWCHHRFPCCRGSGKSNMGTS...         9606   \n",
       "A0PK11      MPGWFKKAWYGLASLLSFSSFILIIVALVVPHWLSGKILCQTGVDL...         9606   \n",
       "\n",
       "                                                       go_ids  \\\n",
       "Uniprot                                                         \n",
       "A0A0C5B5G6  GO:0001649; GO:0003677; GO:0005615; GO:0005634...   \n",
       "A0A1B0GTW7  GO:0004222; GO:0005737; GO:0006508; GO:0007155...   \n",
       "A0JNW5      GO:0005769; GO:0005829; GO:0034498; GO:0042803...   \n",
       "A0JP26                                                    NaN   \n",
       "A0PK11      GO:0007605; GO:0032421; GO:0060088; GO:0060171...   \n",
       "\n",
       "                                                  keyword_ids  \\\n",
       "Uniprot                                                         \n",
       "A0A0C5B5G6  KW-0238; KW-0496; KW-0539; KW-0892; KW-1185; K...   \n",
       "A0A1B0GTW7  KW-0025; KW-0225; KW-0325; KW-1056; KW-0378; K...   \n",
       "A0JNW5      KW-0025; KW-0175; KW-0963; KW-0967; KW-0597; K...   \n",
       "A0JP26            KW-0025; KW-0040; KW-0175; KW-1185; KW-0677   \n",
       "A0PK11      KW-1003; KW-0966; KW-0209; KW-0225; KW-0325; K...   \n",
       "\n",
       "                                                     keywords     tcdb_id  \\\n",
       "Uniprot                                                                     \n",
       "A0A0C5B5G6  DNA-binding;Mitochondrion;Nucleus;Osteogenesis...         NaN   \n",
       "A0A1B0GTW7  Alternative splicing;Disease variant;Glycoprot...         NaN   \n",
       "A0JNW5      Alternative splicing;Coiled coil;Cytoplasm;End...         NaN   \n",
       "A0JP26      Alternative splicing;ANK repeat;Coiled coil;Re...         NaN   \n",
       "A0PK11      Cell membrane;Cell projection;Deafness;Disease...  9.A.46.1.2   \n",
       "\n",
       "                                  interpro tcdb_class tcdb_substrates  \\\n",
       "Uniprot                                                                 \n",
       "A0A0C5B5G6                             NaN        0.0             NaN   \n",
       "A0A1B0GTW7                      IPR001577;        0.0             NaN   \n",
       "A0JNW5                IPR026728;IPR026854;        0.0             NaN   \n",
       "A0JP26      IPR002110;IPR036770;IPR039497;        0.0             NaN   \n",
       "A0PK11                          IPR026748;        9.A             NaN   \n",
       "\n",
       "           keywords_substrates keywords_transport_related keywords_location  \n",
       "Uniprot                                                                      \n",
       "A0A0C5B5G6                                                                   \n",
       "A0A1B0GTW7                                                                   \n",
       "A0JNW5                                                                       \n",
       "A0JP26                                                                       \n",
       "A0PK11                                                                       "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from subpred.dataset import create_dataset, SUBSTRATE_KEYWORDS\n",
    "\n",
    "df_uniprot = create_dataset(\n",
    "    input_file=\"../data/raw/uniprot/uniprot_2022_05_evidence1-2_nofragments.tsv\",\n",
    "    # keywords_classes = None,\n",
    "    # keywords_classes_all = SUBSTRATE_KEYWORDS,\n",
    "    # keywords_filter = None,\n",
    "    multi_substrate=\"keep\",\n",
    "    # outliers=outliers,\n",
    "    verbose=True,\n",
    "    # tax_ids_filter=[3702, 9606, 83333, 559292],\n",
    "    # sequence_clustering=70,\n",
    "    evidence_code=2,\n",
    "    invalid_amino_acids=\"remove_amino_acids\",\n",
    "    # gene_names_only = True,\n",
    "    # force_update=True,\n",
    "    # remove_sequence_fragments = True,\n",
    "    # force_update = False,\n",
    "    tcdb_substrates_file=\"../data/raw/tcdb/tcdb_substrates.tsv\",\n",
    "    swissprot_only=False,\n",
    ")\n",
    "print(df_uniprot.shape)\n",
    "df_uniprot.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(146597, 16)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_swissprot = df_uniprot[df_uniprot.reviewed == \"reviewed\"]\n",
    "df_swissprot.shape"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotation datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>keyword</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>DNA-binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>Mitochondrion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>Nucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>Osteogenesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>Reference proteome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Uniprot             keyword\n",
       "0  A0A0C5B5G6         DNA-binding\n",
       "1  A0A0C5B5G6       Mitochondrion\n",
       "2  A0A0C5B5G6             Nucleus\n",
       "3  A0A0C5B5G6        Osteogenesis\n",
       "4  A0A0C5B5G6  Reference proteome"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from subpred.dataset import get_go_df, get_keywords_df, get_tcdb_substrates\n",
    "\n",
    "df_kw_uniprot = get_keywords_df(df_uniprot)\n",
    "df_kw_uniprot.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>dataset</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>keyword</td>\n",
       "      <td>DNA-binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>keyword</td>\n",
       "      <td>Mitochondrion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>keyword</td>\n",
       "      <td>Nucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>keyword</td>\n",
       "      <td>Osteogenesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>keyword</td>\n",
       "      <td>Reference proteome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Uniprot  dataset          annotation\n",
       "0  A0A0C5B5G6  keyword         DNA-binding\n",
       "1  A0A0C5B5G6  keyword       Mitochondrion\n",
       "2  A0A0C5B5G6  keyword             Nucleus\n",
       "3  A0A0C5B5G6  keyword        Osteogenesis\n",
       "4  A0A0C5B5G6  keyword  Reference proteome"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kw_uniprot_long = df_kw_uniprot.assign(dataset=\"keyword\").rename(\n",
    "    columns={\"keyword\": \"annotation\"}\n",
    ")[[\"Uniprot\", \"dataset\", \"annotation\"]]\n",
    "df_kw_uniprot_long.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "from subpred.go_utils import GeneOntology\n",
    "\n",
    "go = GeneOntology(\"../data/raw/ontologies/go.owl\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>go_id</th>\n",
       "      <th>go_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>GO:0001649</td>\n",
       "      <td>osteoblast differentiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>GO:0003677</td>\n",
       "      <td>DNA binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>GO:0005615</td>\n",
       "      <td>extracellular space</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>GO:0005634</td>\n",
       "      <td>nucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>GO:0005739</td>\n",
       "      <td>mitochondrion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Uniprot       go_id                     go_term\n",
       "0  A0A0C5B5G6  GO:0001649  osteoblast differentiation\n",
       "1  A0A0C5B5G6  GO:0003677                 DNA binding\n",
       "2  A0A0C5B5G6  GO:0005615         extracellular space\n",
       "3  A0A0C5B5G6  GO:0005634                     nucleus\n",
       "4  A0A0C5B5G6  GO:0005739               mitochondrion"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go_uniprot = get_go_df(df_uniprot, go)\n",
    "df_go_uniprot.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>dataset</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4736560</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>go_term</td>\n",
       "      <td>osteoblast differentiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4736561</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>go_term</td>\n",
       "      <td>DNA binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4736562</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>go_term</td>\n",
       "      <td>extracellular space</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4736563</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>go_term</td>\n",
       "      <td>nucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4736564</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>go_term</td>\n",
       "      <td>mitochondrion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            Uniprot  dataset                  annotation\n",
       "4736560  A0A0C5B5G6  go_term  osteoblast differentiation\n",
       "4736561  A0A0C5B5G6  go_term                 DNA binding\n",
       "4736562  A0A0C5B5G6  go_term         extracellular space\n",
       "4736563  A0A0C5B5G6  go_term                     nucleus\n",
       "4736564  A0A0C5B5G6  go_term               mitochondrion"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go_uniprot_long = df_go_uniprot.melt(\n",
    "    id_vars=[\"Uniprot\"],\n",
    "    value_vars=[\"go_id\", \"go_term\"],\n",
    "    var_name=\"dataset\",\n",
    "    value_name=\"annotation\",\n",
    ")\n",
    "df_go_uniprot_long = df_go_uniprot_long[df_go_uniprot_long.dataset == \"go_term\"]\n",
    "df_go_uniprot_long.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>tcdb_id</th>\n",
       "      <th>chebi_id</th>\n",
       "      <th>chebi_term</th>\n",
       "      <th>tcdb_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A6NI61</td>\n",
       "      <td>1.N.2.1.1</td>\n",
       "      <td>CHEBI:25367</td>\n",
       "      <td>molecule</td>\n",
       "      <td>1.N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A6NKB5</td>\n",
       "      <td>9.A.80.1.1</td>\n",
       "      <td>CHEBI:24636</td>\n",
       "      <td>proton</td>\n",
       "      <td>9.A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O00168</td>\n",
       "      <td>1.A.27.1.8</td>\n",
       "      <td>CHEBI:22563</td>\n",
       "      <td>anion</td>\n",
       "      <td>1.A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O00337</td>\n",
       "      <td>2.A.41.2.3</td>\n",
       "      <td>CHEBI:9175</td>\n",
       "      <td>sodium(1+)</td>\n",
       "      <td>2.A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O00337</td>\n",
       "      <td>2.A.41.2.3</td>\n",
       "      <td>CHEBI:9893</td>\n",
       "      <td>uridine</td>\n",
       "      <td>2.A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Uniprot     tcdb_id     chebi_id  chebi_term tcdb_class\n",
       "0  A6NI61   1.N.2.1.1  CHEBI:25367    molecule        1.N\n",
       "1  A6NKB5  9.A.80.1.1  CHEBI:24636      proton        9.A\n",
       "2  O00168  1.A.27.1.8  CHEBI:22563       anion        1.A\n",
       "3  O00337  2.A.41.2.3   CHEBI:9175  sodium(1+)        2.A\n",
       "4  O00337  2.A.41.2.3   CHEBI:9893     uridine        2.A"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TODO TCDB classes\n",
    "df_tcdb_uniprot = get_tcdb_substrates(df_uniprot)\n",
    "df_tcdb_uniprot = df_tcdb_uniprot.assign(\n",
    "    tcdb_class=df_tcdb_uniprot.tcdb_id.apply(lambda x: x[:3])\n",
    ")\n",
    "df_tcdb_uniprot.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>dataset</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A6NI61</td>\n",
       "      <td>tcdb_id</td>\n",
       "      <td>1.N.2.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A6NKB5</td>\n",
       "      <td>tcdb_id</td>\n",
       "      <td>9.A.80.1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O00168</td>\n",
       "      <td>tcdb_id</td>\n",
       "      <td>1.A.27.1.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>O00337</td>\n",
       "      <td>tcdb_id</td>\n",
       "      <td>2.A.41.2.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O00337</td>\n",
       "      <td>tcdb_id</td>\n",
       "      <td>2.A.41.2.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Uniprot  dataset  annotation\n",
       "0  A6NI61  tcdb_id   1.N.2.1.1\n",
       "1  A6NKB5  tcdb_id  9.A.80.1.1\n",
       "2  O00168  tcdb_id  1.A.27.1.8\n",
       "3  O00337  tcdb_id  2.A.41.2.3\n",
       "4  O00337  tcdb_id  2.A.41.2.3"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcdb_uniprot_long = df_tcdb_uniprot.melt(\n",
    "    id_vars=[\"Uniprot\"],\n",
    "    value_vars=[\"tcdb_id\", \"chebi_id\", \"chebi_term\"],\n",
    "    var_name=\"dataset\",\n",
    "    value_name=\"annotation\",\n",
    ")\n",
    "df_tcdb_uniprot_long.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>interpro</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A1B0GTW7</td>\n",
       "      <td>IPR001577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0JNW5</td>\n",
       "      <td>IPR026728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0JNW5</td>\n",
       "      <td>IPR026854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0JP26</td>\n",
       "      <td>IPR002110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0JP26</td>\n",
       "      <td>IPR036770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4190605</th>\n",
       "      <td>X5MFI4</td>\n",
       "      <td>IPR029044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4190606</th>\n",
       "      <td>X5MI49</td>\n",
       "      <td>IPR008630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4190607</th>\n",
       "      <td>X5MI49</td>\n",
       "      <td>IPR029044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4190608</th>\n",
       "      <td>X5MPI5</td>\n",
       "      <td>IPR008630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4190609</th>\n",
       "      <td>X5MPI5</td>\n",
       "      <td>IPR029044</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4190610 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Uniprot   interpro\n",
       "0        A0A1B0GTW7  IPR001577\n",
       "1            A0JNW5  IPR026728\n",
       "2            A0JNW5  IPR026854\n",
       "3            A0JP26  IPR002110\n",
       "4            A0JP26  IPR036770\n",
       "...             ...        ...\n",
       "4190605      X5MFI4  IPR029044\n",
       "4190606      X5MI49  IPR008630\n",
       "4190607      X5MI49  IPR029044\n",
       "4190608      X5MPI5  IPR008630\n",
       "4190609      X5MPI5  IPR029044\n",
       "\n",
       "[4190610 rows x 2 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_uniprot_interpro = (\n",
    "    df_uniprot.interpro.dropna().str.rstrip(\";\").str.split(\";\").explode().str.strip()\n",
    ")\n",
    "assert df_uniprot_interpro[~df_uniprot_interpro.str.match(\"IPR[0-9]{6}\")].shape[0] == 0\n",
    "df_uniprot_interpro = df_uniprot_interpro.reset_index(\n",
    "    drop=False\n",
    ")  # [df_uniprot_interpro.isnull()]\n",
    "df_uniprot_interpro\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>dataset</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A1B0GTW7</td>\n",
       "      <td>interpro</td>\n",
       "      <td>IPR001577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0JNW5</td>\n",
       "      <td>interpro</td>\n",
       "      <td>IPR026728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0JNW5</td>\n",
       "      <td>interpro</td>\n",
       "      <td>IPR026854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0JP26</td>\n",
       "      <td>interpro</td>\n",
       "      <td>IPR002110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0JP26</td>\n",
       "      <td>interpro</td>\n",
       "      <td>IPR036770</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Uniprot   dataset annotation\n",
       "0  A0A1B0GTW7  interpro  IPR001577\n",
       "1      A0JNW5  interpro  IPR026728\n",
       "2      A0JNW5  interpro  IPR026854\n",
       "3      A0JP26  interpro  IPR002110\n",
       "4      A0JP26  interpro  IPR036770"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_uniprot_interpro_long = df_uniprot_interpro.assign(dataset=\"interpro\").rename(\n",
    "    columns={\"interpro\": \"annotation\"}\n",
    ")[[\"Uniprot\", \"dataset\", \"annotation\"]]\n",
    "df_uniprot_interpro_long.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_uniprot_organism_long = (\n",
    "    df_uniprot.organism_id.reset_index(drop=False)\n",
    "    .drop_duplicates()\n",
    "    .assign(dataset=\"organism_id\")\n",
    "    .rename(columns={\"organism_id\": \"annotation\"})[[\"Uniprot\", \"dataset\", \"annotation\"]]\n",
    ")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Dataframe containing all annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>dataset</th>\n",
       "      <th>annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>keyword</td>\n",
       "      <td>DNA-binding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>keyword</td>\n",
       "      <td>Mitochondrion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>keyword</td>\n",
       "      <td>Nucleus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>keyword</td>\n",
       "      <td>Osteogenesis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0C5B5G6</td>\n",
       "      <td>keyword</td>\n",
       "      <td>Reference proteome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Uniprot  dataset          annotation\n",
       "0  A0A0C5B5G6  keyword         DNA-binding\n",
       "1  A0A0C5B5G6  keyword       Mitochondrion\n",
       "2  A0A0C5B5G6  keyword             Nucleus\n",
       "3  A0A0C5B5G6  keyword        Osteogenesis\n",
       "4  A0A0C5B5G6  keyword  Reference proteome"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_annotations = pd.concat(\n",
    "    [\n",
    "        df_kw_uniprot_long,\n",
    "        df_go_uniprot_long,\n",
    "        df_tcdb_uniprot_long,\n",
    "        df_uniprot_interpro_long,\n",
    "        df_uniprot_organism_long,\n",
    "    ],\n",
    "    ignore_index=True,\n",
    ")\n",
    "\n",
    "df_annotations = df_annotations[~df_annotations.duplicated()]\n",
    "\n",
    "assert df_annotations[df_annotations.Uniprot.isnull()].shape[0] == 0\n",
    "assert df_annotations[df_annotations.dataset.isnull()].shape[0] == 0\n",
    "assert df_annotations[df_annotations.annotation.isnull()].shape[0] == 0\n",
    "\n",
    "df_annotations.head()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TODOs\n",
    "\n",
    "- Make AAC, PAAC faster\n",
    "- Should I only keep proteins for which all/most types of annotation are available?\n",
    "    - Most proteins do not have TCDB id, so that would not be feasible\n",
    "- Add non-Uniprot GO terms?\n",
    "    - Fewer terms, but we know the qualifiers and evidence codes\n",
    "- Encode sequences\n",
    "    - AAC, PAAC\n",
    "    - Seqvec\n",
    "    - Internal encoding of clustering program\n",
    "- Calculate clusters\n",
    "    - CD-Hit\n",
    "    - MMseqs2 (used by PDB) (TODO install in env, export new env)\n",
    "        - Also provides search functionality, for example for annotating organism. Can this be used to predict things with homology?\n",
    "    - Sklearn unsupervised learning\n",
    "- Analysis\n",
    "    - Enrichment analysis of clusters, sort\n",
    "        - GSEA etc.\n",
    "        - Hypergeometric test\n",
    "    - Rand index\n",
    "- Only use Protein level evidence proteins?\n",
    "- PCA\n",
    "    - https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3982804/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Evidence at transcript level    772752\n",
       "Evidence at protein level       249205\n",
       "Name: protein_existence, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_uniprot.protein_existence.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Trying PAAC + sklearn"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TODO implement, test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from subpred.compositions import calculate_aac, calculate_paac\n",
    "\n",
    "aac_new = calculate_aac(df_uniprot.sequence, n_threads=-1)\n",
    "paac_new = calculate_paac(df_uniprot.sequence, n_threads=-1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6.3 s ± 291 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n"
     ]
    }
   ],
   "source": [
    "%timeit aac_new = calculate_comp(df_uniprot.sequence,k=1, n_threads = -1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Trying mmseqs2:\n",
    "\n",
    "- Most important parameters: https://github.com/soedinglab/MMseqs2/wiki#clustering-criteria\n",
    "- Output format: Long TSV TODO write parser\n",
    "- Disadvantage: No pairwise scores, no PCA plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'reviewed'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 7\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39msubpred\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mfasta\u001b[39;00m \u001b[39mimport\u001b[39;00m write_fasta\n\u001b[1;32m      4\u001b[0m fasta_file_name \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39m/home/ad/tmp.fasta\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m      6\u001b[0m df_uniprot_test \u001b[39m=\u001b[39m df_uniprot[\n\u001b[0;32m----> 7\u001b[0m     (df_uniprot\u001b[39m.\u001b[39;49mreviewed \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mreviewed\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m      8\u001b[0m     \u001b[39m&\u001b[39m (df_uniprot\u001b[39m.\u001b[39mprotein_existence \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mEvidence at protein level\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m      9\u001b[0m ]\n\u001b[1;32m     10\u001b[0m df_uniprot_test \u001b[39m=\u001b[39m df_uniprot_test\u001b[39m.\u001b[39msequence\u001b[39m.\u001b[39mreset_index(drop\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m)\n\u001b[1;32m     11\u001b[0m df_uniprot_test\u001b[39m.\u001b[39mUniprot \u001b[39m=\u001b[39m df_uniprot_test\u001b[39m.\u001b[39mUniprot\u001b[39m.\u001b[39mapply(\u001b[39mlambda\u001b[39;00m x: \u001b[39m\"\u001b[39m\u001b[39m>\u001b[39m\u001b[39m\"\u001b[39m \u001b[39m+\u001b[39m x)\n",
      "File \u001b[0;32m~/mambaforge/envs/subpred4/lib/python3.10/site-packages/pandas/core/generic.py:5902\u001b[0m, in \u001b[0;36mNDFrame.__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5895\u001b[0m \u001b[39mif\u001b[39;00m (\n\u001b[1;32m   5896\u001b[0m     name \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_internal_names_set\n\u001b[1;32m   5897\u001b[0m     \u001b[39mand\u001b[39;00m name \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_metadata\n\u001b[1;32m   5898\u001b[0m     \u001b[39mand\u001b[39;00m name \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_accessors\n\u001b[1;32m   5899\u001b[0m     \u001b[39mand\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_info_axis\u001b[39m.\u001b[39m_can_hold_identifiers_and_holds_name(name)\n\u001b[1;32m   5900\u001b[0m ):\n\u001b[1;32m   5901\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m[name]\n\u001b[0;32m-> 5902\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mobject\u001b[39;49m\u001b[39m.\u001b[39;49m\u001b[39m__getattribute__\u001b[39;49m(\u001b[39mself\u001b[39;49m, name)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'reviewed'"
     ]
    }
   ],
   "source": [
    "# Fasta export TODO temp file\n",
    "from subpred.fasta import write_fasta\n",
    "\n",
    "fasta_file_name = \"/home/ad/tmp.fasta\"\n",
    "\n",
    "df_uniprot_test = df_uniprot[\n",
    "    (df_uniprot.reviewed == \"reviewed\")\n",
    "    & (df_uniprot.protein_existence == \"Evidence at protein level\")\n",
    "]\n",
    "df_uniprot_test = df_uniprot_test.sequence.reset_index(drop=False)\n",
    "df_uniprot_test.Uniprot = df_uniprot_test.Uniprot.apply(lambda x: \">\" + x)\n",
    "print(df_uniprot_test.shape)\n",
    "df_uniprot_test = df_uniprot_test[~df_uniprot_test.duplicated()]\n",
    "print(df_uniprot_test.shape)\n",
    "fasta_data = list(df_uniprot_test.to_records(index=False))\n",
    "\n",
    "write_fasta(fasta_file_name, fasta_data)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "mmseqs easy-cluster tmp.fasta tmp_out.fasta tmp\n",
    "\n",
    "\n",
    "--min-seq-id 0.5 -c 0.8 --cov-mode 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Trying Seqvec:\n",
    "\n",
    "hdf needs pytables depencency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Dataset(s) incompatible with Pandas data types, not table, or no datasets found in HDF5 file.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mpandas\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mpd\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m df_secvec \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39;49mread_hdf(\u001b[39m\"\u001b[39;49m\u001b[39m/home/ad/seqvec_test/uniref50_v2/weights.hdf5\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n",
      "File \u001b[0;32m~/mambaforge/envs/subpred4_dev/lib/python3.10/site-packages/pandas/io/pytables.py:425\u001b[0m, in \u001b[0;36mread_hdf\u001b[0;34m(path_or_buf, key, mode, errors, where, start, stop, columns, iterator, chunksize, **kwargs)\u001b[0m\n\u001b[1;32m    423\u001b[0m groups \u001b[39m=\u001b[39m store\u001b[39m.\u001b[39mgroups()\n\u001b[1;32m    424\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(groups) \u001b[39m==\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[0;32m--> 425\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[1;32m    426\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mDataset(s) incompatible with Pandas data types, \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    427\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mnot table, or no datasets found in HDF5 file.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    428\u001b[0m     )\n\u001b[1;32m    429\u001b[0m candidate_only_group \u001b[39m=\u001b[39m groups[\u001b[39m0\u001b[39m]\n\u001b[1;32m    431\u001b[0m \u001b[39m# For the HDF file to have only one dataset, all other groups\u001b[39;00m\n\u001b[1;32m    432\u001b[0m \u001b[39m# should then be metadata groups for that candidate group. (This\u001b[39;00m\n\u001b[1;32m    433\u001b[0m \u001b[39m# assumes that the groups() method enumerates parent groups\u001b[39;00m\n\u001b[1;32m    434\u001b[0m \u001b[39m# before their children.)\u001b[39;00m\n",
      "\u001b[0;31mValueError\u001b[0m: Dataset(s) incompatible with Pandas data types, not table, or no datasets found in HDF5 file."
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df_secvec = pd.read_hdf(\"/home/ad/seqvec_test/uniref50_v2/weights.hdf5\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'fasta'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mbiovec\u001b[39;00m\n",
      "File \u001b[0;32m~/mambaforge/envs/test/lib/python3.8/site-packages/biovec/__init__.py:1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mbiovec\u001b[39;00m \u001b[39mimport\u001b[39;00m models\n",
      "File \u001b[0;32m~/mambaforge/envs/test/lib/python3.8/site-packages/biovec/models/__init__.py:1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mbiovec\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mmodels\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mprot_vec\u001b[39;00m \u001b[39mimport\u001b[39;00m \u001b[39m*\u001b[39m\n",
      "File \u001b[0;32m~/mambaforge/envs/test/lib/python3.8/site-packages/biovec/models/prot_vec.py:2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mgensim\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mmodels\u001b[39;00m \u001b[39mimport\u001b[39;00m word2vec\n\u001b[0;32m----> 2\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mbiovec\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mutils\u001b[39;00m \u001b[39mimport\u001b[39;00m \u001b[39m*\u001b[39m\n\u001b[1;32m      5\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mload_protvec\u001b[39m(model_fname):\n\u001b[1;32m      6\u001b[0m     \u001b[39mreturn\u001b[39;00m word2vec\u001b[39m.\u001b[39mWord2Vec\u001b[39m.\u001b[39mload(model_fname)\n",
      "File \u001b[0;32m~/mambaforge/envs/test/lib/python3.8/site-packages/biovec/utils.py:1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mpyfasta\u001b[39;00m \u001b[39mimport\u001b[39;00m Fasta\n\u001b[1;32m      2\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mtqdm\u001b[39;00m \u001b[39mimport\u001b[39;00m tqdm\n\u001b[1;32m      5\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39msplit_ngrams\u001b[39m(seq, n):\n",
      "File \u001b[0;32m~/mambaforge/envs/test/lib/python3.8/site-packages/pyfasta/__init__.py:3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m__future__\u001b[39;00m \u001b[39mimport\u001b[39;00m print_function\n\u001b[1;32m      2\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39msys\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mfasta\u001b[39;00m \u001b[39mimport\u001b[39;00m Fasta, complement, DuplicateHeaderException\n\u001b[1;32m      4\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mrecords\u001b[39;00m \u001b[39mimport\u001b[39;00m \u001b[39m*\u001b[39m\n\u001b[1;32m      5\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39msplit_fasta\u001b[39;00m \u001b[39mimport\u001b[39;00m split\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'fasta'"
     ]
    }
   ],
   "source": [
    "import biovec\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 40 from PyObject",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[3], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mallennlp\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mcommands\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39melmo\u001b[39;00m \u001b[39mimport\u001b[39;00m ElmoEmbedder\n\u001b[1;32m      2\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mpathlib\u001b[39;00m \u001b[39mimport\u001b[39;00m Path\n\u001b[1;32m      4\u001b[0m model_dir \u001b[39m=\u001b[39m Path(\u001b[39m'\u001b[39m\u001b[39m/home/ad/seqvec_test/uniref50_v2\u001b[39m\u001b[39m'\u001b[39m)\n",
      "File \u001b[0;32m~/mambaforge/envs/test/lib/python3.8/site-packages/allennlp/commands/__init__.py:8\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39moverrides\u001b[39;00m \u001b[39mimport\u001b[39;00m overrides\n\u001b[1;32m      7\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mallennlp\u001b[39;00m \u001b[39mimport\u001b[39;00m __version__\n\u001b[0;32m----> 8\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mallennlp\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mcommands\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mconfigure\u001b[39;00m \u001b[39mimport\u001b[39;00m Configure\n\u001b[1;32m      9\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mallennlp\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mcommands\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39melmo\u001b[39;00m \u001b[39mimport\u001b[39;00m Elmo\n\u001b[1;32m     10\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mallennlp\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mcommands\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mevaluate\u001b[39;00m \u001b[39mimport\u001b[39;00m Evaluate\n",
      "File \u001b[0;32m~/mambaforge/envs/test/lib/python3.8/site-packages/allennlp/commands/configure.py:23\u001b[0m\n\u001b[1;32m     20\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39margparse\u001b[39;00m\n\u001b[1;32m     22\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mflask_cors\u001b[39;00m \u001b[39mimport\u001b[39;00m CORS\n\u001b[0;32m---> 23\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mgevent\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mpywsgi\u001b[39;00m \u001b[39mimport\u001b[39;00m WSGIServer\n\u001b[1;32m     25\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mallennlp\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mcommands\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39msubcommand\u001b[39;00m \u001b[39mimport\u001b[39;00m Subcommand\n\u001b[1;32m     26\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mallennlp\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mservice\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mconfig_explorer\u001b[39;00m \u001b[39mimport\u001b[39;00m make_app\n",
      "File \u001b[0;32m~/mambaforge/envs/test/lib/python3.8/site-packages/gevent/__init__.py:87\u001b[0m\n\u001b[1;32m     84\u001b[0m             _switchinterval \u001b[39m=\u001b[39m interval\n\u001b[1;32m     86\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mgevent\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_config\u001b[39;00m \u001b[39mimport\u001b[39;00m config\n\u001b[0;32m---> 87\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mgevent\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_hub_local\u001b[39;00m \u001b[39mimport\u001b[39;00m get_hub\n\u001b[1;32m     88\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mgevent\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_hub_primitives\u001b[39;00m \u001b[39mimport\u001b[39;00m iwait_on_objects \u001b[39mas\u001b[39;00m iwait\n\u001b[1;32m     89\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mgevent\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_hub_primitives\u001b[39;00m \u001b[39mimport\u001b[39;00m wait_on_objects \u001b[39mas\u001b[39;00m wait\n",
      "File \u001b[0;32m~/mambaforge/envs/test/lib/python3.8/site-packages/gevent/_hub_local.py:101\u001b[0m\n\u001b[1;32m     98\u001b[0m     _threadlocal\u001b[39m.\u001b[39mloop \u001b[39m=\u001b[39m loop\n\u001b[1;32m    100\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mgevent\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_util\u001b[39;00m \u001b[39mimport\u001b[39;00m import_c_accel\n\u001b[0;32m--> 101\u001b[0m import_c_accel(\u001b[39mglobals\u001b[39;49m(), \u001b[39m'\u001b[39;49m\u001b[39mgevent.__hub_local\u001b[39;49m\u001b[39m'\u001b[39;49m)\n",
      "File \u001b[0;32m~/mambaforge/envs/test/lib/python3.8/site-packages/gevent/_util.py:105\u001b[0m, in \u001b[0;36mimport_c_accel\u001b[0;34m(globs, cname)\u001b[0m\n\u001b[1;32m     96\u001b[0m \u001b[39mwith\u001b[39;00m warnings\u001b[39m.\u001b[39mcatch_warnings():\n\u001b[1;32m     97\u001b[0m     \u001b[39m# Python 3.7 likes to produce\u001b[39;00m\n\u001b[1;32m     98\u001b[0m     \u001b[39m# \"ImportWarning: can't resolve\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    102\u001b[0m     \u001b[39m# Cython, but it doesn't seem to have any consequences, it's\u001b[39;00m\n\u001b[1;32m    103\u001b[0m     \u001b[39m# just annoying to see and can mess up our unittests.\u001b[39;00m\n\u001b[1;32m    104\u001b[0m     warnings\u001b[39m.\u001b[39msimplefilter(\u001b[39m'\u001b[39m\u001b[39mignore\u001b[39m\u001b[39m'\u001b[39m, \u001b[39mImportWarning\u001b[39;00m)\n\u001b[0;32m--> 105\u001b[0m     mod \u001b[39m=\u001b[39m importlib\u001b[39m.\u001b[39;49mimport_module(cname)\n\u001b[1;32m    107\u001b[0m \u001b[39m# By adopting the entire __dict__, we get a more accurate\u001b[39;00m\n\u001b[1;32m    108\u001b[0m \u001b[39m# __file__ and module repr, plus we don't leak any imported\u001b[39;00m\n\u001b[1;32m    109\u001b[0m \u001b[39m# things we no longer need.\u001b[39;00m\n\u001b[1;32m    110\u001b[0m globs\u001b[39m.\u001b[39mclear()\n",
      "File \u001b[0;32m~/mambaforge/envs/test/lib/python3.8/importlib/__init__.py:127\u001b[0m, in \u001b[0;36mimport_module\u001b[0;34m(name, package)\u001b[0m\n\u001b[1;32m    125\u001b[0m             \u001b[39mbreak\u001b[39;00m\n\u001b[1;32m    126\u001b[0m         level \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m \u001b[39m1\u001b[39m\n\u001b[0;32m--> 127\u001b[0m \u001b[39mreturn\u001b[39;00m _bootstrap\u001b[39m.\u001b[39;49m_gcd_import(name[level:], package, level)\n",
      "File \u001b[0;32msrc/gevent/__greenlet_primitives.pxd:12\u001b[0m, in \u001b[0;36minit gevent.__hub_local\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: greenlet.greenlet size changed, may indicate binary incompatibility. Expected 144 from C header, got 40 from PyObject"
     ]
    }
   ],
   "source": [
    "from allennlp.commands.elmo import ElmoEmbedder\n",
    "from pathlib import Path\n",
    "\n",
    "model_dir = Path(\"/home/ad/seqvec_test/uniref50_v2\")\n",
    "weights = model_dir / \"weights.hdf5\"\n",
    "options = model_dir / \"options.json\"\n",
    "embedder = ElmoEmbedder(options, weights, cuda_device=0)  # cuda_device=-1 for CPU\n",
    "\n",
    "seq = \"SEQWENCE\"  # your amino acid sequence\n",
    "embedding = embedder.embed_sentence(list(seq))  # List-of-Lists with shape [3,L,1024]\n",
    "\n",
    "\n",
    "seq1 = \"SEQWENCE\"  # your amino acid sequence\n",
    "seq2 = \"PROTEIN\"\n",
    "seqs = [list(seq1), list(seq2)]\n",
    "seqs.sort(key=len)  # sorting is crucial for speed\n",
    "embedding = embedder.embed_sentences(\n",
    "    seqs\n",
    ")  # returns: List-of-Lists with shape [3,L,1024]\n",
    "\n",
    "\n",
    "import torch\n",
    "\n",
    "residue_embd = torch.tensor(embedding).sum(dim=0)  # Tensor with shape [L,1024]\n",
    "protein_embd = (\n",
    "    torch.tensor(embedding).sum(dim=0).mean(dim=0)\n",
    ")  # Vector with shape [1024]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "subpred4",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "ec8de6dd8f2d395005c42468f5d13cdd089376f2ad37eab4b2481f4d8703f0b2"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
