{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from subpred.util import load_df\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO Different ways to filter for passive transporters, different definitions\n",
    "# TODO What to do about multiple substrates? Create stats on them\n",
    "# TODO Plots: Dim reduction, Hclust, Kmeans\n",
    "# TODO stats on 2.B, 2.C, 2.D\n",
    "# TODO possible to make bacteria-only dataset?\n",
    "# TODO sequence clustering\n",
    "# TODO what to do about complexes?\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When clustering the entire dataset of *E. coli* transporters in the previous notebook, we discovered that clusters form based on transport mechanism or protein family, not on substrates, which is what we wanted. The idea now is to only look at one type of transmembrane transporter, to find out if we can divide that set into substrates. We already had the suspicion in Manuscript 1 that the *A. thaliana* model worked to well because virtually all of the transporters in our dataset were secondary active. \n",
    "\n",
    "In this notebook, we will try creating datasets that only contain secondary active transporters, which typically transport small hydrophilic molecules across membranes. This excludes channels (passive transporters), binding proteins from transport complexes such as ABC transporters, primary active transporters, and others. We will look at those in other notebooks.\n",
    "\n",
    "Secondary active transporters are gradient-driven transporters, typically with alpha-helical structures. The biggest category should be the Porters, in the form of Uniporters, Symporters and Antiporters. The latter two come with the challenge that they fall into two or more substrate classes. We should create different categories for the individual combinations, and see if they make good substrate classes. The mechanism usually works by a conformational change in the protein, through the presence of a substrate.\n",
    "\n",
    "We will also explore four different ways of creating the passive transporter dataset: TCDB, Gene Ontology, Keywords and Interpro domains. The respective datasets will be compared in terms of size and overlap. TCDB might have the most biologically accurate data, but is not available for most organisms. *E. coli* is a rare example where most of the transporters have an entry. "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Protein dataset"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Loading the Uniprot data for all organisms:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_names</th>\n",
       "      <th>protein_names</th>\n",
       "      <th>reviewed</th>\n",
       "      <th>protein_existence</th>\n",
       "      <th>sequence</th>\n",
       "      <th>organism_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniprot</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A0A0C5B5G6</th>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>Mitochondrial-derived peptide MOTS-c (Mitochon...</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MRWQEMGYIFYPRKLR</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A1B0GTW7</th>\n",
       "      <td>CIROP LMLN2</td>\n",
       "      <td>Ciliated left-right organizer metallopeptidase...</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MLLLLLLLLLLPPLVLRVAASRCLHDETQKSVSLLRPPFSQLPSKS...</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0JNW5</th>\n",
       "      <td>BLTP3B KIAA0701 SHIP164 UHRF1BP1L</td>\n",
       "      <td>Bridge-like lipid transfer protein family memb...</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MAGIIKKQILKHLSRFTKNLSPDKINLSTLKGEGELKNLELDEEVL...</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0JP26</th>\n",
       "      <td>POTEB3</td>\n",
       "      <td>POTE ankyrin domain family member B3</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MVAEVCSMPAASAVKKPFDLRSKMGKWCHHRFPCCRGSGKSNMGTS...</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0PK11</th>\n",
       "      <td>CLRN2</td>\n",
       "      <td>Clarin-2</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MPGWFKKAWYGLASLLSFSSFILIIVALVVPHWLSGKILCQTGVDL...</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X5L4R4</th>\n",
       "      <td>NOD-2</td>\n",
       "      <td>Nucleotide-binding oligomerization domain-cont...</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>MSPGCYKGWPFNCHLSHEEDKRRNETLLQEAETSNLQITASFVSGL...</td>\n",
       "      <td>586796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X5MBL2</th>\n",
       "      <td>GT34D</td>\n",
       "      <td>Putative galacto(Gluco)mannan alpha-1,6-galact...</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>KVLYDRAFNSSDDQSALVYLLLKEKDKWADRIFIEHKYYLNGYWLD...</td>\n",
       "      <td>3352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X5MFI4</th>\n",
       "      <td>GT34D</td>\n",
       "      <td>Putative galacto(Gluco)mannan alpha-1,6-galact...</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>MDEDVLCKGPLHGGSARSLKGSLKRLKRIMESLNDGLIFMGGAVSA...</td>\n",
       "      <td>3352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X5MI49</th>\n",
       "      <td>GT34A</td>\n",
       "      <td>Putative galacto(Gluco)mannan alpha-1,6-galact...</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>MVNDSKLETISGNMVQKRKSFDGLPFWTVSIAGGLLLCWSLWRICF...</td>\n",
       "      <td>3352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X5MPI5</th>\n",
       "      <td>GT34C</td>\n",
       "      <td>Xyloglucan alpha-1,6-xylosyltransferase</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>MRESVKRGLPSVRGPANGRSLALPKARSIQKTFNNLKITILCGFVT...</td>\n",
       "      <td>3352</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1021957 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   gene_names  \\\n",
       "Uniprot                                         \n",
       "A0A0C5B5G6                            MT-RNR1   \n",
       "A0A1B0GTW7                        CIROP LMLN2   \n",
       "A0JNW5      BLTP3B KIAA0701 SHIP164 UHRF1BP1L   \n",
       "A0JP26                                 POTEB3   \n",
       "A0PK11                                  CLRN2   \n",
       "...                                       ...   \n",
       "X5L4R4                                  NOD-2   \n",
       "X5MBL2                                  GT34D   \n",
       "X5MFI4                                  GT34D   \n",
       "X5MI49                                  GT34A   \n",
       "X5MPI5                                  GT34C   \n",
       "\n",
       "                                                protein_names  reviewed  \\\n",
       "Uniprot                                                                   \n",
       "A0A0C5B5G6  Mitochondrial-derived peptide MOTS-c (Mitochon...      True   \n",
       "A0A1B0GTW7  Ciliated left-right organizer metallopeptidase...      True   \n",
       "A0JNW5      Bridge-like lipid transfer protein family memb...      True   \n",
       "A0JP26                   POTE ankyrin domain family member B3      True   \n",
       "A0PK11                                               Clarin-2      True   \n",
       "...                                                       ...       ...   \n",
       "X5L4R4      Nucleotide-binding oligomerization domain-cont...     False   \n",
       "X5MBL2      Putative galacto(Gluco)mannan alpha-1,6-galact...     False   \n",
       "X5MFI4      Putative galacto(Gluco)mannan alpha-1,6-galact...     False   \n",
       "X5MI49      Putative galacto(Gluco)mannan alpha-1,6-galact...     False   \n",
       "X5MPI5                Xyloglucan alpha-1,6-xylosyltransferase     False   \n",
       "\n",
       "            protein_existence  \\\n",
       "Uniprot                         \n",
       "A0A0C5B5G6                  1   \n",
       "A0A1B0GTW7                  1   \n",
       "A0JNW5                      1   \n",
       "A0JP26                      1   \n",
       "A0PK11                      1   \n",
       "...                       ...   \n",
       "X5L4R4                      2   \n",
       "X5MBL2                      2   \n",
       "X5MFI4                      2   \n",
       "X5MI49                      2   \n",
       "X5MPI5                      2   \n",
       "\n",
       "                                                     sequence  organism_id  \n",
       "Uniprot                                                                     \n",
       "A0A0C5B5G6                                   MRWQEMGYIFYPRKLR         9606  \n",
       "A0A1B0GTW7  MLLLLLLLLLLPPLVLRVAASRCLHDETQKSVSLLRPPFSQLPSKS...         9606  \n",
       "A0JNW5      MAGIIKKQILKHLSRFTKNLSPDKINLSTLKGEGELKNLELDEEVL...         9606  \n",
       "A0JP26      MVAEVCSMPAASAVKKPFDLRSKMGKWCHHRFPCCRGSGKSNMGTS...         9606  \n",
       "A0PK11      MPGWFKKAWYGLASLLSFSSFILIIVALVVPHWLSGKILCQTGVDL...         9606  \n",
       "...                                                       ...          ...  \n",
       "X5L4R4      MSPGCYKGWPFNCHLSHEEDKRRNETLLQEAETSNLQITASFVSGL...       586796  \n",
       "X5MBL2      KVLYDRAFNSSDDQSALVYLLLKEKDKWADRIFIEHKYYLNGYWLD...         3352  \n",
       "X5MFI4      MDEDVLCKGPLHGGSARSLKGSLKRLKRIMESLNDGLIFMGGAVSA...         3352  \n",
       "X5MI49      MVNDSKLETISGNMVQKRKSFDGLPFWTVSIAGGLLLCWSLWRICF...         3352  \n",
       "X5MPI5      MRESVKRGLPSVRGPANGRSLALPKARSIQKTFNNLKITILCGFVT...         3352  \n",
       "\n",
       "[1021957 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_uniprot = load_df(\"uniprot\")\n",
    "df_uniprot\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filtering for *E coli K12*, since that strain has the highest number of functional annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_names</th>\n",
       "      <th>protein_names</th>\n",
       "      <th>reviewed</th>\n",
       "      <th>protein_existence</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniprot</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P00509</th>\n",
       "      <td>aspC b0928 JW0911</td>\n",
       "      <td>Aspartate aminotransferase (AspAT) (EC 2.6.1.1...</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MFENITAAPADPILGLADLFRADERPGKINLGIGVYKDETGKTPVL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00803</th>\n",
       "      <td>lepB b2568 JW2552</td>\n",
       "      <td>Signal peptidase I (SPase I) (EC 3.4.21.89) (L...</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MANMFALILVIATLVTGILWCVDKFFFAPKRRERQAAAQAAAGDSL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00804</th>\n",
       "      <td>lspA lsp b0027 JW0025</td>\n",
       "      <td>Lipoprotein signal peptidase (EC 3.4.23.36) (P...</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MSQSICSTGLRWLWLVVVVLIIDLGSKYLILQNFALGDTVPLFPSL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00861</th>\n",
       "      <td>lysA b2838 JW2806</td>\n",
       "      <td>Diaminopimelate decarboxylase (DAP decarboxyla...</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MPHSLFSTDTDLTAENLLRLPAEFGCPVWVYDAQIIRRQIAALKQF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00946</th>\n",
       "      <td>manA pmi b1613 JW1605</td>\n",
       "      <td>Mannose-6-phosphate isomerase (EC 5.3.1.8) (Ph...</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>MQKLINSVQNYAWGSKTALTELYGMENPSSQPMAELWMGAHPKSSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P76154</th>\n",
       "      <td>ydfK b1544 JW1537</td>\n",
       "      <td>Cold shock protein YdfK</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>MKSKDTLKWFPAQLPEVRIILGDAVVEVAKQGRPINTRTLLDYIEG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P0AEG8</th>\n",
       "      <td>dsrB b1952 JW1936</td>\n",
       "      <td>Protein DsrB</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>MKVNDRVTVKTDGGPRRPGVVLAVEEFSEGTMYLVSLEDYPLGIWF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P33668</th>\n",
       "      <td>ybbC b0498 JW0487</td>\n",
       "      <td>Uncharacterized protein YbbC</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>MKYSSIFSMLSFFILFACNETAVYGSDENIIFMRYVEKLHLDKYSV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A7H2C7B0</th>\n",
       "      <td>speFL ECK4660 b4803</td>\n",
       "      <td>Leader peptide SpeFL (Arrest peptide SpeFL)</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>MENNSRTMPHIRRTTHIMKFAHRNSFDFHFFNAR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A0A6YVN8</th>\n",
       "      <td>D-tagatose 3-epimerase</td>\n",
       "      <td>D-tagatose 3-epimerase</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>MNKVGMFYTYWSTEWMVDFPATAKRIAGLGFDLMEISLGEFHNLSD...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3284 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_names  \\\n",
       "Uniprot                              \n",
       "P00509           aspC b0928 JW0911   \n",
       "P00803           lepB b2568 JW2552   \n",
       "P00804       lspA lsp b0027 JW0025   \n",
       "P00861           lysA b2838 JW2806   \n",
       "P00946       manA pmi b1613 JW1605   \n",
       "...                            ...   \n",
       "P76154           ydfK b1544 JW1537   \n",
       "P0AEG8           dsrB b1952 JW1936   \n",
       "P33668           ybbC b0498 JW0487   \n",
       "A0A7H2C7B0     speFL ECK4660 b4803   \n",
       "A0A0A6YVN8  D-tagatose 3-epimerase   \n",
       "\n",
       "                                                protein_names  reviewed  \\\n",
       "Uniprot                                                                   \n",
       "P00509      Aspartate aminotransferase (AspAT) (EC 2.6.1.1...      True   \n",
       "P00803      Signal peptidase I (SPase I) (EC 3.4.21.89) (L...      True   \n",
       "P00804      Lipoprotein signal peptidase (EC 3.4.23.36) (P...      True   \n",
       "P00861      Diaminopimelate decarboxylase (DAP decarboxyla...      True   \n",
       "P00946      Mannose-6-phosphate isomerase (EC 5.3.1.8) (Ph...      True   \n",
       "...                                                       ...       ...   \n",
       "P76154                                Cold shock protein YdfK      True   \n",
       "P0AEG8                                           Protein DsrB      True   \n",
       "P33668                           Uncharacterized protein YbbC      True   \n",
       "A0A7H2C7B0        Leader peptide SpeFL (Arrest peptide SpeFL)     False   \n",
       "A0A0A6YVN8                             D-tagatose 3-epimerase     False   \n",
       "\n",
       "            protein_existence  \\\n",
       "Uniprot                         \n",
       "P00509                      1   \n",
       "P00803                      1   \n",
       "P00804                      1   \n",
       "P00861                      1   \n",
       "P00946                      1   \n",
       "...                       ...   \n",
       "P76154                      2   \n",
       "P0AEG8                      2   \n",
       "P33668                      2   \n",
       "A0A7H2C7B0                  2   \n",
       "A0A0A6YVN8                  1   \n",
       "\n",
       "                                                     sequence  \n",
       "Uniprot                                                        \n",
       "P00509      MFENITAAPADPILGLADLFRADERPGKINLGIGVYKDETGKTPVL...  \n",
       "P00803      MANMFALILVIATLVTGILWCVDKFFFAPKRRERQAAAQAAAGDSL...  \n",
       "P00804      MSQSICSTGLRWLWLVVVVLIIDLGSKYLILQNFALGDTVPLFPSL...  \n",
       "P00861      MPHSLFSTDTDLTAENLLRLPAEFGCPVWVYDAQIIRRQIAALKQF...  \n",
       "P00946      MQKLINSVQNYAWGSKTALTELYGMENPSSQPMAELWMGAHPKSSS...  \n",
       "...                                                       ...  \n",
       "P76154      MKSKDTLKWFPAQLPEVRIILGDAVVEVAKQGRPINTRTLLDYIEG...  \n",
       "P0AEG8      MKVNDRVTVKTDGGPRRPGVVLAVEEFSEGTMYLVSLEDYPLGIWF...  \n",
       "P33668      MKYSSIFSMLSFFILFACNETAVYGSDENIIFMRYVEKLHLDKYSV...  \n",
       "A0A7H2C7B0                 MENNSRTMPHIRRTTHIMKFAHRNSFDFHFFNAR  \n",
       "A0A0A6YVN8  MNKVGMFYTYWSTEWMVDFPATAKRIAGLGFDLMEISLGEFHNLSD...  \n",
       "\n",
       "[3284 rows x 5 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_uniprot_ecoli = df_uniprot[df_uniprot.organism_id == 83333].drop(\n",
    "    \"organism_id\", axis=1\n",
    ")\n",
    "df_uniprot_ecoli\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What is the distribution of evidence codes among Swissprot and TrEMBL?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reviewed</th>\n",
       "      <th>protein_existence</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>3118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   reviewed  protein_existence  count\n",
       "0     False                  1      1\n",
       "1     False                  2      1\n",
       "2      True                  1   3118\n",
       "3      True                  2    164"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_uniprot_ecoli.groupby([\"reviewed\", \"protein_existence\"]).size().reset_index().rename(\n",
    "    columns={0: \"count\"}\n",
    ")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This organism seems to be researched very well. TrEMBL only adds two additional samples to the dataset. Which ones are those?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_names</th>\n",
       "      <th>protein_names</th>\n",
       "      <th>reviewed</th>\n",
       "      <th>protein_existence</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniprot</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A0A7H2C7B0</th>\n",
       "      <td>speFL ECK4660 b4803</td>\n",
       "      <td>Leader peptide SpeFL (Arrest peptide SpeFL)</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "      <td>MENNSRTMPHIRRTTHIMKFAHRNSFDFHFFNAR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A0A0A6YVN8</th>\n",
       "      <td>D-tagatose 3-epimerase</td>\n",
       "      <td>D-tagatose 3-epimerase</td>\n",
       "      <td>False</td>\n",
       "      <td>1</td>\n",
       "      <td>MNKVGMFYTYWSTEWMVDFPATAKRIAGLGFDLMEISLGEFHNLSD...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_names  \\\n",
       "Uniprot                              \n",
       "A0A7H2C7B0     speFL ECK4660 b4803   \n",
       "A0A0A6YVN8  D-tagatose 3-epimerase   \n",
       "\n",
       "                                          protein_names  reviewed  \\\n",
       "Uniprot                                                             \n",
       "A0A7H2C7B0  Leader peptide SpeFL (Arrest peptide SpeFL)     False   \n",
       "A0A0A6YVN8                       D-tagatose 3-epimerase     False   \n",
       "\n",
       "            protein_existence  \\\n",
       "Uniprot                         \n",
       "A0A7H2C7B0                  2   \n",
       "A0A0A6YVN8                  1   \n",
       "\n",
       "                                                     sequence  \n",
       "Uniprot                                                        \n",
       "A0A7H2C7B0                 MENNSRTMPHIRRTTHIMKFAHRNSFDFHFFNAR  \n",
       "A0A0A6YVN8  MNKVGMFYTYWSTEWMVDFPATAKRIAGLGFDLMEISLGEFHNLSD...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_uniprot_ecoli[~df_uniprot_ecoli.reviewed]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A metal ion binding protein with otherwise unknown function or location, and an expression factor. Nothing related to transport, we can remove them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_names</th>\n",
       "      <th>protein_names</th>\n",
       "      <th>protein_existence</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniprot</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P00509</th>\n",
       "      <td>aspC b0928 JW0911</td>\n",
       "      <td>Aspartate aminotransferase (AspAT) (EC 2.6.1.1...</td>\n",
       "      <td>1</td>\n",
       "      <td>MFENITAAPADPILGLADLFRADERPGKINLGIGVYKDETGKTPVL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00803</th>\n",
       "      <td>lepB b2568 JW2552</td>\n",
       "      <td>Signal peptidase I (SPase I) (EC 3.4.21.89) (L...</td>\n",
       "      <td>1</td>\n",
       "      <td>MANMFALILVIATLVTGILWCVDKFFFAPKRRERQAAAQAAAGDSL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00804</th>\n",
       "      <td>lspA lsp b0027 JW0025</td>\n",
       "      <td>Lipoprotein signal peptidase (EC 3.4.23.36) (P...</td>\n",
       "      <td>1</td>\n",
       "      <td>MSQSICSTGLRWLWLVVVVLIIDLGSKYLILQNFALGDTVPLFPSL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00861</th>\n",
       "      <td>lysA b2838 JW2806</td>\n",
       "      <td>Diaminopimelate decarboxylase (DAP decarboxyla...</td>\n",
       "      <td>1</td>\n",
       "      <td>MPHSLFSTDTDLTAENLLRLPAEFGCPVWVYDAQIIRRQIAALKQF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P00946</th>\n",
       "      <td>manA pmi b1613 JW1605</td>\n",
       "      <td>Mannose-6-phosphate isomerase (EC 5.3.1.8) (Ph...</td>\n",
       "      <td>1</td>\n",
       "      <td>MQKLINSVQNYAWGSKTALTELYGMENPSSQPMAELWMGAHPKSSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P77564</th>\n",
       "      <td>ydhW b1672 JW1662</td>\n",
       "      <td>Uncharacterized protein YdhW</td>\n",
       "      <td>2</td>\n",
       "      <td>MGKMNHQDELPLAKVSEVDEAKRQWLQGMRHPVDTVTEPEPAEILA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P76157</th>\n",
       "      <td>ynfN b1551 JW5254</td>\n",
       "      <td>Uncharacterized protein YnfN</td>\n",
       "      <td>2</td>\n",
       "      <td>MREYPNGEKTHLTVMAAGFPSLTGDHKVIYVAADRHVTSEEILEAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P76154</th>\n",
       "      <td>ydfK b1544 JW1537</td>\n",
       "      <td>Cold shock protein YdfK</td>\n",
       "      <td>2</td>\n",
       "      <td>MKSKDTLKWFPAQLPEVRIILGDAVVEVAKQGRPINTRTLLDYIEG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P0AEG8</th>\n",
       "      <td>dsrB b1952 JW1936</td>\n",
       "      <td>Protein DsrB</td>\n",
       "      <td>2</td>\n",
       "      <td>MKVNDRVTVKTDGGPRRPGVVLAVEEFSEGTMYLVSLEDYPLGIWF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P33668</th>\n",
       "      <td>ybbC b0498 JW0487</td>\n",
       "      <td>Uncharacterized protein YbbC</td>\n",
       "      <td>2</td>\n",
       "      <td>MKYSSIFSMLSFFILFACNETAVYGSDENIIFMRYVEKLHLDKYSV...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3282 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_names  \\\n",
       "Uniprot                          \n",
       "P00509       aspC b0928 JW0911   \n",
       "P00803       lepB b2568 JW2552   \n",
       "P00804   lspA lsp b0027 JW0025   \n",
       "P00861       lysA b2838 JW2806   \n",
       "P00946   manA pmi b1613 JW1605   \n",
       "...                        ...   \n",
       "P77564       ydhW b1672 JW1662   \n",
       "P76157       ynfN b1551 JW5254   \n",
       "P76154       ydfK b1544 JW1537   \n",
       "P0AEG8       dsrB b1952 JW1936   \n",
       "P33668       ybbC b0498 JW0487   \n",
       "\n",
       "                                             protein_names  protein_existence  \\\n",
       "Uniprot                                                                         \n",
       "P00509   Aspartate aminotransferase (AspAT) (EC 2.6.1.1...                  1   \n",
       "P00803   Signal peptidase I (SPase I) (EC 3.4.21.89) (L...                  1   \n",
       "P00804   Lipoprotein signal peptidase (EC 3.4.23.36) (P...                  1   \n",
       "P00861   Diaminopimelate decarboxylase (DAP decarboxyla...                  1   \n",
       "P00946   Mannose-6-phosphate isomerase (EC 5.3.1.8) (Ph...                  1   \n",
       "...                                                    ...                ...   \n",
       "P77564                        Uncharacterized protein YdhW                  2   \n",
       "P76157                        Uncharacterized protein YnfN                  2   \n",
       "P76154                             Cold shock protein YdfK                  2   \n",
       "P0AEG8                                        Protein DsrB                  2   \n",
       "P33668                        Uncharacterized protein YbbC                  2   \n",
       "\n",
       "                                                  sequence  \n",
       "Uniprot                                                     \n",
       "P00509   MFENITAAPADPILGLADLFRADERPGKINLGIGVYKDETGKTPVL...  \n",
       "P00803   MANMFALILVIATLVTGILWCVDKFFFAPKRRERQAAAQAAAGDSL...  \n",
       "P00804   MSQSICSTGLRWLWLVVVVLIIDLGSKYLILQNFALGDTVPLFPSL...  \n",
       "P00861   MPHSLFSTDTDLTAENLLRLPAEFGCPVWVYDAQIIRRQIAALKQF...  \n",
       "P00946   MQKLINSVQNYAWGSKTALTELYGMENPSSQPMAELWMGAHPKSSS...  \n",
       "...                                                    ...  \n",
       "P77564   MGKMNHQDELPLAKVSEVDEAKRQWLQGMRHPVDTVTEPEPAEILA...  \n",
       "P76157   MREYPNGEKTHLTVMAAGFPSLTGDHKVIYVAADRHVTSEEILEAA...  \n",
       "P76154   MKSKDTLKWFPAQLPEVRIILGDAVVEVAKQGRPINTRTLLDYIEG...  \n",
       "P0AEG8   MKVNDRVTVKTDGGPRRPGVVLAVEEFSEGTMYLVSLEDYPLGIWF...  \n",
       "P33668   MKYSSIFSMLSFFILFACNETAVYGSDENIIFMRYVEKLHLDKYSV...  \n",
       "\n",
       "[3282 rows x 4 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_uniprot_ecoli = df_uniprot_ecoli[df_uniprot_ecoli.reviewed].drop(\n",
    "    [\"reviewed\"], axis=1\n",
    ")\n",
    "df_uniprot_ecoli\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It could also be interesting to have a bacteria-only dataset, and compare that to *E. coli*. Is there a tsv file that maps organism id to kingdom?"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Annotations for filtering out passive transporters\n",
    "\n",
    "Different databases have different definitions of transporter classes, and different annotations.\n",
    "\n",
    "#### TCDB:\n",
    "\n",
    "The *TCDB class 2* represents the *Electrochemical Potential-driven Transporters*. These are split into subclasses 2.A (Porters) and 2.B, 2.C and 2.D . Class 2.A seems to correspond to the typical definition of secondary active transporters, so we should look at the remaining three subclasses.\n",
    "\n",
    "#### Gene Ontology:\n",
    "\n",
    "The term *secondary active transmembrane transporter activity* (GO:0015291) is a molecular function annotation. The definition is:\n",
    "\n",
    "*Enables the transfer of a solute from one side of a membrane to the other, up its concentration gradient. The transporter binds the solute and undergoes a series of conformational changes. Transport works equally well in either direction and is driven by a chemiosmotic source of energy, not direct ATP coupling. Secondary active transporters include symporters and antiporters.*\n",
    "\n",
    "The GO entry cites the TCDB paper as its source, so it seems to use the same definitions.\n",
    "\n",
    "Child terms include: symporter activity, uniporter activity, antiporter activity.\n",
    "\n",
    "We should look into electronically inferred annotations (IEA). How would they impact the number of secondary active transport annotations in our dataset?\n",
    "\n",
    "Qualifiers for the annotations should be filtered by \"enables\", which means that the protein is directly responsible for the function.\n",
    "\n",
    "#### Keywords:\n",
    "\n",
    "- Symport (KW-0769): Protein involved in the transport of solutes across a biological membrane in one direction, which depends on the transport of another solute in the same direction. One molecule can move up an electrochemical gradient because the movement of the other molecule is more favorable. Example: the sodium/glucose co-transport.\n",
    "- Antiport (KW-0050): Protein involved in the transport of a solute across a biological membrane coupled, directly, to the transport of a different solute in the opposite direction.\n",
    "- There is no keywords for uniport, when looking at examples it seems like they are just annotated with their substrate keyword and \"Transport\". Some Uniporters in Uniprot seem to be annotated with Symport or Antiport in GO. Uniprot typically depends on a concentration gradient between the two compartments, independently of the movement of any other molecular species. Example is the GLUT family of sugar transporters, which is responsible for sugar uptake in mammals (Reviews [1](https://doi.org/10.3390/ijms23158698), [2](https://doi.org/10.1007/s00424-020-02411-3)).\n",
    "\n",
    "\n",
    "#### Interpro:\n",
    "\n",
    "The biggest superfamily of secondary active transporters is the MFS family. All members of that family contain the same sequence domain. Generally, the TCDB is structured along protein families, which are typically defined by the existence of a particula domain. If we create a dataset of all secondary active transport families, we can use Interpro annotations as well. \n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TCDB annotations analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tcdb_id</th>\n",
       "      <th>variable</th>\n",
       "      <th>value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.A.52.2.2</td>\n",
       "      <td>chebi_id</td>\n",
       "      <td>CHEBI:23337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.A.52.2.2</td>\n",
       "      <td>chebi_id</td>\n",
       "      <td>CHEBI:25517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.A.22.2.5</td>\n",
       "      <td>chebi_id</td>\n",
       "      <td>CHEBI:9175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.A.22.2.5</td>\n",
       "      <td>chebi_id</td>\n",
       "      <td>CHEBI:8345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.A.90.2.4</td>\n",
       "      <td>chebi_id</td>\n",
       "      <td>CHEBI:8816</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34308</th>\n",
       "      <td>2.A.47.1.5</td>\n",
       "      <td>Uniprot</td>\n",
       "      <td>Q9W7I2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34309</th>\n",
       "      <td>2.A.18.3.2</td>\n",
       "      <td>Uniprot</td>\n",
       "      <td>Q9XE48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34310</th>\n",
       "      <td>2.A.18.3.3</td>\n",
       "      <td>Uniprot</td>\n",
       "      <td>Q9XE49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34311</th>\n",
       "      <td>2.A.29.11.2</td>\n",
       "      <td>Uniprot</td>\n",
       "      <td>Q9ZNY4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34312</th>\n",
       "      <td>1.C.71.2.3</td>\n",
       "      <td>Uniprot</td>\n",
       "      <td>T1WIB6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34313 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           tcdb_id  variable        value\n",
       "0       2.A.52.2.2  chebi_id  CHEBI:23337\n",
       "1       2.A.52.2.2  chebi_id  CHEBI:25517\n",
       "2       2.A.22.2.5  chebi_id   CHEBI:9175\n",
       "3       2.A.22.2.5  chebi_id   CHEBI:8345\n",
       "4       2.A.90.2.4  chebi_id   CHEBI:8816\n",
       "...            ...       ...          ...\n",
       "34308   2.A.47.1.5   Uniprot       Q9W7I2\n",
       "34309   2.A.18.3.2   Uniprot       Q9XE48\n",
       "34310   2.A.18.3.3   Uniprot       Q9XE49\n",
       "34311  2.A.29.11.2   Uniprot       Q9ZNY4\n",
       "34312   1.C.71.2.3   Uniprot       T1WIB6\n",
       "\n",
       "[34313 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcdb = load_df(\"tcdb_substrates\")\n",
    "df_tcdb\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filtering for E Coli and transforming from long table to sparse table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>tcdb_id</th>\n",
       "      <th>substrate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A5A618</td>\n",
       "      <td>3.D.4.3.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A5A627</td>\n",
       "      <td>1.C.103.1.1</td>\n",
       "      <td>inorganic anion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A5A627</td>\n",
       "      <td>1.C.103.1.1</td>\n",
       "      <td>molecule</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C1P5Z7</td>\n",
       "      <td>4.A.1.1.1</td>\n",
       "      <td>D-glucopyranose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C1P611</td>\n",
       "      <td>1.C.135.1.1</td>\n",
       "      <td>inorganic ion</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1201</th>\n",
       "      <td>Q47706</td>\n",
       "      <td>1.B.25.1.5</td>\n",
       "      <td>beta-D-glucosiduronic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1202</th>\n",
       "      <td>Q59385</td>\n",
       "      <td>3.A.3.5.5</td>\n",
       "      <td>copper(1+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <td>Q59385</td>\n",
       "      <td>3.A.3.5.5</td>\n",
       "      <td>silver(1+)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1204</th>\n",
       "      <td>Q6BEX0</td>\n",
       "      <td>3.A.1.2.25</td>\n",
       "      <td>UDP-sugar</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1205</th>\n",
       "      <td>Q6BF25</td>\n",
       "      <td>1.C.64.1.13</td>\n",
       "      <td>inorganic ion</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1206 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Uniprot      tcdb_id                   substrate\n",
       "0     A5A618    3.D.4.3.2                         NaN\n",
       "1     A5A627  1.C.103.1.1             inorganic anion\n",
       "2     A5A627  1.C.103.1.1                    molecule\n",
       "3     C1P5Z7    4.A.1.1.1             D-glucopyranose\n",
       "4     C1P611  1.C.135.1.1               inorganic ion\n",
       "...      ...          ...                         ...\n",
       "1201  Q47706   1.B.25.1.5  beta-D-glucosiduronic acid\n",
       "1202  Q59385    3.A.3.5.5                  copper(1+)\n",
       "1203  Q59385    3.A.3.5.5                  silver(1+)\n",
       "1204  Q6BEX0   3.A.1.2.25                   UDP-sugar\n",
       "1205  Q6BF25  1.C.64.1.13               inorganic ion\n",
       "\n",
       "[1206 rows x 3 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcdb_uniprot = df_tcdb[df_tcdb.variable == \"Uniprot\"][[\"tcdb_id\", \"value\"]].rename(\n",
    "    columns={\"value\": \"Uniprot\"}\n",
    ")\n",
    "tcdb_chebi = df_tcdb[df_tcdb.variable == \"chebi_term\"][[\"tcdb_id\", \"value\"]].rename(\n",
    "    columns={\"value\": \"substrate\"}\n",
    ")\n",
    "\n",
    "df_tcdb_sparse = pd.merge(tcdb_uniprot, tcdb_chebi, how=\"outer\", on=\"tcdb_id\")\n",
    "df_tcdb_sparse_ecoli = (\n",
    "    df_tcdb_sparse[df_tcdb_sparse.Uniprot.isin(df_uniprot_ecoli.index)][\n",
    "        [\"Uniprot\", \"tcdb_id\", \"substrate\"]\n",
    "    ]\n",
    "    .sort_values([\"Uniprot\", \"tcdb_id\", \"substrate\"])\n",
    "    .reset_index(drop=True)\n",
    ")\n",
    "df_tcdb_sparse_ecoli\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Filtering for class 2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>tcdb_id</th>\n",
       "      <th>substrate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P02920</td>\n",
       "      <td>2.A.1.5.1</td>\n",
       "      <td>4-nitrophenyl alpha-D-galactoside</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P02920</td>\n",
       "      <td>2.A.1.5.1</td>\n",
       "      <td>4-nitrophenyl-beta-D-galactoside</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P02920</td>\n",
       "      <td>2.A.1.5.1</td>\n",
       "      <td>lactose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P02920</td>\n",
       "      <td>2.A.1.5.1</td>\n",
       "      <td>maltose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P02920</td>\n",
       "      <td>2.A.1.5.1</td>\n",
       "      <td>melibiose</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>Q46839</td>\n",
       "      <td>2.A.14.1.2</td>\n",
       "      <td>lactate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519</th>\n",
       "      <td>Q46892</td>\n",
       "      <td>2.A.8.1.7</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520</th>\n",
       "      <td>Q46909</td>\n",
       "      <td>2.A.1.1.95</td>\n",
       "      <td>metabolite</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521</th>\n",
       "      <td>Q47142</td>\n",
       "      <td>2.A.1.27.1</td>\n",
       "      <td>hydratropic acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>522</th>\n",
       "      <td>Q47377</td>\n",
       "      <td>2.A.7.22.1</td>\n",
       "      <td>polymyxin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>523 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Uniprot     tcdb_id                          substrate\n",
       "0    P02920   2.A.1.5.1  4-nitrophenyl alpha-D-galactoside\n",
       "1    P02920   2.A.1.5.1   4-nitrophenyl-beta-D-galactoside\n",
       "2    P02920   2.A.1.5.1                            lactose\n",
       "3    P02920   2.A.1.5.1                            maltose\n",
       "4    P02920   2.A.1.5.1                          melibiose\n",
       "..      ...         ...                                ...\n",
       "518  Q46839  2.A.14.1.2                            lactate\n",
       "519  Q46892   2.A.8.1.7                                NaN\n",
       "520  Q46909  2.A.1.1.95                         metabolite\n",
       "521  Q47142  2.A.1.27.1                   hydratropic acid\n",
       "522  Q47377  2.A.7.22.1                          polymyxin\n",
       "\n",
       "[523 rows x 3 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcdb_sparse_ecoli_class2 = df_tcdb_sparse_ecoli[\n",
    "    df_tcdb_sparse_ecoli.tcdb_id.str.startswith(\"2\")\n",
    "].reset_index(drop=True)\n",
    "df_tcdb_sparse_ecoli_class2\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What is the superfamily distribution?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.A    219\n",
       "2.C      2\n",
       "Name: tcdb_id, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcdb_sparse_ecoli_class2.tcdb_id.drop_duplicates().transform(\n",
    "    lambda x: x[:3]\n",
    ").value_counts()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Almost all proteins are in the Porters family, except for four proteins in the Ion-gradient-driven energizers family. According to TCDB, their mechanism is poorly understood, but they could be moving protons into the periplasmic space to enable the uptake of substrates through the passive transporters in the outer membrane. They use \"proton motive force\". Three of them seem to be related."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>tcdb_id</th>\n",
       "      <th>substrate</th>\n",
       "      <th>protein_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>P02929</td>\n",
       "      <td>2.C.1.1.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Protein TonB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>P0A855</td>\n",
       "      <td>2.C.1.2.1</td>\n",
       "      <td>phospholipid</td>\n",
       "      <td>Tol-Pal system protein TolB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>P19934</td>\n",
       "      <td>2.C.1.2.1</td>\n",
       "      <td>phospholipid</td>\n",
       "      <td>Tol-Pal system protein TolA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>898</th>\n",
       "      <td>P45955</td>\n",
       "      <td>2.C.1.2.1</td>\n",
       "      <td>phospholipid</td>\n",
       "      <td>Cell division coordinator CpoB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Uniprot    tcdb_id     substrate                   protein_names\n",
       "18   P02929  2.C.1.1.1           NaN                    Protein TonB\n",
       "110  P0A855  2.C.1.2.1  phospholipid     Tol-Pal system protein TolB\n",
       "543  P19934  2.C.1.2.1  phospholipid     Tol-Pal system protein TolA\n",
       "898  P45955  2.C.1.2.1  phospholipid  Cell division coordinator CpoB"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcdb_sparse_ecoli[df_tcdb_sparse_ecoli.tcdb_id.str.startswith(\"2.C\")].merge(\n",
    "    df_uniprot_ecoli.protein_names, how=\"left\", left_on=\"Uniprot\", right_index=True\n",
    ")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many proteins in E coli have a secondary active transport system annotation?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "238"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "secondary_active_transporters_tcdb = set(df_tcdb_sparse_ecoli_class2.Uniprot.unique())\n",
    "len(secondary_active_transporters_tcdb)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many different secondary active transport systems are present in E Coli?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "221"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcdb_sparse_ecoli_class2.tcdb_id.unique().shape[0]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What are the proteins that share the same transport system?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    208\n",
       "2     10\n",
       "3      2\n",
       "4      1\n",
       "Name: Uniprot, dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transport_system_number_of_proteins = (\n",
    "    df_tcdb_sparse_ecoli_class2.drop(\"substrate\", axis=1)\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"tcdb_id\")\n",
    "    .Uniprot.apply(len)\n",
    ")\n",
    "transport_system_number_of_proteins.value_counts()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "208 transport systems are annotated with one protein. Only one is annotated with more than three proteins. What is the reason for annotating a transport system with multiple proteins?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tcdb_id\n",
       "2.A.64.1.1    4\n",
       "Name: Uniprot, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transport_system_number_of_proteins[transport_system_number_of_proteins == 4]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The transport system with four proteins is a protein complex (TatABCE), that explains it. \n",
    "Is that true for all of them?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tcdb_id</th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>gene_names</th>\n",
       "      <th>protein_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.A.108.2.3</td>\n",
       "      <td>P31545</td>\n",
       "      <td>efeB ycdB b1019 JW1004</td>\n",
       "      <td>Deferrochelatase (EC 4.98.1.1) (Peroxidase Efe...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2.A.108.2.3</td>\n",
       "      <td>P0AB24</td>\n",
       "      <td>efeO ycdO b1018 JW1003</td>\n",
       "      <td>Iron uptake system component EfeO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2.A.113.1.1</td>\n",
       "      <td>P76425</td>\n",
       "      <td>rcnA yohM b2106 JW2093</td>\n",
       "      <td>Nickel/cobalt efflux system RcnA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2.A.113.1.1</td>\n",
       "      <td>P64534</td>\n",
       "      <td>rcnB yohN b2107 JW5346</td>\n",
       "      <td>Nickel/cobalt homeostasis protein RcnB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.A.37.1.1</td>\n",
       "      <td>P03819</td>\n",
       "      <td>kefC trkC b0047 JW0046</td>\n",
       "      <td>Glutathione-regulated potassium-efflux system ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2.A.37.1.1</td>\n",
       "      <td>P0A754</td>\n",
       "      <td>kefF yabF b0046 JW0045</td>\n",
       "      <td>Glutathione-regulated potassium-efflux system ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2.A.38.1.1</td>\n",
       "      <td>P0AGI8</td>\n",
       "      <td>trkA b3290 JW3251</td>\n",
       "      <td>Trk system potassium uptake protein TrkA (K(+)...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2.A.38.1.1</td>\n",
       "      <td>P0AFZ7</td>\n",
       "      <td>trkH b3849 JW5576</td>\n",
       "      <td>Trk system potassium uptake protein TrkH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2.A.56.1.2</td>\n",
       "      <td>P37674</td>\n",
       "      <td>yiaM b3577 JW3549</td>\n",
       "      <td>2,3-diketo-L-gulonate TRAP transporter small p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2.A.56.1.2</td>\n",
       "      <td>P37676</td>\n",
       "      <td>yiaO b3579 JW3551</td>\n",
       "      <td>2,3-diketo-L-gulonate-binding periplasmic prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>2.A.6.1.4</td>\n",
       "      <td>P38054</td>\n",
       "      <td>cusA ybdE b0575 JW0564</td>\n",
       "      <td>Cation efflux system protein CusA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>2.A.6.1.4</td>\n",
       "      <td>P77239</td>\n",
       "      <td>cusB ylcD b0574 JW0563</td>\n",
       "      <td>Cation efflux system protein CusB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>2.A.6.1.4</td>\n",
       "      <td>P77214</td>\n",
       "      <td>cusF cusX ylcC b0573 JW0562</td>\n",
       "      <td>Cation efflux system protein CusF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>2.A.6.2.12</td>\n",
       "      <td>P76398</td>\n",
       "      <td>mdtB yegN b2075 JW2060</td>\n",
       "      <td>Multidrug resistance protein MdtB (Multidrug t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>2.A.6.2.12</td>\n",
       "      <td>P76399</td>\n",
       "      <td>mdtC yegO b2076 JW2061</td>\n",
       "      <td>Multidrug resistance protein MdtC (Multidrug t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>2.A.6.2.2</td>\n",
       "      <td>P31224</td>\n",
       "      <td>acrB acrE b0462 JW0451</td>\n",
       "      <td>Multidrug efflux pump subunit AcrB (AcrAB-TolC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>2.A.6.2.2</td>\n",
       "      <td>P0AAW9</td>\n",
       "      <td>acrZ ybhT b0762 JW5102</td>\n",
       "      <td>Multidrug efflux pump accessory protein AcrZ (...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>2.A.6.4.1</td>\n",
       "      <td>P0AG90</td>\n",
       "      <td>secD b0408 JW0398</td>\n",
       "      <td>Protein translocase subunit SecD (Sec transloc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>2.A.6.4.1</td>\n",
       "      <td>P0AG93</td>\n",
       "      <td>secF b0409 JW0399</td>\n",
       "      <td>Protein translocase subunit SecF (Sec transloc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>2.A.64.1.1</td>\n",
       "      <td>P69428</td>\n",
       "      <td>tatA mttA1 yigT b3836 JW3813</td>\n",
       "      <td>Sec-independent protein translocase protein TatA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>2.A.64.1.1</td>\n",
       "      <td>P69425</td>\n",
       "      <td>tatB mttA2 ysgB b3838 JW5580</td>\n",
       "      <td>Sec-independent protein translocase protein TatB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>2.A.64.1.1</td>\n",
       "      <td>P69423</td>\n",
       "      <td>tatC mttB yigU yigV b3839 JW3815</td>\n",
       "      <td>Sec-independent protein translocase protein TatC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>2.A.64.1.1</td>\n",
       "      <td>P0A843</td>\n",
       "      <td>tatE ybeC b0627 JW0622</td>\n",
       "      <td>Sec-independent protein translocase protein TatE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>2.A.7.22.1</td>\n",
       "      <td>Q47377</td>\n",
       "      <td>arnE pmrL yfbW b4544 JW2252</td>\n",
       "      <td>Probable 4-amino-4-deoxy-L-arabinose-phosphoun...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>2.A.7.22.1</td>\n",
       "      <td>P76474</td>\n",
       "      <td>arnF pmrM yfbJ b2258 JW5373</td>\n",
       "      <td>Probable 4-amino-4-deoxy-L-arabinose-phosphoun...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>2.A.79.2.1</td>\n",
       "      <td>P0ADD2</td>\n",
       "      <td>yjjB b4363 JW4327</td>\n",
       "      <td>Probable succinate transporter subunit YjjB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>2.A.79.2.1</td>\n",
       "      <td>P0ADD5</td>\n",
       "      <td>yjjP b4364 JW5796</td>\n",
       "      <td>Probable succinate transporter subunit YjjP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>2.C.1.2.1</td>\n",
       "      <td>P45955</td>\n",
       "      <td>cpoB ybgF b0742 JW0732</td>\n",
       "      <td>Cell division coordinator CpoB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>2.C.1.2.1</td>\n",
       "      <td>P19934</td>\n",
       "      <td>tolA cim excC lky b0739 JW0729</td>\n",
       "      <td>Tol-Pal system protein TolA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>2.C.1.2.1</td>\n",
       "      <td>P0A855</td>\n",
       "      <td>tolB b0740 JW5100</td>\n",
       "      <td>Tol-Pal system protein TolB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        tcdb_id Uniprot                        gene_names  \\\n",
       "0   2.A.108.2.3  P31545            efeB ycdB b1019 JW1004   \n",
       "1   2.A.108.2.3  P0AB24            efeO ycdO b1018 JW1003   \n",
       "2   2.A.113.1.1  P76425            rcnA yohM b2106 JW2093   \n",
       "3   2.A.113.1.1  P64534            rcnB yohN b2107 JW5346   \n",
       "4    2.A.37.1.1  P03819            kefC trkC b0047 JW0046   \n",
       "5    2.A.37.1.1  P0A754            kefF yabF b0046 JW0045   \n",
       "6    2.A.38.1.1  P0AGI8                 trkA b3290 JW3251   \n",
       "7    2.A.38.1.1  P0AFZ7                 trkH b3849 JW5576   \n",
       "8    2.A.56.1.2  P37674                 yiaM b3577 JW3549   \n",
       "9    2.A.56.1.2  P37676                 yiaO b3579 JW3551   \n",
       "10    2.A.6.1.4  P38054            cusA ybdE b0575 JW0564   \n",
       "11    2.A.6.1.4  P77239            cusB ylcD b0574 JW0563   \n",
       "12    2.A.6.1.4  P77214       cusF cusX ylcC b0573 JW0562   \n",
       "13   2.A.6.2.12  P76398            mdtB yegN b2075 JW2060   \n",
       "14   2.A.6.2.12  P76399            mdtC yegO b2076 JW2061   \n",
       "15    2.A.6.2.2  P31224            acrB acrE b0462 JW0451   \n",
       "16    2.A.6.2.2  P0AAW9            acrZ ybhT b0762 JW5102   \n",
       "17    2.A.6.4.1  P0AG90                 secD b0408 JW0398   \n",
       "18    2.A.6.4.1  P0AG93                 secF b0409 JW0399   \n",
       "19   2.A.64.1.1  P69428      tatA mttA1 yigT b3836 JW3813   \n",
       "20   2.A.64.1.1  P69425      tatB mttA2 ysgB b3838 JW5580   \n",
       "21   2.A.64.1.1  P69423  tatC mttB yigU yigV b3839 JW3815   \n",
       "22   2.A.64.1.1  P0A843            tatE ybeC b0627 JW0622   \n",
       "23   2.A.7.22.1  Q47377       arnE pmrL yfbW b4544 JW2252   \n",
       "24   2.A.7.22.1  P76474       arnF pmrM yfbJ b2258 JW5373   \n",
       "25   2.A.79.2.1  P0ADD2                 yjjB b4363 JW4327   \n",
       "26   2.A.79.2.1  P0ADD5                 yjjP b4364 JW5796   \n",
       "27    2.C.1.2.1  P45955            cpoB ybgF b0742 JW0732   \n",
       "28    2.C.1.2.1  P19934    tolA cim excC lky b0739 JW0729   \n",
       "29    2.C.1.2.1  P0A855                 tolB b0740 JW5100   \n",
       "\n",
       "                                        protein_names  \n",
       "0   Deferrochelatase (EC 4.98.1.1) (Peroxidase Efe...  \n",
       "1                   Iron uptake system component EfeO  \n",
       "2                    Nickel/cobalt efflux system RcnA  \n",
       "3              Nickel/cobalt homeostasis protein RcnB  \n",
       "4   Glutathione-regulated potassium-efflux system ...  \n",
       "5   Glutathione-regulated potassium-efflux system ...  \n",
       "6   Trk system potassium uptake protein TrkA (K(+)...  \n",
       "7            Trk system potassium uptake protein TrkH  \n",
       "8   2,3-diketo-L-gulonate TRAP transporter small p...  \n",
       "9   2,3-diketo-L-gulonate-binding periplasmic prot...  \n",
       "10                  Cation efflux system protein CusA  \n",
       "11                  Cation efflux system protein CusB  \n",
       "12                  Cation efflux system protein CusF  \n",
       "13  Multidrug resistance protein MdtB (Multidrug t...  \n",
       "14  Multidrug resistance protein MdtC (Multidrug t...  \n",
       "15  Multidrug efflux pump subunit AcrB (AcrAB-TolC...  \n",
       "16  Multidrug efflux pump accessory protein AcrZ (...  \n",
       "17  Protein translocase subunit SecD (Sec transloc...  \n",
       "18  Protein translocase subunit SecF (Sec transloc...  \n",
       "19   Sec-independent protein translocase protein TatA  \n",
       "20   Sec-independent protein translocase protein TatB  \n",
       "21   Sec-independent protein translocase protein TatC  \n",
       "22   Sec-independent protein translocase protein TatE  \n",
       "23  Probable 4-amino-4-deoxy-L-arabinose-phosphoun...  \n",
       "24  Probable 4-amino-4-deoxy-L-arabinose-phosphoun...  \n",
       "25        Probable succinate transporter subunit YjjB  \n",
       "26        Probable succinate transporter subunit YjjP  \n",
       "27                     Cell division coordinator CpoB  \n",
       "28                        Tol-Pal system protein TolA  \n",
       "29                        Tol-Pal system protein TolB  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transport_system_number_of_proteins[\n",
    "    transport_system_number_of_proteins > 1\n",
    "].index.to_series().reset_index(drop=True).drop_duplicates().to_frame().merge(\n",
    "    df_tcdb_sparse_ecoli_class2, how=\"left\", on=\"tcdb_id\"\n",
    ").merge(\n",
    "    df_uniprot_ecoli[[\"gene_names\", \"protein_names\"]],\n",
    "    how=\"left\",\n",
    "    left_on=\"Uniprot\",\n",
    "    right_index=True,\n",
    ").drop(\n",
    "    \"substrate\", axis=1\n",
    ").drop_duplicates().sort_values(\n",
    "    [\"tcdb_id\", \"gene_names\"]\n",
    ").reset_index(\n",
    "    drop=True\n",
    ")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Yes, they all have matching names. Some of them could be operons?"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many have chebi terms for substrates?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "transport systems without chebi annotations: 38\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "2.A.3.1.15    0\n",
       "2.A.2.3.9     0\n",
       "2.A.5.5.1     0\n",
       "2.A.1.2.65    0\n",
       "2.A.1.26.1    0\n",
       "             ..\n",
       "2.A.9.3.1     1\n",
       "2.A.95.1.1    1\n",
       "2.A.95.1.5    1\n",
       "2.A.96.1.1    3\n",
       "2.C.1.2.1     1\n",
       "Length: 221, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp = df_tcdb_sparse_ecoli_class2[[\"tcdb_id\", \"substrate\"]].drop_duplicates()\n",
    "transport_systems_nosubstrates = set(\n",
    "    tmp[tmp.substrate.isnull() | (tmp.substrate == \"molecule\")].tcdb_id\n",
    ")\n",
    "print(\n",
    "    \"transport systems without chebi annotations:\", len(transport_systems_nosubstrates)\n",
    ")\n",
    "tmp = (\n",
    "    tmp[~tmp.tcdb_id.isin(transport_systems_nosubstrates)].groupby(\"tcdb_id\").apply(len)\n",
    ")\n",
    "\n",
    "transport_system_number_of_substrates = pd.concat(\n",
    "    [pd.Series(index=list(transport_systems_nosubstrates), data=0), tmp]\n",
    ")\n",
    "transport_system_number_of_substrates\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What is the distribution of substrate counts for the transport systems?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:xlabel='n_substrates'>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAh8AAAG0CAYAAACSbkVhAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAA9hAAAPYQGoP6dpAAApWUlEQVR4nO3de1zUdb7H8fdwcRgS0BC5rIjjiust0yOu5ZUuuqfM8tjpsmrqMfdoaoae8pKlqCdAO6kVHVp9lGmul93KynVNydK2zLwkVt5q88YGRBcC8gIK3/OHD+c0oeYofEfw9Xw8fo9H8/vN/OYzhPDiN7+ZcRhjjAAAACwJ8PcAAADgykJ8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYF+XuAn6usrFReXp7CwsLkcDj8PQ4AALgAxhiVlpYqLi5OAQHnP7Zx2cVHXl6e4uPj/T0GAAC4CLm5uWrSpMl5r3PZxUdYWJik08OHh4f7eRoAAHAhSkpKFB8f7/k9fj6XXXyceaolPDyc+AAAoJa5kFMmOOEUAABYRXwAAACriA8AAGDVZXfOBwAAl6qiokInT5709xh1TnBwsAIDAy95P8QHAKDOMMaooKBAP/zwg79HqbMaNGigmJiYS3ovLuIDAFBnnAmPxo0bKzQ0lDerrEbGGB07dkyFhYWSpNjY2IveF/EBAKgTKioqPOERGRnp73HqJJfLJUkqLCxU48aNL/opGE44BQDUCWfO8QgNDfXzJHXbma/vpZxTQ3wAAOoUnmqpWdXx9SU+AACAVcQHAACwihNOAQB1XrPJa6ze36GMvlbvr7bhyAcAANChQ4fkcDiUk5NT4/dFfAAAAKuIDwAALgOVlZWaPXu2WrRoIafTqaZNm+qJJ56QJH366ae68cYb5XK5FBkZqf/8z//Ujz/+6LltcnKyUlJSvPbXv39/DRs2zHO5WbNmSktL0/DhwxUWFqamTZtqwYIFnu1ut1uS1LFjRzkcDiUnJ9fYY+WcjxpQnc8t8rwhAFwZpkyZooULF2revHnq3r278vPztW/fPh07dkz/+q//quuuu07btm1TYWGhRowYobFjx+qll17y6T6eeuopzZo1S48++qheeeUVPfDAA+rZs6datWqlrVu36re//a3efvtttW3bVvXq1auZByriAwAAvystLdXTTz+tzMxMDR06VJL061//Wt27d9fChQt1/PhxLVmyRFdddZUkKTMzU/369dPs2bMVHR19wfdz6623avTo0ZKkSZMmad68edq4caNatWqlqKgoSVJkZKRiYmKq+RF642kXAAD8bO/evSorK9NNN9101m3XXnutJzwkqVu3bqqsrNT+/ft9up/27dt7/tvhcCgmJsbzWS02ER8AAPjZmc9MORtjzDnfVfTM+oCAABljvLad7e3Pg4ODq9y+srLS13EvGfEBAICfJSYmyuVyacOGDVW2tWnTRjk5OTp69Khn3QcffKCAgAC1bNlSkhQVFaX8/HzP9oqKCn322Wc+zXDmHI+KioqLeQg+IT4AAPCzkJAQTZo0SRMnTtSSJUv05ZdfasuWLXrhhRc0aNAghYSEaOjQofrss8/07rvv6sEHH9R9993nOd/jxhtv1Jo1a7RmzRrt27dPo0eP1g8//ODTDI0bN5bL5dJbb72lr7/+WsXFxTXwSE/jhFMAQJ1XG145+PjjjysoKEjTpk1TXl6eYmNjNWrUKIWGhmrdunV66KGH1LlzZ4WGhurOO+/U3LlzPbcdPny4du3apSFDhigoKEjjx4/XDTfc4NP9BwUF6ZlnntHMmTM1bdo09ejRQxs3bqzmR3maw/z8SSI/KykpUUREhIqLixUeHu7vcS4KL7UFAPtOnDihgwcPyu12KyQkxN/j1Fnn+jr78vubp10AAIBVxAcAALCK+AAAAFYRHwAAwCriAwBQp/jjTbOuJNXx9eWltgCAOqFevXoKCAhQXl6eoqKiVK9evXO+Myh8Z4xReXm5vvnmGwUEBFzSB88RHwCAOiEgIEBut1v5+fnKy8vz9zh1VmhoqJo2baqAgIt/8oT4AADUGfXq1VPTpk116tQpK28TfqUJDAxUUFDQJR9RIj4AAHWKw+FQcHBwlQ9Rw+WDE04BAIBVxAcAALCK+AAAAFYRHwAAwCqf4uPUqVN67LHH5Ha75XK51Lx5c82cOdPrDUeMMUpNTVVcXJxcLpeSk5O1e/fuah8cAADUTj7Fx+zZs/X8888rMzNTe/fu1Zw5c/Tkk0/q2Wef9Vxnzpw5mjt3rjIzM7Vt2zbFxMSod+/eKi0trfbhAQBA7ePTS20//PBD3XHHHerbt68kqVmzZlq+fLm2b98u6fRRj/nz52vq1KkaMGCAJGnx4sWKjo7WsmXLNHLkyCr7LCsrU1lZmedySUnJRT8YAABw+fPpyEf37t21YcMGff7555KkXbt26f3339ett94qSTp48KAKCgrUp08fz22cTqd69eqlzZs3n3Wf6enpioiI8Czx8fEX+1gAAEAt4NORj0mTJqm4uFitWrVSYGCgKioq9MQTT+j3v/+9JKmgoECSFB0d7XW76OhoHT58+Kz7nDJliiZMmOC5XFJSQoAAAFCH+RQfK1eu1NKlS7Vs2TK1bdtWOTk5SklJUVxcnIYOHeq53s/fdtUYc863YnU6nXI6nRcxOgAAqI18io9HHnlEkydP1r333itJuuaaa3T48GGlp6dr6NChiomJkXT6CEhsbKzndoWFhVWOhgAAgCuTT+d8HDt2rMqn2AUGBnpeaut2uxUTE6Ps7GzP9vLycm3atEldu3athnEBAEBt59ORj379+umJJ55Q06ZN1bZtW+3cuVNz587V8OHDJZ1+uiUlJUVpaWlKTExUYmKi0tLSFBoaqoEDB9bIAwAAALWLT/Hx7LPP6vHHH9fo0aNVWFiouLg4jRw5UtOmTfNcZ+LEiTp+/LhGjx6toqIidenSRevXr1dYWFi1Dw8AAGofhzHG+HuInyopKVFERISKi4sVHh7u73EuSrPJa6ptX4cy+lbbvgAAqCm+/P7ms10AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABglc/x8dVXX2nw4MGKjIxUaGioOnTooB07dni2G2OUmpqquLg4uVwuJScna/fu3dU6NAAAqL18io+ioiJ169ZNwcHBWrt2rfbs2aOnnnpKDRo08Fxnzpw5mjt3rjIzM7Vt2zbFxMSod+/eKi0tre7ZAQBALRTky5Vnz56t+Ph4LVq0yLOuWbNmnv82xmj+/PmaOnWqBgwYIElavHixoqOjtWzZMo0cObLKPsvKylRWVua5XFJS4utjAAAAtYhPRz7efPNNJSUl6a677lLjxo3VsWNHLVy40LP94MGDKigoUJ8+fTzrnE6nevXqpc2bN591n+np6YqIiPAs8fHxF/lQAABAbeBTfBw4cEBZWVlKTEzUunXrNGrUKI0bN05LliyRJBUUFEiSoqOjvW4XHR3t2fZzU6ZMUXFxsWfJzc29mMcBAABqCZ+edqmsrFRSUpLS0tIkSR07dtTu3buVlZWlIUOGeK7ncDi8bmeMqbLuDKfTKafT6evcAACglvLpyEdsbKzatGnjta5169Y6cuSIJCkmJkaSqhzlKCwsrHI0BAAAXJl8io9u3bpp//79Xus+//xzJSQkSJLcbrdiYmKUnZ3t2V5eXq5Nmzapa9eu1TAuAACo7Xx62mX8+PHq2rWr0tLSdPfdd2vr1q1asGCBFixYIOn00y0pKSlKS0tTYmKiEhMTlZaWptDQUA0cOLBGHgAAAKhdfIqPzp07a9WqVZoyZYpmzpwpt9ut+fPna9CgQZ7rTJw4UcePH9fo0aNVVFSkLl26aP369QoLC6v24QEAQO3jMMYYfw/xUyUlJYqIiFBxcbHCw8P9Pc5FaTZ5TbXt61BG32rbFwAANcWX3998tgsAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALAqyN8DwJ5mk9dU274OZfSttn0BAK4sHPkAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKuIDwAAYBXxAQAArCI+AACAVcQHAACwivgAAABWER8AAMAq4gMAAFhFfAAAAKsuKT7S09PlcDiUkpLiWWeMUWpqquLi4uRyuZScnKzdu3df6pwAAKCOuOj42LZtmxYsWKD27dt7rZ8zZ47mzp2rzMxMbdu2TTExMerdu7dKS0sveVgAAFD7XVR8/Pjjjxo0aJAWLlyohg0betYbYzR//nxNnTpVAwYMULt27bR48WIdO3ZMy5Ytq7ahAQBA7XVR8TFmzBj17dtXN998s9f6gwcPqqCgQH369PGsczqd6tWrlzZv3nzWfZWVlamkpMRrAQAAdVeQrzdYsWKFPv74Y23btq3KtoKCAklSdHS01/ro6GgdPnz4rPtLT0/XjBkzfB0DAADUUj4d+cjNzdVDDz2kpUuXKiQk5JzXczgcXpeNMVXWnTFlyhQVFxd7ltzcXF9GAgAAtYxPRz527NihwsJCderUybOuoqJC7733njIzM7V//35Jp4+AxMbGeq5TWFhY5WjIGU6nU06n82JmBwAAtZBPRz5uuukmffrpp8rJyfEsSUlJGjRokHJyctS8eXPFxMQoOzvbc5vy8nJt2rRJXbt2rfbhAQBA7ePTkY+wsDC1a9fOa91VV12lyMhIz/qUlBSlpaUpMTFRiYmJSktLU2hoqAYOHFh9UwMAgFrL5xNOf8nEiRN1/PhxjR49WkVFRerSpYvWr1+vsLCw6r4rAABQC11yfGzcuNHrssPhUGpqqlJTUy911wAAoA7is10AAIBV1f60i03NJq+ptn0dyuhbbfsCAADnxpEPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFb5FB/p6enq3LmzwsLC1LhxY/Xv31/79+/3uo4xRqmpqYqLi5PL5VJycrJ2795drUMDAIDay6f42LRpk8aMGaMtW7YoOztbp06dUp8+fXT06FHPdebMmaO5c+cqMzNT27ZtU0xMjHr37q3S0tJqHx4AANQ+Qb5c+a233vK6vGjRIjVu3Fg7duxQz549ZYzR/PnzNXXqVA0YMECStHjxYkVHR2vZsmUaOXJk9U0OAABqpUs656O4uFiSdPXVV0uSDh48qIKCAvXp08dzHafTqV69emnz5s1n3UdZWZlKSkq8FgAAUHdddHwYYzRhwgR1795d7dq1kyQVFBRIkqKjo72uGx0d7dn2c+np6YqIiPAs8fHxFzsSAACoBS46PsaOHatPPvlEy5cvr7LN4XB4XTbGVFl3xpQpU1RcXOxZcnNzL3YkAABQC/h0zscZDz74oN5880299957atKkiWd9TEyMpNNHQGJjYz3rCwsLqxwNOcPpdMrpdF7MGAAAoBby6ciHMUZjx47Va6+9pnfeeUdut9tru9vtVkxMjLKzsz3rysvLtWnTJnXt2rV6JgYAALWaT0c+xowZo2XLlumNN95QWFiY5zyOiIgIuVwuORwOpaSkKC0tTYmJiUpMTFRaWppCQ0M1cODAGnkAAACgdvEpPrKysiRJycnJXusXLVqkYcOGSZImTpyo48ePa/To0SoqKlKXLl20fv16hYWFVcvAAACgdvMpPowxv3gdh8Oh1NRUpaamXuxMAACgDuOzXQAAgFUX9WoXoDo1m7ymWvd3KKNvte4PAFC9OPIBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYF+XsA4HLWbPKaatvXoYy+1bavy3UuALgQHPkAAABWER8AAMAq4gMAAFjFOR8Aqg3nogC4EBz5AAAAVhEfAADAKuIDAABYRXwAAACriA8AAGAV8QEAAKwiPgAAgFW8zwcAoNbgvWTqBo58AAAAq4gPAABgFfEBAACsIj4AAIBVxAcAALCK+AAAAFYRHwAAwCriAwAAWMWbjAGAn1yub5h1uc6FuoMjHwAAwCriAwAAWEV8AAAAq2rsnI///d//1ZNPPqn8/Hy1bdtW8+fPV48ePWrq7gDgnDiHATXtcv0eu1znqpEjHytXrlRKSoqmTp2qnTt3qkePHrrlllt05MiRmrg7AABQi9RIfMydO1f333+/RowYodatW2v+/PmKj49XVlZWTdwdAACoRar9aZfy8nLt2LFDkydP9lrfp08fbd68ucr1y8rKVFZW5rlcXFwsSSopKfnF+6osO3aJ0/6/C7m/C8VcvqnOuaTLdzbm8g1z+eZKmEu6fGdjrv/fboz55Z2ZavbVV18ZSeaDDz7wWv/EE0+Yli1bVrn+9OnTjSQWFhYWFhaWOrDk5ub+YivU2AmnDofD67Ixpso6SZoyZYomTJjguVxZWanvv/9ekZGRZ72+r0pKShQfH6/c3FyFh4df8v6qC3P5hrl8w1y+u1xnYy7fMJdvqnMuY4xKS0sVFxf3i9et9vho1KiRAgMDVVBQ4LW+sLBQ0dHRVa7vdDrldDq91jVo0KC6x1J4ePhl9T/8DObyDXP5hrl8d7nOxly+YS7fVNdcERERF3S9aj/htF69eurUqZOys7O91mdnZ6tr167VfXcAAKCWqZGnXSZMmKD77rtPSUlJuv7667VgwQIdOXJEo0aNqom7AwAAtUiNxMc999yj7777TjNnzlR+fr7atWunv/3tb0pISKiJuzsvp9Op6dOnV3lqx9+YyzfM5Rvm8t3lOhtz+Ya5fOOvuRzGXMhrYgAAAKoHn+0CAACsIj4AAIBVxAcAALCK+AAAAFYRH8Al4pxtAPBNjb29uj/885//VFZWljZv3qyCggI5HA5FR0era9euGjVqlOLj4/09Iuogp9OpXbt2qXXr1v4eBagx+fn5ysrK0vvvv6/8/HwFBgbK7Xarf//+GjZsmAIDA/09ImqROvNS2/fff1+33HKL4uPj1adPH0VHR8sYo8LCQmVnZys3N1dr165Vt27d/D1qFbm5uZo+fbpefPFF6/d9/Phx7dixQ1dffbXatGnjte3EiRP685//rCFDhlifa+/evdqyZYuuv/56tWrVSvv27dPTTz+tsrIyDR48WDfeeKP1mX76GUQ/9fTTT2vw4MGKjIyUJM2dO9fmWFUUFRVp8eLF+uKLLxQbG6uhQ4f6Jbx37typBg0ayO12S5KWLl2qrKwsHTlyRAkJCRo7dqzuvfde63NJ0oMPPqi7775bPXr08Mv9n8+zzz6r7du3q2/fvrr77rv18ssvKz09XZWVlRowYIBmzpypoCC7fzdu375dN998s9xut1wulz766CMNGjRI5eXlWrdunVq3bq1169YpLCzM6lyoxS75Y2wvE0lJSSYlJeWc21NSUkxSUpLFiS5cTk6OCQgIsH6/+/fvNwkJCcbhcJiAgADTq1cvk5eX59leUFDgl7nWrl1r6tWrZ66++moTEhJi1q5da6KioszNN99sbrrpJhMUFGQ2bNhgfS6Hw2E6dOhgkpOTvRaHw2E6d+5skpOTzQ033GB9rtjYWPPtt98aY4w5cOCAiYmJMTExMaZ3796mSZMmJiIiwuzdu9f6XB07djTvvPOOMcaYhQsXGpfLZcaNG2eysrJMSkqKqV+/vnnhhResz2WM8XzPJyYmmoyMDJOfn++XOX5u5syZJiwszNx5550mJibGZGRkmMjISPPf//3fJi0tzURFRZlp06ZZn6tbt24mNTXVc/nll182Xbp0McYY8/3335sOHTqYcePGWZ/rQhQUFJgZM2b47f5zc3NNaWlplfXl5eVm06ZNfpjImG+//da888475rvvvjPGGPPNN9+YjIwMM2PGDLNnzx4rM9SZ+AgJCTH79u075/a9e/eakJAQixP9vzfeeOO8y7x58/zyS75///7mtttuM99884354osvTL9+/Yzb7TaHDx82xvgvPq6//nozdepUY4wxy5cvNw0bNjSPPvqoZ/ujjz5qevfubX2utLQ043a7q4RPUFCQ2b17t/V5znA4HObrr782xhhz7733muTkZHP06FFjjDEnTpwwt912m/n3f/9363OFhoZ6vpc6duxo/vjHP3pt/9Of/mTatGljfS5jTn/N3n77bfPQQw+ZRo0ameDgYHP77beb1atXm4qKCr/MZIwxzZs3N6+++qox5vQfJYGBgWbp0qWe7a+99ppp0aKF9blcLpf58ssvPZcrKipMcHCwKSgoMMYYs379ehMXF2d9rgvhrz/u8vLyTOfOnU1AQIAJDAw0Q4YM8YoQf/18/eijj0xERIRxOBymYcOGZvv27cbtdpvExETTokUL43K5zI4dO2p8jjoTH26327z44ovn3P7iiy8at9ttcaL/d+avLIfDcc7FH9+EjRs3Np988onXutGjR5umTZuaL7/80m//OMLDw80XX3xhjDn9Qy4oKMjrH8Onn35qoqOjrc9ljDFbt241LVu2NP/1X/9lysvLjTGXV3ycLY62bNlimjRpYn2uyMhIs337dmPM6e+1nJwcr+3/+Mc/jMvlsj6XMd5fs/LycrNy5Urzu9/9zgQGBpq4uDjz6KOPer4HbXK5XJ5gM8aY4OBg89lnn3kuHzp0yISGhlqfKyEhwbz//vuey3l5ecbhcJhjx44ZY4w5ePCg3/6427Vr13mXlStX+uXn2JAhQ8x1111ntm3bZrKzs01SUpLp1KmT+f77740xp+PD4XBYn+vmm282I0aMMCUlJebJJ580TZo0MSNGjPBsv//++03//v1rfI46Ex/PPfecqVevnhkzZox5/fXXzYcffmi2bNliXn/9dTNmzBjjdDpNVlaWX2aLi4szq1atOuf2nTt3+uUfR1hY2FkPsY0dO9Y0adLEvPfee36PD2OMqV+/vtdfXYcOHfLbDzpjjCktLTVDhgwx7du3N5988okJDg72e3wUFhYaY05/r/30l5Uxp38xOJ1O63MNHjzY3H///cYYY+666y7z2GOPeW1PS0sz11xzjfW5jPGOj586fPiwmT59uklISPDL977b7TZr1641xhjz+eefm4CAAPPnP//Zs33NmjWmWbNm1ud66KGHTLt27czatWvNO++8Y2644QaTnJzs2f7WW2+ZX//619bnMub8f9ydWe+P/5dxcXHmo48+8lw+ceKEueOOO0yHDh3Md99957c/7ho2bOj5uV9eXm4CAgK85vz444/Nr371qxqfo87EhzHGrFixwnTp0sUEBQV5vvmCgoJMly5dzMqVK/02V79+/czjjz9+zu05OTl+KeDOnTubJUuWnHXbmDFjTIMGDfzyj6N9+/aeH8DGnD7ScfLkSc/lv//97347ivVTy5cvN9HR0SYgIMDv8XHNNdeYjh07mvr165vXXnvNa/umTZus/DD5ua+++so0a9bM9OzZ00yYMMG4XC7TvXt384c//MH07NnT1KtXz6xZs8b6XMacOz7OqKysNOvXr7c40WlTp041UVFRZsSIEcbtdpspU6aYpk2bmqysLPP888+b+Ph4M378eOtzlZaWmrvvvtvzs7Vr167mwIEDnu3r1q3ziiSbGjVqZF544QVz6NChsy5r1qzxy8+xq666ynz++ede606ePGn69+/v+cPFX3MdPHjQc/nnf9wdPnzYyh93deqltvfcc4/uuecenTx5Ut9++60kqVGjRgoODvbrXI888oiOHj16zu0tWrTQu+++a3Gi0/7t3/5Ny5cv13333VdlW2ZmpiorK/X8889bn+uBBx5QRUWF53K7du28tq9du9Yvr3b5uXvvvVfdu3fXjh07/PKJzWdMnz7d63JoaKjX5dWrV/vlVR1xcXHauXOnMjIytHr1ahljtHXrVuXm5qpbt2764IMPlJSUZH0uSUpISDjvS0MdDod69+5tcaLTZsyYIZfLpS1btmjkyJGaNGmS2rdvr4kTJ+rYsWPq16+fZs2aZX2u+vXra+XKlTpx4oROnTql+vXre23v06eP9ZnO6NSpk/Ly8s75b/CHH37wy3vxNG/eXJ988okSExM964KCgvSXv/xFd911l2677TbrM0lSfHy8Dhw4oGbNmkmSVqxYodjYWM/2/Px8NWrUqMbnqDMvtQUAXHlWrVqlo0ePavDgwWfdXlRUpDfffFNDhw61OtekSZOUk5OjdevWVdl26tQp3XnnnVq9erUqKyutzjVjxgz95je/OefL3KdOnap9+/bp1VdfrdE5iA8AAKrZqVOndOzYMYWHh591e0VFhf75z3/69ajp2Rw7dkyBgYFyOp01ej+8vToAoM7Kzc3V8OHDrd9vUFDQOcNDkvLy8jRjxgyLE12Y7777Tg888ECN3w9HPgAAddauXbv0L//yL17nkV0OrvS56tQJpwCAK8ubb7553u0HDhywNIk35jo/jnwAAGqtgIAAORyO876ixeFwWD/CwFy/MEeN7h0AgBoUGxurV199VZWVlWddPv74Y+a6DOciPgAAtVanTp3O+wvzl/7KrynMdX6c8wEAqLUu1zdxZK7z45wPAABgFU+7AAAAq4gPAABgFfEBAACsIj4AAIBVxAeAi3bo0CE5HA7l5OT4exQAtQjxAcBvqjteNm7cKIfDoR9++KFa9gegZhAfAC575eXl/h4BQDUiPoA6IDk5WePGjdPEiRN19dVXKyYmRqmpqRd029TUVDVt2lROp1NxcXEaN26cZ5vD4dDrr7/udf0GDRropZde8lq3b98+de3aVSEhIWrbtq02btzo2VZUVKRBgwYpKipKLpdLiYmJWrRokSTJ7XZLkjp27CiHw6Hk5GRJ0rBhw9S/f3+lp6crLi5OLVu2lCQtXbpUSUlJCgsLU0xMjAYOHKjCwkJJp4+i3HDDDZKkhg0byuFwaNiwYZIkY4zmzJmj5s2by+Vy6dprr9Urr7xyQTMCqH68wylQRyxevFgTJkzQRx99pA8//FDDhg1Tt27d1Lt373Pe5pVXXtG8efO0YsUKtW3bVgUFBdq1a5fP9/3II49o/vz5atOmjebOnavbb79dBw8eVGRkpB5//HHt2bNHa9euVaNGjfSPf/xDx48flyRt3bpVv/3tb/X222+rbdu2qlevnmefGzZsUHh4uLKzsz1v91xeXq5Zs2bpN7/5jQoLCzV+/HgNGzZMf/vb3xQfH69XX31Vd955p/bv36/w8HC5XC5J0mOPPabXXntNWVlZSkxM1HvvvafBgwcrKipKvXr1Ou+MAGqAAVDr9erVy3Tv3t1rXefOnc2kSZPOe7unnnrKtGzZ0pSXl591uySzatUqr3URERFm0aJFxhhjDh48aCSZjIwMz/aTJ0+aJk2amNmzZxtjjOnXr5/5j//4j7Pu/8ztd+7c6bV+6NChJjo62pSVlZ13/q1btxpJprS01BhjzLvvvmskmaKiIs91fvzxRxMSEmI2b97sddv777/f/P73v//FGQFUP552AeqI9u3be12OjY31PCVxLnfddZeOHz+u5s2b6w9/+INWrVqlU6dO+Xzf119/vee/g4KClJSUpL1790qSHnjgAa1YsUIdOnTQxIkTtXnz5gva5zXXXON1JESSdu7cqTvuuEMJCQkKCwvzPE1z5MiRc+5nz549OnHihHr37q369et7liVLlujLL7+8pBkBXBziA6gjgoODvS47HA5VVlae9zbx8fHav3+/nnvuOblcLo0ePVo9e/bUyZMnPfswP/v4pzPbfonD4ZAk3XLLLTp8+LBSUlKUl5enm266SQ8//PAv3v6qq67yunz06FH16dNH9evX19KlS7Vt2zatWrVK0vlPSD3zNVizZo1ycnI8y549ezznfVzsjAAuDvEBXOFcLpduv/12PfPMM9q4caM+/PBDffrpp5KkqKgo5efne677xRdf6NixY1X2sWXLFs9/nzp1Sjt27FCrVq0866KiojRs2DAtXbpU8+fP14IFCyTJc2SjoqLiF+fct2+fvv32W2VkZKhHjx5q1apVlSM7Z9tfmzZt5HQ6deTIEbVo0cJriY+P/8UZAVQ/TjgFrmAvvfSSKioq1KVLF4WGhurll1+Wy+VSQkKCJOnGG29UZmamrrvuOlVWVmrSpElVjrBI0nPPPafExES1bt1a8+bNU1FRkYYPHy5JmjZtmjp16qS2bduqrKxMf/3rX9W6dWtJUuPGjeVyufTWW2+pSZMmCgkJUURExFlnbdq0qerVq6dnn31Wo0aN0meffaZZs2Z5XSchIUEOh0N//etfdeutt8rlciksLEwPP/ywxo8fr8rKSnXv3l0lJSXavHmz6tevr6FDh553RgDVjyMfwBWsQYMGWrhwobp166b27dtrw4YNWr16tSIjIyVJTz31lOLj49WzZ08NHDhQDz/8sEJDQ6vsJyMjQ7Nnz9a1116rv//973rjjTfUqFEjSaePRkyZMkXt27dXz549FRgYqBUrVkg6fX7IM888oz/+8Y+Ki4vTHXfccc5Zo6Ki9NJLL+kvf/mL2rRpo4yMDP3P//yP13V+9atfacaMGZo8ebKio6M1duxYSdKsWbM0bdo0paenq3Xr1vrd736n1atXe17qe74ZAVQ/h/n5E7oAAAA1iCMfAADAKuIDqMP+9Kc/eb289KdL27Zt/T0egCsUT7sAdVhpaam+/vrrs24LDg72nFgKADYRHwAAwCqedgEAAFYRHwAAwCriAwAAWEV8AAAAq4gPAABgFfEBAACsIj4AAIBV/wcKRrh+iCbTjAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "transport_system_number_of_substrates.value_counts().reset_index(name=\"count\").rename(\n",
    "    columns={\"index\": \"n_substrates\"}\n",
    ").sort_values(\"n_substrates\").plot.bar(x=\"n_substrates\", y=\"count\")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What are the transport systems with a large number of substrates?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.A.1.2.19    13\n",
       "dtype: int64"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transport_system_number_of_substrates[transport_system_number_of_substrates > 12]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The transport system with 13 substrates is the \"Multidrug resistance exporter\" (MdfA). The large number of substrates also comes from well-known drugs that the transporters are known to carry."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.A.1.2.62    12\n",
       "dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transport_system_number_of_substrates[transport_system_number_of_substrates == 12]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The 12 substrate transporter is also a multidrug resistance pump called PunC."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.A.1.5.1    7\n",
       "dtype: int64"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transport_system_number_of_substrates[transport_system_number_of_substrates == 7]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The transporter with 7 substrates is lacY, a proton-sugar symporter"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene ontology"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "from subpred.ontology import (\n",
    "    GeneOntology,\n",
    "    GO_FILE,\n",
    "    add_go_ancestors,\n",
    "    EVIDENCE_CODE_TO_DESCRIPTION,\n",
    ")\n",
    "\n",
    "go_owl = GeneOntology(GO_FILE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>go_id</th>\n",
       "      <th>go_term</th>\n",
       "      <th>evidence_category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A5A627</td>\n",
       "      <td>IDA</td>\n",
       "      <td>GO:0003674</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>experimental_evidence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A5A627</td>\n",
       "      <td>IDA</td>\n",
       "      <td>GO:0005215</td>\n",
       "      <td>transporter activity</td>\n",
       "      <td>experimental_evidence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A5A627</td>\n",
       "      <td>IDA</td>\n",
       "      <td>GO:0005216</td>\n",
       "      <td>ion channel activity</td>\n",
       "      <td>experimental_evidence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A5A627</td>\n",
       "      <td>IDA</td>\n",
       "      <td>GO:0005253</td>\n",
       "      <td>anion channel activity</td>\n",
       "      <td>experimental_evidence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A5A627</td>\n",
       "      <td>IDA</td>\n",
       "      <td>GO:0008509</td>\n",
       "      <td>anion transmembrane transporter activity</td>\n",
       "      <td>experimental_evidence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76601</th>\n",
       "      <td>Q93K97</td>\n",
       "      <td>IBA</td>\n",
       "      <td>GO:0016818</td>\n",
       "      <td>hydrolase activity, acting on acid anhydrides,...</td>\n",
       "      <td>phylogenetically_inferred</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76602</th>\n",
       "      <td>Q93K97</td>\n",
       "      <td>IBA</td>\n",
       "      <td>GO:0019144</td>\n",
       "      <td>ADP-sugar diphosphatase activity</td>\n",
       "      <td>phylogenetically_inferred</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76603</th>\n",
       "      <td>Q93K97</td>\n",
       "      <td>IPI</td>\n",
       "      <td>GO:0003674</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>experimental_evidence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76604</th>\n",
       "      <td>Q93K97</td>\n",
       "      <td>IPI</td>\n",
       "      <td>GO:0005488</td>\n",
       "      <td>binding</td>\n",
       "      <td>experimental_evidence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76605</th>\n",
       "      <td>Q93K97</td>\n",
       "      <td>IPI</td>\n",
       "      <td>GO:0005515</td>\n",
       "      <td>protein binding</td>\n",
       "      <td>experimental_evidence</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76606 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Uniprot evidence_code       go_id  \\\n",
       "0      A5A627           IDA  GO:0003674   \n",
       "1      A5A627           IDA  GO:0005215   \n",
       "2      A5A627           IDA  GO:0005216   \n",
       "3      A5A627           IDA  GO:0005253   \n",
       "4      A5A627           IDA  GO:0008509   \n",
       "...       ...           ...         ...   \n",
       "76601  Q93K97           IBA  GO:0016818   \n",
       "76602  Q93K97           IBA  GO:0019144   \n",
       "76603  Q93K97           IPI  GO:0003674   \n",
       "76604  Q93K97           IPI  GO:0005488   \n",
       "76605  Q93K97           IPI  GO:0005515   \n",
       "\n",
       "                                                 go_term  \\\n",
       "0                                     molecular_function   \n",
       "1                                   transporter activity   \n",
       "2                                   ion channel activity   \n",
       "3                                 anion channel activity   \n",
       "4               anion transmembrane transporter activity   \n",
       "...                                                  ...   \n",
       "76601  hydrolase activity, acting on acid anhydrides,...   \n",
       "76602                   ADP-sugar diphosphatase activity   \n",
       "76603                                 molecular_function   \n",
       "76604                                            binding   \n",
       "76605                                    protein binding   \n",
       "\n",
       "               evidence_category  \n",
       "0          experimental_evidence  \n",
       "1          experimental_evidence  \n",
       "2          experimental_evidence  \n",
       "3          experimental_evidence  \n",
       "4          experimental_evidence  \n",
       "...                          ...  \n",
       "76601  phylogenetically_inferred  \n",
       "76602  phylogenetically_inferred  \n",
       "76603      experimental_evidence  \n",
       "76604      experimental_evidence  \n",
       "76605      experimental_evidence  \n",
       "\n",
       "[76606 rows x 5 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go = load_df(\"go\")\n",
    "df_go_ecoli = df_go[df_go.Uniprot.isin(df_uniprot_ecoli.index)].reset_index(drop=True)\n",
    "df_go_ecoli = add_go_ancestors(df_go_ecoli, go_owl)\n",
    "df_go_ecoli = df_go_ecoli.assign(\n",
    "    go_term=df_go_ecoli.go_id.transform(go_owl.get_label)\n",
    ")\n",
    "\n",
    "df_go_ecoli_mf = (\n",
    "    df_go_ecoli[(df_go_ecoli.qualifier == \"enables\") & (df_go_ecoli.aspect == \"F\")]\n",
    "    .reset_index(drop=True)\n",
    "    .drop([\"qualifier\", \"aspect\"], axis=1)\n",
    ")\n",
    "df_go_ecoli_mf = df_go_ecoli_mf.sort_values(df_go_ecoli_mf.columns.tolist())\n",
    "df_go_ecoli_mf = df_go_ecoli_mf.assign(\n",
    "    evidence_category=df_go_ecoli_mf.evidence_code.map(EVIDENCE_CODE_TO_DESCRIPTION)\n",
    ")\n",
    "df_go_ecoli_mf = df_go_ecoli_mf.reset_index(drop=True)\n",
    "df_go_ecoli_mf\n",
    "# TODO df_go_ecoli is used below, but does not have ancestors!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "electronic_annotation        96\n",
       "experimental_evidence        80\n",
       "phylogenetically_inferred    44\n",
       "computational_analysis        8\n",
       "Name: evidence_category, dtype: int64"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go_ecoli_mf_secondary = df_go_ecoli_mf[\n",
    "    df_go_ecoli_mf.go_term == \"secondary active transmembrane transporter activity\"\n",
    "]\n",
    "df_go_ecoli_mf_secondary.evidence_category.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Is there overlap between the evidence codes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "non-iea annotations:\n",
      "92\n",
      "iea annotations:\n",
      "96\n",
      "overlap:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "127"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"non-iea annotations:\")\n",
    "print(\n",
    "    df_go_ecoli_mf_secondary[df_go_ecoli_mf_secondary.evidence_code != \"IEA\"][\n",
    "        [\"Uniprot\", \"go_id\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .shape[0]\n",
    ")\n",
    "print(\"iea annotations:\")\n",
    "print(\n",
    "    df_go_ecoli_mf_secondary[df_go_ecoli_mf_secondary.evidence_code == \"IEA\"][\n",
    "        [\"Uniprot\", \"go_id\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .shape[0]\n",
    ")\n",
    "print(\"overlap:\")\n",
    "pd.concat(\n",
    "    [\n",
    "        df_go_ecoli_mf_secondary[df_go_ecoli_mf_secondary.evidence_code != \"IEA\"][\n",
    "            [\"Uniprot\", \"go_id\"]\n",
    "        ],\n",
    "        df_go_ecoli_mf_secondary[df_go_ecoli_mf_secondary.evidence_code == \"IEA\"][\n",
    "            [\"Uniprot\", \"go_id\"]\n",
    "        ],\n",
    "    ]\n",
    ").drop_duplicates().shape[0]\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Yes, there is a considerable overlap. Including IEA terms increases the number of annotated proteins from 92 to 127."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "del(df_go_ecoli_mf_secondary)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many proteins in E coli are annotated with \"secondary active transmembrane transporter activity\"?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "127"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "secondary_active_transporters_go = set(\n",
    "    df_go_ecoli_mf[\n",
    "        df_go_ecoli_mf.go_term == \"secondary active transmembrane transporter activity\"\n",
    "    ].Uniprot.unique()\n",
    ")\n",
    "len(secondary_active_transporters_go)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Only about half as many compared to the TCDB class 2! Including all in the preprocessing above qualifiers did not change that. What is the distribution of proteins among the child terms of transmembrane transporter activity?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "480"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_transporters_go = set(\n",
    "    df_go_ecoli_mf[\n",
    "        df_go_ecoli_mf.go_term == \"transmembrane transporter activity\"\n",
    "    ].Uniprot.unique()\n",
    ")\n",
    "len(all_transporters_go)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Only half as many compared to the TCDB class 2! Including all qualifiers did not change that. What is the distribution of proteins among the child terms of transmembrane transporter activity?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_child_term_distribution(go_term:str):\n",
    "    # prints count of child terms of a particular go term, as dict\n",
    "    direct_child_terms = {\n",
    "        go_owl.get_label(go_id)\n",
    "        for go_id in go_owl.get_children(\n",
    "            go_owl.get_identifier(go_term)\n",
    "        )\n",
    "    }\n",
    "    direct_child_terms_count = {\n",
    "        child_term: df_go_ecoli_mf[df_go_ecoli_mf.go_term == child_term]\n",
    "        .Uniprot.unique()\n",
    "        .shape[0]\n",
    "        for child_term in direct_child_terms\n",
    "    }\n",
    "    return dict(sorted(direct_child_terms_count.items(), key=lambda item: item[1], reverse=True))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Distribution of child terms of \"transmembrane transporter activity\":"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'active transmembrane transporter activity': 240,\n",
       " 'ion transmembrane transporter activity': 196,\n",
       " 'inorganic molecular entity transmembrane transporter activity': 133,\n",
       " 'organic acid transmembrane transporter activity': 78,\n",
       " 'carbohydrate transmembrane transporter activity': 70,\n",
       " 'passive transmembrane transporter activity': 55,\n",
       " 'amide transmembrane transporter activity': 35,\n",
       " 'carbohydrate derivative transmembrane transporter activity': 29,\n",
       " 'xenobiotic transmembrane transporter activity': 27,\n",
       " 'efflux transmembrane transporter activity': 25,\n",
       " 'macromolecule transmembrane transporter activity': 17,\n",
       " 'organic hydroxy compound transmembrane transporter activity': 16,\n",
       " 'sulfur compound transmembrane transporter activity': 16,\n",
       " 'modified amino acid transmembrane transporter activity': 13,\n",
       " 'nucleobase transmembrane transporter activity': 12,\n",
       " 'polyamine transmembrane transporter activity': 12,\n",
       " 'siderophore-iron transmembrane transporter activity': 12,\n",
       " 'nucleobase-containing compound transmembrane transporter activity': 11,\n",
       " 'organophosphate ester transmembrane transporter activity': 9,\n",
       " 'heme transmembrane transporter activity': 6,\n",
       " 'vitamin transmembrane transporter activity': 5,\n",
       " 'azole transmembrane transporter activity': 4,\n",
       " 'iron chelate transmembrane transporter activity': 3,\n",
       " 'toxin transmembrane transporter activity': 2,\n",
       " 'amine transmembrane transporter activity': 2,\n",
       " 'salt transmembrane transporter activity': 1,\n",
       " 'poly-beta-1,6-N-acetyl-D-glucosamine transmembrane transporter activity': 1,\n",
       " 'bicyclomycin transmembrane transporter activity': 0,\n",
       " 'lactoferrin transmembrane transporter activity': 0,\n",
       " 'icosanoid transmembrane transporter activity': 0,\n",
       " 'mobile ion carrier activity': 0,\n",
       " 'neurotransmitter transmembrane transporter activity': 0,\n",
       " 'chlorophyll catabolite transmembrane transporter activity': 0,\n",
       " 'acetate ester transmembrane transporter activity': 0,\n",
       " 'alkane transmembrane transporter activity': 0,\n",
       " 'auxin transmembrane transporter activity': 0,\n",
       " 'lipo-chitin oligosaccharide transmembrane transporter activity': 0,\n",
       " 'oxidative phosphorylation uncoupler activity': 0,\n",
       " 'ferric triacetylfusarinine C transmembrane transporter activity': 0,\n",
       " 'organomercurial transmembrane transporter activity': 0,\n",
       " 'carbon dioxide transmembrane transporter activity': 0,\n",
       " 'biopterin transmembrane transporter activity': 0}"
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_child_term_distribution(\"transmembrane transporter activity\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The largest group is made up of active transporters. This includes secondary active, and primary active (e.g. the ATP-driven transport complexes). Some of these terms, like active and passive transport, are related to transport mechanism, while others, like \"carbohydrate transmembrane transporter activity\" and \"amide transmembrane transporter activity\" are related to substrate. The average protein is annotated with 2.14 of the child terms, so these are not mutually exclusive:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total annotations: 1030 annotations and 480 proteins\n"
     ]
    }
   ],
   "source": [
    "print(f\"total annotations: {sum(get_child_term_distribution('transmembrane transporter activity').values())} annotations and {len(all_transporters_go)} proteins\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Distribution of child terms of \"active transmembrane transporter activity\":"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'secondary active transmembrane transporter activity': 127,\n",
       " 'active ion transmembrane transporter activity': 107,\n",
       " 'primary active transmembrane transporter activity': 87,\n",
       " 'protein-N(PI)-phosphohistidine-sugar phosphotransferase activity': 25,\n",
       " 'protein-phosphocysteine-sugar phosphotransferase activity': 23,\n",
       " 'sorbitol transmembrane transporter activity': 0,\n",
       " 'monoamine transmembrane transporter activity': 0}"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_child_term_distribution(\"active transmembrane transporter activity\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Are the missing 2.A proteins distributed among the other non-primary types of active transport?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "active ion transmembrane transporter activity\n",
      "71\n",
      "sorbitol transmembrane transporter activity\n",
      "0\n",
      "primary active transmembrane transporter activity\n",
      "2\n",
      "secondary active transmembrane transporter activity\n",
      "111\n",
      "protein-N(PI)-phosphohistidine-sugar phosphotransferase activity\n",
      "0\n",
      "protein-phosphocysteine-sugar phosphotransferase activity\n",
      "0\n",
      "monoamine transmembrane transporter activity\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "tcdb_class2_proteins = set(df_tcdb_sparse_ecoli_class2.Uniprot.unique())\n",
    "for term in {go_owl.get_label(x) for x in go_owl.get_children(go_owl.get_identifier(\"active transmembrane transporter activity\"))}:\n",
    "    term_proteins = set(\n",
    "        df_go_ecoli_mf[df_go_ecoli_mf.go_term == term\n",
    "        ].Uniprot.unique()\n",
    "    )\n",
    "    print(term)\n",
    "    print(len(term_proteins & tcdb_class2_proteins))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "80"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(\n",
    "    df_go_ecoli_mf[\n",
    "        df_go_ecoli_mf.go_term == \"active ion transmembrane transporter activity\"\n",
    "    ].Uniprot.unique()\n",
    ")  & set(\n",
    "    df_go_ecoli_mf[\n",
    "        df_go_ecoli_mf.go_term == \"secondary active transmembrane transporter activity\"\n",
    "    ].Uniprot.unique()\n",
    ")& set(\n",
    "    df_go_ecoli_mf[\n",
    "        df_go_ecoli_mf.go_term == \"transmembrane transporter activity\"\n",
    "    ].Uniprot.unique()\n",
    ")\n",
    "\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "active ion transmembrane transporter activity\n",
      "71\n",
      "sorbitol transmembrane transporter activity\n",
      "0\n",
      "primary active transmembrane transporter activity\n",
      "2\n",
      "secondary active transmembrane transporter activity\n",
      "111\n",
      "protein-N(PI)-phosphohistidine-sugar phosphotransferase activity\n",
      "0\n",
      "protein-phosphocysteine-sugar phosphotransferase activity\n",
      "0\n",
      "monoamine transmembrane transporter activity\n",
      "0\n"
     ]
    }
   ],
   "source": [
    "tcdb_class2_proteins = set(df_tcdb_sparse_ecoli_class2.Uniprot.unique())\n",
    "for term in {go_owl.get_label(x) for x in go_owl.get_children(go_owl.get_identifier(\"active transmembrane transporter activity\"))}:\n",
    "    term_proteins = set(\n",
    "        df_go_ecoli_mf[df_go_ecoli_mf.go_term == term\n",
    "        ].Uniprot.unique()\n",
    "    )\n",
    "    print(term)\n",
    "    print(len(term_proteins & tcdb_class2_proteins))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Proteins with class 2 annotations in TCDB, but no \"secondary active transmembrane transporter activity\" go term:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_names</th>\n",
       "      <th>protein_names</th>\n",
       "      <th>protein_existence</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Uniprot</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>P0AD30</th>\n",
       "      <td>yfcA b2327 JW2324</td>\n",
       "      <td>Probable membrane transporter protein YfcA</td>\n",
       "      <td>1</td>\n",
       "      <td>METFNSLFMVSPLLLGVLFFVAMLAGFIDSIAGGGGLLTIPALMAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P37746</th>\n",
       "      <td>rfbX wzx b2037 JW2022</td>\n",
       "      <td>Putative O-antigen transporter</td>\n",
       "      <td>1</td>\n",
       "      <td>MNTNKLSLRRNVIYLAVVQGSNYLLPLLTFPYLVRTLGPENFGIFG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P0ADD2</th>\n",
       "      <td>yjjB b4363 JW4327</td>\n",
       "      <td>Probable succinate transporter subunit YjjB</td>\n",
       "      <td>1</td>\n",
       "      <td>MGVIEFLLALAQDMILAAIPAVGFAMVFNVPVRALRWCALLGSIGH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P0ABG7</th>\n",
       "      <td>mrdB rodA b0634 JW0629</td>\n",
       "      <td>Peptidoglycan glycosyltransferase MrdB (PGT) (...</td>\n",
       "      <td>1</td>\n",
       "      <td>MTDNPNKKTFWDKVHLDPTMLLILLALLVYSALVIWSASGQDIGMM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P36554</th>\n",
       "      <td>mdtD yegB b2077 JW2062</td>\n",
       "      <td>Putative multidrug resistance protein MdtD</td>\n",
       "      <td>1</td>\n",
       "      <td>MTDLPDSTRWQLWIVAFGFFMQSLDTTIVNTALPSMAQSLGESPLH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P0AB24</th>\n",
       "      <td>efeO ycdO b1018 JW1003</td>\n",
       "      <td>Iron uptake system component EfeO</td>\n",
       "      <td>1</td>\n",
       "      <td>MTINFRRNALQLSVAALFSSAFMANAADVPQVKVTVTDKQCEPMTI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P38055</th>\n",
       "      <td>ydjE b1769 JW1758</td>\n",
       "      <td>Inner membrane metabolite transport protein YdjE</td>\n",
       "      <td>1</td>\n",
       "      <td>MEQYDQIGARLDRLPLARFHYRIFGIISFSLLLTGFLSYSGNVVLA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P77389</th>\n",
       "      <td>ydhP b1657 JW1649</td>\n",
       "      <td>Inner membrane transport protein YdhP</td>\n",
       "      <td>1</td>\n",
       "      <td>MKINYPLLALAIGAFGIGTTEFSPMGLLPVIARGVDVSIPAAGMLI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P0AA70</th>\n",
       "      <td>yedA b1959 JW1942</td>\n",
       "      <td>Uncharacterized inner membrane transporter YedA</td>\n",
       "      <td>1</td>\n",
       "      <td>MRFRQLLPLFGALFALYIIWGSTYFVIRIGVESWPPLMMAGVRFLA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P21503</th>\n",
       "      <td>ycaD b0898 JW0881</td>\n",
       "      <td>Uncharacterized MFS-type transporter YcaD</td>\n",
       "      <td>1</td>\n",
       "      <td>MSTYTQPVMLLLSGLLLLTLAIAVLNTLVPLWLAQEHMSTWQVGVV...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>127 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene_names  \\\n",
       "Uniprot                           \n",
       "P0AD30        yfcA b2327 JW2324   \n",
       "P37746    rfbX wzx b2037 JW2022   \n",
       "P0ADD2        yjjB b4363 JW4327   \n",
       "P0ABG7   mrdB rodA b0634 JW0629   \n",
       "P36554   mdtD yegB b2077 JW2062   \n",
       "...                         ...   \n",
       "P0AB24   efeO ycdO b1018 JW1003   \n",
       "P38055        ydjE b1769 JW1758   \n",
       "P77389        ydhP b1657 JW1649   \n",
       "P0AA70        yedA b1959 JW1942   \n",
       "P21503        ycaD b0898 JW0881   \n",
       "\n",
       "                                             protein_names  protein_existence  \\\n",
       "Uniprot                                                                         \n",
       "P0AD30          Probable membrane transporter protein YfcA                  1   \n",
       "P37746                      Putative O-antigen transporter                  1   \n",
       "P0ADD2         Probable succinate transporter subunit YjjB                  1   \n",
       "P0ABG7   Peptidoglycan glycosyltransferase MrdB (PGT) (...                  1   \n",
       "P36554          Putative multidrug resistance protein MdtD                  1   \n",
       "...                                                    ...                ...   \n",
       "P0AB24                   Iron uptake system component EfeO                  1   \n",
       "P38055    Inner membrane metabolite transport protein YdjE                  1   \n",
       "P77389               Inner membrane transport protein YdhP                  1   \n",
       "P0AA70     Uncharacterized inner membrane transporter YedA                  1   \n",
       "P21503           Uncharacterized MFS-type transporter YcaD                  1   \n",
       "\n",
       "                                                  sequence  \n",
       "Uniprot                                                     \n",
       "P0AD30   METFNSLFMVSPLLLGVLFFVAMLAGFIDSIAGGGGLLTIPALMAA...  \n",
       "P37746   MNTNKLSLRRNVIYLAVVQGSNYLLPLLTFPYLVRTLGPENFGIFG...  \n",
       "P0ADD2   MGVIEFLLALAQDMILAAIPAVGFAMVFNVPVRALRWCALLGSIGH...  \n",
       "P0ABG7   MTDNPNKKTFWDKVHLDPTMLLILLALLVYSALVIWSASGQDIGMM...  \n",
       "P36554   MTDLPDSTRWQLWIVAFGFFMQSLDTTIVNTALPSMAQSLGESPLH...  \n",
       "...                                                    ...  \n",
       "P0AB24   MTINFRRNALQLSVAALFSSAFMANAADVPQVKVTVTDKQCEPMTI...  \n",
       "P38055   MEQYDQIGARLDRLPLARFHYRIFGIISFSLLLTGFLSYSGNVVLA...  \n",
       "P77389   MKINYPLLALAIGAFGIGTTEFSPMGLLPVIARGVDVSIPAAGMLI...  \n",
       "P0AA70   MRFRQLLPLFGALFALYIIWGSTYFVIRIGVESWPPLMMAGVRFLA...  \n",
       "P21503   MSTYTQPVMLLLSGLLLLTLAIAVLNTLVPLWLAQEHMSTWQVGVV...  \n",
       "\n",
       "[127 rows x 4 columns]"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "term_proteins = set(df_go_ecoli_mf[df_go_ecoli_mf.go_term == \"secondary active transmembrane transporter activity\"].Uniprot.unique())\n",
    "df_uniprot_ecoli_non_go_tcdb = df_uniprot_ecoli.loc[list(tcdb_class2_proteins - term_proteins)]\n",
    "df_uniprot_ecoli_non_go_tcdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "238\n",
      "127\n",
      "127\n",
      "111\n"
     ]
    }
   ],
   "source": [
    "print(len(tcdb_class2_proteins))\n",
    "print(len(term_proteins))\n",
    "print(len(tcdb_class2_proteins - term_proteins))\n",
    "print(len(tcdb_class2_proteins & term_proteins))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's look at one protein as an example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>tcdb_id</th>\n",
       "      <th>substrate</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>2.A.108.2.3</td>\n",
       "      <td>iron(2+)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Uniprot      tcdb_id substrate\n",
       "62  P0AB24  2.A.108.2.3  iron(2+)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tcdb_sparse_ecoli_class2[df_tcdb_sparse_ecoli_class2.Uniprot ==\"P0AB24\" ]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Uniprot description: Involved in Fe2+ uptake. Could be an iron-binding and/or electron-transfer component\n",
    "\n",
    "The TCDB system description is: \"The acid-induced, low pH, ferrous iron (Fe2+) uptake transporter, EfeUOB (YcdN); cryptic in E. coli K12 (Cao et al. 2007; Grosse et al., 2006).\". Could it be that these are not transmembrane proteins, but transporters in the periplasmic space? GO terms of  P0AB24:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>aspect</th>\n",
       "      <th>go_id</th>\n",
       "      <th>go_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>83137</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IMP</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0006950</td>\n",
       "      <td>response to stress</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83138</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IMP</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0006979</td>\n",
       "      <td>response to oxidative stress</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83139</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IMP</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0008150</td>\n",
       "      <td>biological_process</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83140</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IMP</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0050896</td>\n",
       "      <td>response to stimulus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83141</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IMP</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0009314</td>\n",
       "      <td>response to radiation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83142</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IMP</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0009416</td>\n",
       "      <td>response to light stimulus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83143</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IMP</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0009628</td>\n",
       "      <td>response to abiotic stimulus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83144</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IMP</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0009411</td>\n",
       "      <td>response to UV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83145</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IMP</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0042221</td>\n",
       "      <td>response to chemical</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83146</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IMP</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0009636</td>\n",
       "      <td>response to toxic substance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83147</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IMP</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0046677</td>\n",
       "      <td>response to antibiotic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83148</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>located_in</td>\n",
       "      <td>IMP</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0042597</td>\n",
       "      <td>periplasmic space</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83149</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>located_in</td>\n",
       "      <td>IMP</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0110165</td>\n",
       "      <td>cellular anatomical entity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83150</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>located_in</td>\n",
       "      <td>IMP</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0030288</td>\n",
       "      <td>outer membrane-bounded periplasmic space</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83151</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>located_in</td>\n",
       "      <td>IMP</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0005575</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83152</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>located_in</td>\n",
       "      <td>IEA</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0042597</td>\n",
       "      <td>periplasmic space</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83153</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>located_in</td>\n",
       "      <td>IEA</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0110165</td>\n",
       "      <td>cellular anatomical entity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83154</th>\n",
       "      <td>P0AB24</td>\n",
       "      <td>located_in</td>\n",
       "      <td>IEA</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0005575</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Uniprot                   qualifier evidence_code aspect       go_id  \\\n",
       "83137  P0AB24  acts_upstream_of_or_within           IMP      P  GO:0006950   \n",
       "83138  P0AB24  acts_upstream_of_or_within           IMP      P  GO:0006979   \n",
       "83139  P0AB24  acts_upstream_of_or_within           IMP      P  GO:0008150   \n",
       "83140  P0AB24  acts_upstream_of_or_within           IMP      P  GO:0050896   \n",
       "83141  P0AB24  acts_upstream_of_or_within           IMP      P  GO:0009314   \n",
       "83142  P0AB24  acts_upstream_of_or_within           IMP      P  GO:0009416   \n",
       "83143  P0AB24  acts_upstream_of_or_within           IMP      P  GO:0009628   \n",
       "83144  P0AB24  acts_upstream_of_or_within           IMP      P  GO:0009411   \n",
       "83145  P0AB24  acts_upstream_of_or_within           IMP      P  GO:0042221   \n",
       "83146  P0AB24  acts_upstream_of_or_within           IMP      P  GO:0009636   \n",
       "83147  P0AB24  acts_upstream_of_or_within           IMP      P  GO:0046677   \n",
       "83148  P0AB24                  located_in           IMP      C  GO:0042597   \n",
       "83149  P0AB24                  located_in           IMP      C  GO:0110165   \n",
       "83150  P0AB24                  located_in           IMP      C  GO:0030288   \n",
       "83151  P0AB24                  located_in           IMP      C  GO:0005575   \n",
       "83152  P0AB24                  located_in           IEA      C  GO:0042597   \n",
       "83153  P0AB24                  located_in           IEA      C  GO:0110165   \n",
       "83154  P0AB24                  located_in           IEA      C  GO:0005575   \n",
       "\n",
       "                                        go_term  \n",
       "83137                        response to stress  \n",
       "83138              response to oxidative stress  \n",
       "83139                        biological_process  \n",
       "83140                      response to stimulus  \n",
       "83141                     response to radiation  \n",
       "83142                response to light stimulus  \n",
       "83143              response to abiotic stimulus  \n",
       "83144                            response to UV  \n",
       "83145                      response to chemical  \n",
       "83146               response to toxic substance  \n",
       "83147                    response to antibiotic  \n",
       "83148                         periplasmic space  \n",
       "83149                cellular anatomical entity  \n",
       "83150  outer membrane-bounded periplasmic space  \n",
       "83151                        cellular_component  \n",
       "83152                         periplasmic space  \n",
       "83153                cellular anatomical entity  \n",
       "83154                        cellular_component  "
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go_ecoli[df_go_ecoli.Uniprot == \"P0AB24\"].assign(go_term = df_go_ecoli.go_id.transform(go_owl.get_label))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This protein is in periplasmic space! What about the cellular locations of the other proteins?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uniprot</th>\n",
       "      <th>qualifier</th>\n",
       "      <th>evidence_code</th>\n",
       "      <th>aspect</th>\n",
       "      <th>go_id</th>\n",
       "      <th>go_term</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8573</th>\n",
       "      <td>P02929</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IGI</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0015889</td>\n",
       "      <td>cobalamin transport</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8574</th>\n",
       "      <td>P02929</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IGI</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0071702</td>\n",
       "      <td>organic substance transport</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8575</th>\n",
       "      <td>P02929</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IGI</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0071705</td>\n",
       "      <td>nitrogen compound transport</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8576</th>\n",
       "      <td>P02929</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IGI</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0006810</td>\n",
       "      <td>transport</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8577</th>\n",
       "      <td>P02929</td>\n",
       "      <td>acts_upstream_of_or_within</td>\n",
       "      <td>IGI</td>\n",
       "      <td>P</td>\n",
       "      <td>GO:0051179</td>\n",
       "      <td>localization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272046</th>\n",
       "      <td>Q47377</td>\n",
       "      <td>located_in</td>\n",
       "      <td>IDA</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0005886</td>\n",
       "      <td>plasma membrane</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272047</th>\n",
       "      <td>Q47377</td>\n",
       "      <td>located_in</td>\n",
       "      <td>IEA</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0016020</td>\n",
       "      <td>membrane</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272048</th>\n",
       "      <td>Q47377</td>\n",
       "      <td>located_in</td>\n",
       "      <td>IEA</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0110165</td>\n",
       "      <td>cellular anatomical entity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272049</th>\n",
       "      <td>Q47377</td>\n",
       "      <td>located_in</td>\n",
       "      <td>IEA</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0005575</td>\n",
       "      <td>cellular_component</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272050</th>\n",
       "      <td>Q47377</td>\n",
       "      <td>located_in</td>\n",
       "      <td>IEA</td>\n",
       "      <td>C</td>\n",
       "      <td>GO:0005886</td>\n",
       "      <td>plasma membrane</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6877 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Uniprot                   qualifier evidence_code aspect       go_id  \\\n",
       "8573    P02929  acts_upstream_of_or_within           IGI      P  GO:0015889   \n",
       "8574    P02929  acts_upstream_of_or_within           IGI      P  GO:0071702   \n",
       "8575    P02929  acts_upstream_of_or_within           IGI      P  GO:0071705   \n",
       "8576    P02929  acts_upstream_of_or_within           IGI      P  GO:0006810   \n",
       "8577    P02929  acts_upstream_of_or_within           IGI      P  GO:0051179   \n",
       "...        ...                         ...           ...    ...         ...   \n",
       "272046  Q47377                  located_in           IDA      C  GO:0005886   \n",
       "272047  Q47377                  located_in           IEA      C  GO:0016020   \n",
       "272048  Q47377                  located_in           IEA      C  GO:0110165   \n",
       "272049  Q47377                  located_in           IEA      C  GO:0005575   \n",
       "272050  Q47377                  located_in           IEA      C  GO:0005886   \n",
       "\n",
       "                            go_term  \n",
       "8573            cobalamin transport  \n",
       "8574    organic substance transport  \n",
       "8575    nitrogen compound transport  \n",
       "8576                      transport  \n",
       "8577                   localization  \n",
       "...                             ...  \n",
       "272046              plasma membrane  \n",
       "272047                     membrane  \n",
       "272048   cellular anatomical entity  \n",
       "272049           cellular_component  \n",
       "272050              plasma membrane  \n",
       "\n",
       "[6877 rows x 6 columns]"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go_ecoli_non_go_tcdb = df_go_ecoli[df_go_ecoli.Uniprot.isin(df_uniprot_ecoli_non_go_tcdb.index.tolist())].assign(go_term = df_go_ecoli.go_id.transform(go_owl.get_label))\n",
    "df_go_ecoli_non_go_tcdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cellular_component                                            366\n",
       "cellular anatomical entity                                    346\n",
       "membrane                                                      324\n",
       "plasma membrane                                               313\n",
       "protein-containing complex                                     23\n",
       "periplasmic space                                              18\n",
       "outer membrane-bounded periplasmic space                       13\n",
       "membrane protein complex                                       13\n",
       "transporter complex                                            11\n",
       "transmembrane transporter complex                               8\n",
       "cell division site                                              7\n",
       "efflux pump complex                                             5\n",
       "plasma membrane protein complex                                 5\n",
       "intracellular protein-containing complex                        4\n",
       "cell envelope Sec protein transport complex                     4\n",
       "TAT protein transport complex                                   3\n",
       "cytosol                                                         3\n",
       "tripartite ATP-independent periplasmic transporter complex      2\n",
       "side of membrane                                                2\n",
       "external side of plasma membrane                                2\n",
       "cell outer membrane                                             2\n",
       "outer membrane                                                  2\n",
       "potassium:proton antiporter complex                             1\n",
       "cell envelope                                                   1\n",
       "divisome complex                                                1\n",
       "division septum                                                 1\n",
       "cell septum                                                     1\n",
       "extrinsic component of cytoplasmic side of plasma membrane      1\n",
       "extrinsic component of membrane                                 1\n",
       "extrinsic component of plasma membrane                          1\n",
       "envelope                                                        1\n",
       "Name: go_term, dtype: int64"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_go_ecoli_non_go_tcdb[df_go_ecoli_non_go_tcdb.aspect == \"C\"].go_term.value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No, most proteins are in plasma membrane. The dataframe can contain the same annotation multiple times with different evidence codes, hence the high number. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "molecular_function                                                                     226\n",
      "transporter activity                                                                   177\n",
      "transmembrane transporter activity                                                     176\n",
      "ion transmembrane transporter activity                                                  48\n",
      "binding                                                                                 40\n",
      "cation transmembrane transporter activity                                               35\n",
      "carboxylic acid transmembrane transporter activity                                      34\n",
      "organic acid transmembrane transporter activity                                         34\n",
      "anion transmembrane transporter activity                                                28\n",
      "organic anion transmembrane transporter activity                                        28\n",
      "amino acid transmembrane transporter activity                                           26\n",
      "protein binding                                                                         22\n",
      "efflux transmembrane transporter activity                                               17\n",
      "nucleobase transmembrane transporter activity                                           17\n",
      "L-amino acid transmembrane transporter activity                                         16\n",
      "inorganic molecular entity transmembrane transporter activity                           15\n",
      "metal ion transmembrane transporter activity                                            15\n",
      "inorganic cation transmembrane transporter activity                                     15\n",
      "transition metal ion transmembrane transporter activity                                 14\n",
      "carbohydrate derivative transmembrane transporter activity                              13\n",
      "xenobiotic transmembrane transporter activity                                           13\n",
      "macromolecule transmembrane transporter activity                                        11\n",
      "purine nucleobase transmembrane transporter activity                                    11\n",
      "ion binding                                                                             11\n",
      "catalytic activity                                                                      10\n",
      "aldonate transmembrane transporter activity                                             10\n",
      "gluconate transmembrane transporter activity                                             9\n",
      "neutral L-amino acid transmembrane transporter activity                                  8\n",
      "lipid-linked peptidoglycan transporter activity                                          7\n",
      "peptidoglycan transmembrane transporter activity                                         7\n",
      "monocarboxylic acid transmembrane transporter activity                                   7\n",
      "metal ion binding                                                                        7\n",
      "organic cyclic compound binding                                                          7\n",
      "heterocyclic compound binding                                                            7\n",
      "L-amino acid efflux transmembrane transporter activity                                   7\n",
      "oxidoreductase activity                                                                  7\n",
      "cation binding                                                                           7\n",
      "basic amino acid transmembrane transporter activity                                      6\n",
      "carbohydrate transmembrane transporter activity                                          6\n",
      "molecular carrier activity                                                               5\n",
      "xanthine transmembrane transporter activity                                              5\n",
      "zinc ion transmembrane transporter activity                                              5\n",
      "organic hydroxy compound transmembrane transporter activity                              4\n",
      "nucleotide binding                                                                       4\n",
      "small molecule binding                                                                   4\n",
      "energy transducer activity                                                               4\n",
      "lipid transporter activity                                                               4\n",
      "L-threonine transmembrane transporter activity                                           4\n",
      "threonine efflux transmembrane transporter activity                                      4\n",
      "molecular transducer activity                                                            4\n",
      "identical protein binding                                                                4\n",
      "transition metal ion binding                                                             4\n",
      "protein transporter activity                                                             4\n",
      "L-lysine transmembrane transporter activity                                              4\n",
      "nucleoside phosphate binding                                                             4\n",
      "protein carrier chaperone                                                                4\n",
      "membrane insertase activity                                                              4\n",
      "anion binding                                                                            4\n",
      "cation efflux transmembrane transporter activity                                         4\n",
      "protein transmembrane transporter activity                                               4\n",
      "guanine transmembrane transporter activity                                               4\n",
      "peroxidase activity                                                                      3\n",
      "oxidoreductase activity, acting on NAD(P)H                                               3\n",
      "tetrapyrrole binding                                                                     3\n",
      "heme binding                                                                             3\n",
      "homoserine transmembrane transporter activity                                            3\n",
      "nickel cation transmembrane transporter activity                                         3\n",
      "NAD(P)H dehydrogenase (quinone) activity                                                 3\n",
      "antioxidant activity                                                                     3\n",
      "oxidoreductase activity, acting on peroxide as acceptor                                  3\n",
      "protein domain specific binding                                                          3\n",
      "aromatic amino acid transmembrane transporter activity                                   3\n",
      "oxidoreductase activity, acting on NAD(P)H, quinone or similar compound as acceptor      3\n",
      "branched-chain amino acid transmembrane transporter activity                             3\n",
      "bile acid transmembrane transporter activity                                             3\n",
      "copper ion binding                                                                       3\n",
      "carbohydrate derivative binding                                                          3\n",
      "ribonucleotide binding                                                                   3\n",
      "manganese ion transmembrane transporter activity                                         3\n",
      "FMN binding                                                                              3\n",
      "L-lysine efflux transmembrane transporter activity                                       3\n",
      "ATP-dependent activity                                                                   2\n",
      "protein-transporting ATPase activity                                                     2\n",
      "ATPase-coupled transmembrane transporter activity                                        2\n",
      "monosaccharide transmembrane transporter activity                                        2\n",
      "L-tyrosine transmembrane transporter activity                                            2\n",
      "L-arginine transmembrane transporter activity                                            2\n",
      "L-phenylalanine transmembrane transporter activity                                       2\n",
      "adenine transmembrane transporter activity                                               2\n",
      "nucleoside transmembrane transporter activity                                            2\n",
      "active transmembrane transporter activity                                                2\n",
      "primary active transmembrane transporter activity                                        2\n",
      "uronic acid transmembrane transporter activity                                           2\n",
      "carbohydrate binding                                                                     2\n",
      "L-valine transmembrane transporter activity                                              2\n",
      "electron transfer activity                                                               2\n",
      "hexosyltransferase activity                                                              2\n",
      "L-isoleucine transmembrane transporter activity                                          2\n",
      "L-leucine transmembrane transporter activity                                             2\n",
      "glycosyltransferase activity                                                             2\n",
      "peptidoglycan glycosyltransferase activity                                               2\n",
      "sugar transmembrane transporter activity                                                 2\n",
      "transferase activity                                                                     2\n",
      "L-idonate transmembrane transporter activity                                             2\n",
      "hexuronate transmembrane transporter activity                                            2\n",
      "nucleobase-containing compound transmembrane transporter activity                        2\n",
      "lyase activity                                                                           1\n",
      "ferrochelatase activity                                                                  1\n",
      "C4-dicarboxylate transmembrane transporter activity                                      1\n",
      "uridine transmembrane transporter activity                                               1\n",
      "silver ion transmembrane transporter activity                                            1\n",
      "metallochaperone activity                                                                1\n",
      "benzoate transmembrane transporter activity                                              1\n",
      "cytidine transmembrane transporter activity                                              1\n",
      "pyrimidine nucleoside transmembrane transporter activity                                 1\n",
      "hydrolase activity                                                                       1\n",
      "siderophore-iron transmembrane transporter activity                                      1\n",
      "enterobactin transmembrane transporter activity                                          1\n",
      "D-serine transmembrane transporter activity                                              1\n",
      "serine transmembrane transporter activity                                                1\n",
      "D-amino acid transmembrane transporter activity                                          1\n",
      "NADPH dehydrogenase (quinone) activity                                                   1\n",
      "NADH dehydrogenase activity                                                              1\n",
      "NADH dehydrogenase (quinone) activity                                                    1\n",
      "protein-containing complex binding                                                       1\n",
      "copper ion transmembrane transporter activity                                            1\n",
      "cadmium ion transmembrane transporter activity                                           1\n",
      "cobalt ion transmembrane transporter activity                                            1\n",
      "ferrous iron transmembrane transporter activity                                          1\n",
      "iron ion transmembrane transporter activity                                              1\n",
      "intramembrane lipid transporter activity                                                 1\n",
      "organic cation transmembrane transporter activity                                        1\n",
      "choline transmembrane transporter activity                                               1\n",
      "sulfuric ester hydrolase activity                                                        1\n",
      "hydrolase activity, acting on ester bonds                                                1\n",
      "lipid binding                                                                            1\n",
      "phosphatidylglycerol binding                                                             1\n",
      "cardiolipin binding                                                                      1\n",
      "phospholipid binding                                                                     1\n",
      "potassium ion transmembrane transporter activity                                         1\n",
      "NAD binding                                                                              1\n",
      "adenyl nucleotide binding                                                                1\n",
      "purine nucleotide binding                                                                1\n",
      "L-tryptophan transmembrane transporter activity                                          1\n",
      "toxin transmembrane transporter activity                                                 1\n",
      "virion binding                                                                           1\n",
      "disordered domain specific binding                                                       1\n",
      "copper chaperone activity                                                                1\n"
     ]
    }
   ],
   "source": [
    "print(df_go_ecoli_non_go_tcdb[df_go_ecoli_non_go_tcdb.aspect == \"F\"].go_term.value_counts().to_string())"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When looking at the Uniprot entries of some examples of inner-membrane proteins that have a TCDB 2.A annotation but not a \"secondary active transmembrane transporter activity\" annotation, it looks like these proteins mostly lack experimental data to receive a \"molecular function\" go term, but they are similar enough in terms of sequence and structure to assign them to a TCDB mechanism. Some of the proteins are annotated only with \"transmembrane transporter activity\" but nothing more specific. Others are only annotated with a substrate (e.g. amino acid transmembrane transporter activity) but not with a mechanism. This could be problematic when filtering by mechanism GO annotation first. Filtering by TCDB is not an option for most organisms, since *E. coli* has the highest percantage of annotations, along with yeast. The third category are transporters in periplasmic space, which are apparently still labeled as membrane transporters. "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Distribution of child terms of \"secondary active transmembrane transporter activity\":"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'symporter activity': 82,\n",
       " 'antiporter activity': 35,\n",
       " 'inorganic phosphate transmembrane transporter activity': 6,\n",
       " 'secondary active monocarboxylate transmembrane transporter activity': 4,\n",
       " 'proton motive force dependent protein transmembrane transporter activity': 3,\n",
       " 'uniporter activity': 2,\n",
       " 'nitrite efflux transmembrane transporter activity': 2,\n",
       " 'secondary active sulfate transmembrane transporter activity': 2,\n",
       " 'neutral L-amino acid secondary active transmembrane transporter activity': 1,\n",
       " 'antimonite secondary active transmembrane transporter activity': 1,\n",
       " 'sulfur-containing amino acid secondary active transmembrane transporter activity': 1,\n",
       " 'secondary active p-aminobenzoyl-glutamate transmembrane transporter activity': 1,\n",
       " 'glutathione-regulated potassium exporter activity': 1,\n",
       " 'arsenite secondary active transmembrane transporter activity': 1,\n",
       " 'zinc efflux active transmembrane transporter activity': 1,\n",
       " 'oxaloacetate secondary active transmembrane transporter activity': 0,\n",
       " 'proton-dependent peptide secondary active transmembrane transporter activity': 0,\n",
       " 'thyroid hormone transmembrane transporter activity': 0,\n",
       " 'nucleoside transmembrane transporter activity, against a concentration gradient': 0,\n",
       " 'tricarboxylate secondary active transmembrane transporter activity': 0,\n",
       " 'secondary active cyanate transmembrane transporter activity': 0,\n",
       " 'bilirubin secondary active transmembrane transporter activity': 0,\n",
       " 'low-affinity secondary active ammonium transmembrane transporter activity': 0,\n",
       " 'secondary active organic cation transmembrane transporter activity': 0,\n",
       " 'secondary active oligopeptide transmembrane transporter activity': 0,\n",
       " 'sodium-independent organic anion transmembrane transporter activity': 0,\n",
       " 'peptide-acetyl-CoA secondary active transmembrane transporter activity': 0,\n",
       " 'high-affinity secondary active nitrite transmembrane transporter activity': 0,\n",
       " 'high-affinity secondary active ammonium transmembrane transporter activity': 0}"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_child_term_distribution(\"secondary active transmembrane transporter activity\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "subpred4",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
